---
title: "R for Mass Spectrometry"
subtitle: "Applications in Proteomics and Metabolomics"
author: "Laurent Gatto, Sebastian Gibb, Johannes Rainer"
date: "`r Sys.Date()`"
output:
  msmbstyle::msmb_html_book:
    highlight: tango
    toc: TRUE
    toc_depth: 1
    split_by: chapter
    margin_references: TRUE
    css: style.css
bibliography: [refs.bib, packages.bib]
link-citations: yes
---

# Preamble

```{r, echo = FALSE}
options(bitmapType="cairo")

```

The aim of the [R for Mass
Spectrometry](https://www.rformassspectrometry.org/) initiative is to
provide efficient, thoroughly documented, tested and flexible R
software for the analysis and interpretation of high throughput mass
spectrometry assays, including proteomics and metabolomics
experiments. The project formalises the longtime collaborative
development efforts of its core members under the RforMassSpectrometry
organisation to facilitate dissemination and accessibility of their
work.

```{r sticker, fig.cap = "The *R for Mass Spectrometry* intiative sticker, designed by Johannes Rainer.", out.width = '50%', fig.margin=TRUE, echo=FALSE}
knitr::include_graphics("https://github.com/rformassspectrometry/stickers/raw/master/sticker/RforMassSpectrometry.png")
```

This material introduces participants to the analysis and exploration
of mass spectrometry (MS) based proteomics data using R and
Bioconductor. The course will cover all levels of MS data, from raw
data to identification and quantitation data, up to the statistical
interpretation of a typical shotgun MS experiment and will focus on
hands-on tutorials. At the end of this course, the participants will
be able to manipulate MS data in R and use existing packages for their
exploratory and statistical proteomics data analysis.

## Targeted audience and assumed background

The course material is targeted to either proteomics practitioners or
data analysts/bioinformaticians that would like to learn how to use R
and Bioconductor to analyse proteomics data. Familiarity with MS or
proteomics in general is desirable, but not essential as we will walk
through and describe a typical MS data as part of learning about the
tools. For approachable introductions to sample preparation, mass
spectrometry, data interpretation and analysis, readers are redirected
to:

- *A beginner's guide to mass spectrometryâ€“based proteomics* [@Sinha:2020]
- *The ABC's (and XYZ's) of peptide sequencing* [@Steen:2004]
- *How do shotgun proteomics algorithms identify proteins?* [@Marcotte:2007]
- *An Introduction to Mass Spectrometry-Based Proteomics* [@Shuken:2023]

A working knowledge of R (R syntax, commonly used functions, basic
data structures such as data frames, vectors, matrices, ... and their
manipulation) is required. Familiarity with other Bioconductor omics
data classes and the tidyverse syntax is useful, but not necessary.


```{r bib, include=FALSE}
# create a bib file for the R packages used in this document
knitr::write_bib(c('base', 'rmarkdown', 'bookdown', 'msmbstyle'), file = 'skeleton.bib')
```

```{r env_0, echo = FALSE, message = FALSE, warning = FALSE}
suppressPackageStartupMessages(library("BiocStyle"))
suppressPackageStartupMessages(library("mzR"))
suppressPackageStartupMessages(library("Spectra"))
suppressPackageStartupMessages(library("QFeatures"))
suppressPackageStartupMessages(library("MsCoreUtils"))
```

## Setup

This material uses the latest version of the R for Mass Spectrometry
package and their dependencies. It might thus be possible that even
the latest Bioconductor stable version isn't recent enough.

To install all the necessary package, please use the latest release of
R and execute:

```{r setup, eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("remotes")
BiocManager::install("tidyverse")
BiocManager::install("factoextra")
BiocManager::install("MsDataHub")
BiocManager::install("mzR")
BiocManager::install("rhdf5")
BiocManager::install("rpx")
BiocManager::install("MsCoreUtils")
BiocManager::install("QFeatures")
BiocManager::install("Spectra")
BiocManager::install("ProtGenerics")
BiocManager::install("PSMatch")
BiocManager::install("pheatmap")
BiocManager::install("limma")
BiocManager::install("MSnID")
BiocManager::install("RforMassSpectrometry/SpectraVis")
```

After installation, you can download some data that will be used in
the latter chapter running the following:

```{r downloaddata, eval = FALSE}
library(rpx)
px <- PXDataset("PXD000001") ## answer yes if asked to create a cache directory
fn <- "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML"
mzf <- pxget(px, fn)
px <- PXDataset("PXD022816")
pxget(px, grep("mzID", pxfiles(px))[1:3])
pxget(px, grep("mzML", pxfiles(px))[1:3])
```


All software versions used to generate this document are recoded at
the end of the book in chapter \@ref(sec-si).

## Questions and help

For questions about specific software or their usage, please refer to
the software's github issue page, or use the [Bioconductor support
site](http://support.bioconductor.org/).

## Acknowledgments

Thank you to [Charlotte Soneson](https://github.com/csoneson) for
fixing many typos in a previous version of this book.

## License

<a rel="license"
href="http://creativecommons.org/licenses/by-sa/4.0/"><img
alt="Creative Commons Licence" style="border-width:0"
src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a><br
/>This material is licensed under a <a rel="license"
href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons
Attribution-ShareAlike 4.0 International License</a>. You are free to
**share** (copy and redistribute the material in any medium or format)
and **adapt** (remix, transform, and build upon the material) for any
purpose, even commercially, as long as you give appropriate credit and
distribute your contributions under the same license as the original.