<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 1 Preamble | R for Mass Spectrometry" />
<meta property="og:type" content="book" />




<meta name="author" content="Laurent Gatto, Sebastian Gibb, Johannes Rainer" />

<meta name="date" content="2025-06-05" />


<meta name="description" content="Chapter 1 Preamble | R for Mass Spectrometry">

<title>Chapter 1 Preamble | R for Mass Spectrometry</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>
<script>
function copy_link(id) {
  var dummy = document.createElement('input'),
  text = window.location.href.split(/[?#]/)[0] + '#' + id;
  document.body.appendChild(dummy);
  dummy.value = text;
  dummy.select();
  document.execCommand('copy');
  document.body.removeChild(dummy);
  
  var tooltip = document.getElementById(id + '-tooltip');
  tooltip.innerHTML = 'Copied!';
}

function reset_tooltip(id) {
  var tooltip = document.getElementById(id);
  tooltip.innerHTML = 'Copy link';
}
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }

code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #204a87; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #204a87; font-weight: bold; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #ce5c00; font-weight: bold; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">R for Mass Spectrometry<p><p class="author">Laurent Gatto, Sebastian Gibb, Johannes Rainer</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a id="active-page" href="index.html" id="toc-preamble"><span class="toc-section-number">1</span> Preamble</a><ul class="toc-sections">
<li class="toc"><a href="#targeted-audience-and-assumed-background"> Targeted audience and assumed background</a></li>
<li class="toc"><a href="#setup"> Setup</a></li>
<li class="toc"><a href="#questions-and-help"> Questions and help</a></li>
<li class="toc"><a href="#citation"> Citation</a></li>
<li class="toc"><a href="#acknowledgments"> Acknowledgments</a></li>
<li class="toc"><a href="#license"> License</a></li>
</ul>
<a href="sec-msintro.html" id="toc-sec-msintro"><span class="toc-section-number">2</span> Introduction</a>
<a href="sec-raw.html" id="toc-sec-raw"><span class="toc-section-number">3</span> Raw MS data</a>
<a href="sec-id.html" id="toc-sec-id"><span class="toc-section-number">4</span> Identification data</a>
<a href="sec-quant.html" id="toc-sec-quant"><span class="toc-section-number">5</span> Quantitative data</a>
<a href="sec-anx.html" id="toc-sec-anx"><span class="toc-section-number">6</span> Annex</a>
<a href="sec-si.html" id="toc-sec-si"><span class="toc-section-number">7</span> Additional materials and session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="preamble" class="section level1" number="1">
<h1>
<span class="header-section-number">Chapter 1</span> Preamble</h1>
<p>The aim of the <a href="https://www.rformassspectrometry.org/">R for Mass
Spectrometry</a> initiative is to
provide efficient, thoroughly documented, tested and flexible R
software for the analysis and interpretation of high throughput mass
spectrometry assays, including proteomics and metabolomics
experiments. The project formalises the longtime collaborative
development efforts of its core members under the RforMassSpectrometry
organisation to facilitate dissemination and accessibility of their
work.</p>
<p>
<span class="marginnote shownote">
<!--
<div class="figure">--><span style="display:block;" id="fig:sticker"></span>
<img src="https://github.com/rformassspectrometry/stickers/raw/master/sticker/RforMassSpectrometry.png" alt="The *R for Mass Spectrometry* intiative sticker, designed by Johannes Rainer." width="50%"><!--
<p class="caption marginnote">-->Figure 1.1: The <em>R for Mass Spectrometry</em> intiative sticker, designed by Johannes Rainer.<!--</p>-->
<!--</div>--></span>
</p>
<p>This material introduces participants to the analysis and exploration
of mass spectrometry (MS) based proteomics data using R and
Bioconductor. The course will cover all levels of MS data, from raw
data to identification and quantitation data, up to the statistical
interpretation of a typical shotgun MS experiment and will focus on
hands-on tutorials. At the end of this course, the participants will
be able to manipulate MS data in R and use existing packages for their
exploratory and statistical proteomics data analysis.</p>
<div id="targeted-audience-and-assumed-background" class="section level2" number="1.1">
<h2>
<span class="header-section-number">1.1</span> Targeted audience and assumed background<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('targeted-audience-and-assumed-background')" onmouseout="reset_tooltip('targeted-audience-and-assumed-background-tooltip')"><span class="tooltiptext" id="targeted-audience-and-assumed-background-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>The course material is targeted to either proteomics practitioners or
data analysts/bioinformaticians that would like to learn how to use R
and Bioconductor to analyse proteomics data. Familiarity with MS or
proteomics in general is desirable, but not essential as we will walk
through and describe a typical MS data as part of learning about the
tools. For approachable introductions to sample preparation, mass
spectrometry, data interpretation and analysis, readers are redirected
to:</p>
<ul>
<li>
<em>A beginner’s guide to mass spectrometry–based proteomics</em> <span class="citation">(<label for="tufte-mn-1" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-1" class="margin-toggle">Sinha and Mann 2020<span class="marginnote">Sinha, Ankit, and Matthias Mann. 2020. <span>“<span class="nocase">A beginner’s guide to mass spectrometry–based proteomics</span>.”</span> <em>The Biochemist</em>, September. <a href="https://doi.org/10.1042/BIO20200057">https://doi.org/10.1042/BIO20200057</a>.</span>)</span>
</li>
<li>
<em>The ABC’s (and XYZ’s) of peptide sequencing</em> <span class="citation">(<label for="tufte-mn-2" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-2" class="margin-toggle">Steen and Mann 2004<span class="marginnote">Steen, Hanno, and Matthias Mann. 2004. <span>“The <span class="nocase">ABC’s</span> (and <span class="nocase">XYZ’s</span>) of Peptide Sequencing.”</span> <em>Nat. Rev. Mol. Cell Biol.</em> 5 (9): 699–711.</span>)</span>
</li>
<li>
<em>How do shotgun proteomics algorithms identify proteins?</em> <span class="citation">(<label for="tufte-mn-3" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-3" class="margin-toggle">Marcotte 2007<span class="marginnote">Marcotte, Edward M. 2007. <span>“How Do Shotgun Proteomics Algorithms Identify Proteins?”</span> <em>Nat. Biotechnol.</em> 25 (7): 755–57.</span>)</span>
</li>
<li>
<em>An Introduction to Mass Spectrometry-Based Proteomics</em> <span class="citation">(<label for="tufte-mn-4" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-4" class="margin-toggle">Shuken 2023<span class="marginnote">Shuken, Steven R. 2023. <span>“An Introduction to Mass <span>Spectrometry-Based</span> Proteomics.”</span> <em>J. Proteome Res.</em>, June.</span>)</span>
</li>
</ul>
<p>A working knowledge of R (R syntax, commonly used functions, basic
data structures such as data frames, vectors, matrices, … and their
manipulation) is required. Familiarity with other Bioconductor omics
data classes and the tidyverse syntax is useful, but not necessary.</p>
</div>
<div id="setup" class="section level2" number="1.2">
<h2>
<span class="header-section-number">1.2</span> Setup<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('setup')" onmouseout="reset_tooltip('setup-tooltip')"><span class="tooltiptext" id="setup-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>This material uses the latest version of the R for Mass Spectrometry
package and their dependencies. It might thus be possible that even
the latest Bioconductor stable version isn’t recent enough.</p>
<p>To install all the necessary package, please use the latest release of
R and execute:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="index.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-2"><a href="index.html#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb1-3"><a href="index.html#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="index.html#cb1-4" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"remotes"</span>)</span>
<span id="cb1-5"><a href="index.html#cb1-5" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"tidyverse"</span>)</span>
<span id="cb1-6"><a href="index.html#cb1-6" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"factoextra"</span>)</span>
<span id="cb1-7"><a href="index.html#cb1-7" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"MsDataHub"</span>)</span>
<span id="cb1-8"><a href="index.html#cb1-8" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"mzR"</span>)</span>
<span id="cb1-9"><a href="index.html#cb1-9" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"rhdf5"</span>)</span>
<span id="cb1-10"><a href="index.html#cb1-10" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"rpx"</span>)</span>
<span id="cb1-11"><a href="index.html#cb1-11" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"MsCoreUtils"</span>)</span>
<span id="cb1-12"><a href="index.html#cb1-12" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"QFeatures"</span>)</span>
<span id="cb1-13"><a href="index.html#cb1-13" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"Spectra"</span>)</span>
<span id="cb1-14"><a href="index.html#cb1-14" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"ProtGenerics"</span>)</span>
<span id="cb1-15"><a href="index.html#cb1-15" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"PSMatch"</span>)</span>
<span id="cb1-16"><a href="index.html#cb1-16" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"pheatmap"</span>)</span>
<span id="cb1-17"><a href="index.html#cb1-17" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"limma"</span>)</span>
<span id="cb1-18"><a href="index.html#cb1-18" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"MSnID"</span>)</span>
<span id="cb1-19"><a href="index.html#cb1-19" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"RforMassSpectrometry/SpectraVis"</span>)</span></code></pre></div>
<p>After installation, you can download some data that will be used in
the latter chapter running the following:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="index.html#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rpx)</span>
<span id="cb2-2"><a href="index.html#cb2-2" aria-hidden="true" tabindex="-1"></a>px <span class="ot">&lt;-</span> <span class="fu">PXDataset</span>(<span class="st">"PXD000001"</span>) <span class="do">## answer yes if asked to create a cache directory</span></span>
<span id="cb2-3"><a href="index.html#cb2-3" aria-hidden="true" tabindex="-1"></a>fn <span class="ot">&lt;-</span> <span class="st">"TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML"</span></span>
<span id="cb2-4"><a href="index.html#cb2-4" aria-hidden="true" tabindex="-1"></a>mzf <span class="ot">&lt;-</span> <span class="fu">pxget</span>(px, fn)</span>
<span id="cb2-5"><a href="index.html#cb2-5" aria-hidden="true" tabindex="-1"></a>px <span class="ot">&lt;-</span> <span class="fu">PXDataset</span>(<span class="st">"PXD022816"</span>)</span>
<span id="cb2-6"><a href="index.html#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">pxget</span>(px, <span class="fu">grep</span>(<span class="st">"mzID"</span>, <span class="fu">pxfiles</span>(px))[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span>
<span id="cb2-7"><a href="index.html#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">pxget</span>(px, <span class="fu">grep</span>(<span class="st">"mzML"</span>, <span class="fu">pxfiles</span>(px))[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span></code></pre></div>
<p>All software versions used to generate this document are recoded at
the end of the book in chapter <a href="sec-si.html#sec-si">7</a>.</p>
</div>
<div id="questions-and-help" class="section level2" number="1.3">
<h2>
<span class="header-section-number">1.3</span> Questions and help<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('questions-and-help')" onmouseout="reset_tooltip('questions-and-help-tooltip')"><span class="tooltiptext" id="questions-and-help-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>For questions about specific software or their usage, please refer to
the software’s github issue page, or use the <a href="http://support.bioconductor.org/">Bioconductor support
site</a>.</p>
</div>
<div id="citation" class="section level2" number="1.4">
<h2>
<span class="header-section-number">1.4</span> Citation<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('citation')" onmouseout="reset_tooltip('citation-tooltip')"><span class="tooltiptext" id="citation-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>If you need to cite this book, please use the following reference:</p>
<p><a href="https://doi.org/10.5281/zenodo.15180829"><img src="https://zenodo.org/badge/349528091.svg" alt="DOI"></a></p>
<p>Laurent Gatto, Sebastian Gibb and Johannes Rainer, <em>R for Mass
Spectrometry</em> (2025)
<a href="https://doi.org/10.5281/zenodo.15180829">DOI:10.5281/zenodo.15180830</a>.</p>
</div>
<div id="acknowledgments" class="section level2" number="1.5">
<h2>
<span class="header-section-number">1.5</span> Acknowledgments<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('acknowledgments')" onmouseout="reset_tooltip('acknowledgments-tooltip')"><span class="tooltiptext" id="acknowledgments-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Thank you to <a href="https://github.com/csoneson">Charlotte Soneson</a> for
fixing many typos in a previous version of this book.</p>
</div>
<div id="license" class="section level2" number="1.6">
<h2>
<span class="header-section-number">1.6</span> License<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('license')" onmouseout="reset_tooltip('license-tooltip')"><span class="tooltiptext" id="license-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p><a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png"></a><br>This material is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons
Attribution-ShareAlike 4.0 International License</a>. You are free to
<strong>share</strong> (copy and redistribute the material in any medium or format)
and <strong>adapt</strong> (remix, transform, and build upon the material) for any
purpose, even commercially, as long as you give appropriate credit and
distribute your contributions under the same license as the original.</p>

</div>
</div>
<p style="text-align: center;">
<a href="sec-msintro.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2025-06-05
 using 
R version 4.5.0 (2025-04-11)
</p>
</div>
</div>



</body>
</html>
