<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 6 Supplementary exercises | R for Mass Spectrometry" />
<meta property="og:type" content="book" />




<meta name="author" content="Laurent Gatto, Sebastian Gibb, Johannes Rainer" />

<meta name="date" content="2025-06-20" />


<meta name="description" content="Chapter 6 Supplementary exercises | R for Mass Spectrometry">

<title>Chapter 6 Supplementary exercises | R for Mass Spectrometry</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>
<script>
function copy_link(id) {
  var dummy = document.createElement('input'),
  text = window.location.href.split(/[?#]/)[0] + '#' + id;
  document.body.appendChild(dummy);
  dummy.value = text;
  dummy.select();
  document.execCommand('copy');
  document.body.removeChild(dummy);
  
  var tooltip = document.getElementById(id + '-tooltip');
  tooltip.innerHTML = 'Copied!';
}

function reset_tooltip(id) {
  var tooltip = document.getElementById(id);
  tooltip.innerHTML = 'Copy link';
}
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }

code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #204a87; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #204a87; font-weight: bold; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #ce5c00; font-weight: bold; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">R for Mass Spectrometry<p><p class="author">Laurent Gatto, Sebastian Gibb, Johannes Rainer</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html" id="toc-preamble"><span class="toc-section-number">1</span> Preamble</a>
<a href="sec-msintro.html" id="toc-sec-msintro"><span class="toc-section-number">2</span> Introduction</a>
<a href="sec-raw.html" id="toc-sec-raw"><span class="toc-section-number">3</span> Raw MS data</a>
<a href="sec-id.html" id="toc-sec-id"><span class="toc-section-number">4</span> Identification data</a>
<a href="sec-quant.html" id="toc-sec-quant"><span class="toc-section-number">5</span> Quantitative data</a>
<a id="active-page" href="sec-ex.html" id="toc-sec-ex"><span class="toc-section-number">6</span> Supplementary exercises</a><ul class="toc-sections">
<li class="toc"><a href="#raw-data-and-identification-results"> Raw data and identification results</a></li>
<li class="toc"><a href="#search-engine"> Search engine</a></li>
<li class="toc"><a href="#quantitative-data-processing"> Quantitative data processing</a></li>
</ul>
<a href="sec-anx.html" id="toc-sec-anx"><span class="toc-section-number">7</span> Annex</a>
<a href="sec-si.html" id="toc-sec-si"><span class="toc-section-number">8</span> Additional materials and session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="sec-ex" class="section level1" number="6">
<h1>
<span class="header-section-number">Chapter 6</span> Supplementary exercises</h1>
<div id="raw-data-and-identification-results" class="section level2" number="6.1">
<h2>
<span class="header-section-number">6.1</span> Raw data and identification results<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('raw-data-and-identification-results')" onmouseout="reset_tooltip('raw-data-and-identification-results-tooltip')"><span class="tooltiptext" id="raw-data-and-identification-results-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<ul>
<li><p>Download the 3 first mzML and mzID files from the
<a href="https://www.ebi.ac.uk/pride/archive/projects/PXD022816">PXD022816</a>
project <span class="citation">(<label for="tufte-mn-16" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-16" class="margin-toggle">Morgenstern, Barzilay, and Levin 2021<span class="marginnote">Morgenstern, David, Rotem Barzilay, and Yishai Levin. 2021. <span>“<span>RawBeans</span>: A Simple, Vendor-Independent, Raw-Data Quality-Control Tool.”</span> <em>Journal of Proteome Research</em>. <a href="https://doi.org/10.1021/acs.jproteome.0c00956">https://doi.org/10.1021/acs.jproteome.0c00956</a>.</span>)</span>.</p></li>
<li><p>Generate a <code>Spectra</code> object and a table of filtered PSMs. Visualise
the total ion chromatograms and check the quality of the
identification data by comparing the density of the decoy and target
PSMs id scores for each file.</p></li>
<li><p>Join the raw and identification data. Beware though that the joining
must now be performed by spectrum ids and by files.</p></li>
<li><p>Extract the PSMs that have been matched to peptides from protein
<code>O43175</code> and compare and cluster the scans. Hint: once you have
created the smaller <code>Spectra</code> object with the scans of interest,
switch to an in-memory backend to seed up the calculations.</p></li>
<li><p>Generate total ion chromatograms for each acquisition and annotate
the MS1 scans with the number of PSMs using the
<code>countIdentifications()</code> function, as shown above. The function will
automatically perform the counts in parallel for each acquisition.</p></li>
</ul>
</div>
<div id="search-engine" class="section level2" number="6.2">
<h2>
<span class="header-section-number">6.2</span> Search engine<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('search-engine')" onmouseout="reset_tooltip('search-engine-tooltip')"><span class="tooltiptext" id="search-engine-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Download the <a href="https://dataverse.uclouvain.be/dataset.xhtml?persistentId=doi:10.14428/DVN/QR0OYG&amp;faces-redirect=true#">spectra and protein
database</a>
needed for the exercise (here is a <a href="https://dataverse.uclouvain.be/api/access/dataset/:persistentId/?persistentId=doi:10.14428/DVN/QR0OYG%5D">direct
link</a>. The
protein database is in fasta format and can be processed as described
in the section <a href="sec-id.html#sec-id-seq">4.8</a> <em>Reading and processing protein
sequences</em>. The MS2 spectra are provided in the Mascot Generic Format
(MGF) format, that can be loaded using the dedicated
<a href="https://rformassspectrometry.github.io/MsBackendMgf/">MsBackendMgf</a>
backend as <code>Spectra</code> objects.</p>
<p>You are asked to write code to identify the spectra, following the
principles defined in the <em>Identification data</em> chapter, include ways
to provide confidence in your identification results, beyond a single
identification score.</p>
<p><strong>Hints:</strong></p>
<ul>
<li>Focus on expected peptides sequences that longer than 6 and shorter
than 28 amino acids to reduce the search space.</li>
<li>Do not search each MS2 scan against the whole database, but focus on
peptides that have a mass that is close to the scan’s precuror mass.</li>
<li>To calculate the mass of a peptides, use <code>m/z * c - proton_mass *   c</code>, where <code>m/z</code> and <code>c</code> is the mass-over-charge and the charge of
the precursor and <code>proton_mass</code> is the mass of a proton (available
with <code>PSMatch::getAtomicMass()[["p"]]</code>).</li>
<li>The <code>PSMatch::getAminoAcids()</code> function returns a <code>data.frame</code> of
amino acid properties.</li>
<li>Consider using <code>spectrapply</code> to iterate of the individual scans of a
<code>Spectra</code> object.</li>
</ul>
</div>
<div id="quantitative-data-processing" class="section level2" number="6.3">
<h2>
<span class="header-section-number">6.3</span> Quantitative data processing<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('quantitative-data-processing')" onmouseout="reset_tooltip('quantitative-data-processing-tooltip')"><span class="tooltiptext" id="quantitative-data-processing-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Following up from the quantitative data analysis seen on chapter
<a href="sec-quant.html#sec-quant">5</a>, the following file includes a third condition C and
a two additional lab, tallying now 27 samples.</p>
<div class="sourceCode" id="cb416"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb416-1"><a href="sec-ex.html#cb416-1" aria-hidden="true" tabindex="-1"></a>f <span class="ot">&lt;-</span> MsDataHub<span class="sc">::</span><span class="fu">cptac_a_b_c_peptides.txt</span>()</span></code></pre></div>
<table>
<thead><tr class="header">
<th align="left"></th>
<th align="right">LTQ-Orbitrap_86</th>
<th align="right">LTQ-OrbitrapO_65</th>
<th align="right">LTQ-OrbitrapW_56</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">6A</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">6B</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">3</td>
</tr>
<tr class="odd">
<td align="left">6C</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">3</td>
</tr>
</tbody>
</table>
<p>The full design is shown below.</p>
<table>
<thead><tr class="header">
<th align="left">TRUE</th>
<th align="left">id</th>
<th align="left">condition</th>
<th align="left">lab</th>
<th align="left">previous</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">6A_1</td>
<td align="left">1</td>
<td align="left">6A</td>
<td align="left">LTQ-Orbitrap_86</td>
<td align="left">new</td>
</tr>
<tr class="even">
<td align="left">6A_2</td>
<td align="left">2</td>
<td align="left">6A</td>
<td align="left">LTQ-Orbitrap_86</td>
<td align="left">new</td>
</tr>
<tr class="odd">
<td align="left">6A_3</td>
<td align="left">3</td>
<td align="left">6A</td>
<td align="left">LTQ-Orbitrap_86</td>
<td align="left">new</td>
</tr>
<tr class="even">
<td align="left">6A_4</td>
<td align="left">4</td>
<td align="left">6A</td>
<td align="left">LTQ-OrbitrapO_65</td>
<td align="left">new</td>
</tr>
<tr class="odd">
<td align="left">6A_5</td>
<td align="left">5</td>
<td align="left">6A</td>
<td align="left">LTQ-OrbitrapO_65</td>
<td align="left">new</td>
</tr>
<tr class="even">
<td align="left">6A_6</td>
<td align="left">6</td>
<td align="left">6A</td>
<td align="left">LTQ-OrbitrapO_65</td>
<td align="left">new</td>
</tr>
<tr class="odd">
<td align="left">6A_7</td>
<td align="left">7</td>
<td align="left">6A</td>
<td align="left">LTQ-OrbitrapW_56</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">6A_8</td>
<td align="left">8</td>
<td align="left">6A</td>
<td align="left">LTQ-OrbitrapW_56</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">6A_9</td>
<td align="left">9</td>
<td align="left">6A</td>
<td align="left">LTQ-OrbitrapW_56</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">6B_1</td>
<td align="left">1</td>
<td align="left">6B</td>
<td align="left">LTQ-Orbitrap_86</td>
<td align="left">new</td>
</tr>
<tr class="odd">
<td align="left">6B_2</td>
<td align="left">2</td>
<td align="left">6B</td>
<td align="left">LTQ-Orbitrap_86</td>
<td align="left">new</td>
</tr>
<tr class="even">
<td align="left">6B_3</td>
<td align="left">3</td>
<td align="left">6B</td>
<td align="left">LTQ-Orbitrap_86</td>
<td align="left">new</td>
</tr>
<tr class="odd">
<td align="left">6B_4</td>
<td align="left">4</td>
<td align="left">6B</td>
<td align="left">LTQ-OrbitrapO_65</td>
<td align="left">new</td>
</tr>
<tr class="even">
<td align="left">6B_5</td>
<td align="left">5</td>
<td align="left">6B</td>
<td align="left">LTQ-OrbitrapO_65</td>
<td align="left">new</td>
</tr>
<tr class="odd">
<td align="left">6B_6</td>
<td align="left">6</td>
<td align="left">6B</td>
<td align="left">LTQ-OrbitrapO_65</td>
<td align="left">new</td>
</tr>
<tr class="even">
<td align="left">6B_7</td>
<td align="left">7</td>
<td align="left">6B</td>
<td align="left">LTQ-OrbitrapW_56</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">6B_8</td>
<td align="left">8</td>
<td align="left">6B</td>
<td align="left">LTQ-OrbitrapW_56</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">6B_9</td>
<td align="left">9</td>
<td align="left">6B</td>
<td align="left">LTQ-OrbitrapW_56</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">6C_1</td>
<td align="left">1</td>
<td align="left">6C</td>
<td align="left">LTQ-Orbitrap_86</td>
<td align="left">new</td>
</tr>
<tr class="even">
<td align="left">6C_2</td>
<td align="left">2</td>
<td align="left">6C</td>
<td align="left">LTQ-Orbitrap_86</td>
<td align="left">new</td>
</tr>
<tr class="odd">
<td align="left">6C_3</td>
<td align="left">3</td>
<td align="left">6C</td>
<td align="left">LTQ-Orbitrap_86</td>
<td align="left">new</td>
</tr>
<tr class="even">
<td align="left">6C_4</td>
<td align="left">4</td>
<td align="left">6C</td>
<td align="left">LTQ-OrbitrapO_65</td>
<td align="left">new</td>
</tr>
<tr class="odd">
<td align="left">6C_5</td>
<td align="left">5</td>
<td align="left">6C</td>
<td align="left">LTQ-OrbitrapO_65</td>
<td align="left">new</td>
</tr>
<tr class="even">
<td align="left">6C_6</td>
<td align="left">6</td>
<td align="left">6C</td>
<td align="left">LTQ-OrbitrapO_65</td>
<td align="left">new</td>
</tr>
<tr class="odd">
<td align="left">6C_7</td>
<td align="left">7</td>
<td align="left">6C</td>
<td align="left">LTQ-OrbitrapW_56</td>
<td align="left">new</td>
</tr>
<tr class="even">
<td align="left">6C_8</td>
<td align="left">8</td>
<td align="left">6C</td>
<td align="left">LTQ-OrbitrapW_56</td>
<td align="left">new</td>
</tr>
<tr class="odd">
<td align="left">6C_9</td>
<td align="left">9</td>
<td align="left">6C</td>
<td align="left">LTQ-OrbitrapW_56</td>
<td align="left">new</td>
</tr>
</tbody>
</table>
<ul>
<li>Repeat the analysis described in chapter <a href="sec-quant.html#sec-quant">5</a> using the
extended dataset, trying to optimise true positive results and
avoiding false positive. Think about the best experimental design
approach, how to best process the data, visualising important steps
along the way, to conclude with a volcano plot and a table tallying
the number of true/false positive/negative results.</li>
</ul>
</div>
</div>
<p style="text-align: center;">
<a href="sec-quant.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-anx.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2025-06-20
 using 
R version 4.5.0 (2025-04-11)
</p>
</div>
</div>



</body>
</html>
