<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 7 Additional materials and session information | R for Mass Spectrometry" />
<meta property="og:type" content="book" />




<meta name="author" content="Laurent Gatto, Sebastian Gibb, Johannes Rainer" />

<meta name="date" content="2025-06-20" />


<meta name="description" content="Chapter 7 Additional materials and session information | R for Mass Spectrometry">

<title>Chapter 7 Additional materials and session information | R for Mass Spectrometry</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>
<script>
function copy_link(id) {
  var dummy = document.createElement('input'),
  text = window.location.href.split(/[?#]/)[0] + '#' + id;
  document.body.appendChild(dummy);
  dummy.value = text;
  dummy.select();
  document.execCommand('copy');
  document.body.removeChild(dummy);
  
  var tooltip = document.getElementById(id + '-tooltip');
  tooltip.innerHTML = 'Copied!';
}

function reset_tooltip(id) {
  var tooltip = document.getElementById(id);
  tooltip.innerHTML = 'Copy link';
}
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }

code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #204a87; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #204a87; font-weight: bold; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #ce5c00; font-weight: bold; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">R for Mass Spectrometry<p><p class="author">Laurent Gatto, Sebastian Gibb, Johannes Rainer</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html" id="toc-preamble"><span class="toc-section-number">1</span> Preamble</a>
<a href="sec-msintro.html" id="toc-sec-msintro"><span class="toc-section-number">2</span> Introduction</a>
<a href="sec-raw.html" id="toc-sec-raw"><span class="toc-section-number">3</span> Raw MS data</a>
<a href="sec-id.html" id="toc-sec-id"><span class="toc-section-number">4</span> Identification data</a>
<a href="sec-quant.html" id="toc-sec-quant"><span class="toc-section-number">5</span> Quantitative data</a>
<a href="sec-anx.html" id="toc-sec-anx"><span class="toc-section-number">6</span> Annex</a>
<a id="active-page" href="sec-si.html" id="toc-sec-si"><span class="toc-section-number">7</span> Additional materials and session information</a><ul class="toc-sections">
<li class="toc"><a href="#additional-materials"> Additional materials</a></li>
<li class="toc"><a href="#compiling-the-book-locally"> Compiling the book locally</a></li>
<li class="toc"><a href="#session-information"> Session information</a></li>
</ul>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="sec-si" class="section level1" number="7">
<h1>
<span class="header-section-number">Chapter 7</span> Additional materials and session information</h1>
<div id="additional-materials" class="section level2" number="7.1">
<h2>
<span class="header-section-number">7.1</span> Additional materials<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('additional-materials')" onmouseout="reset_tooltip('additional-materials-tooltip')"><span class="tooltiptext" id="additional-materials-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<ul>
<li><p>The <a href="https://lgatto.github.io/QFeaturesScpWorkshop2021/">Single-cell proteomics data analysis using <code>QFeatures</code> and
<code>scp</code></a> workshop
is provided as two vignettes. The first one provides a general
introduction to the <code>QFeatures</code> class in the general context of mass
spectrometry-based proteomics data manipulation. The second vignette
focuses on single-cell application and introduces the <code>scp</code> package
<span class="citation">(<label for="tufte-mn-16" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-16" class="margin-toggle">Vanderaa and Gatto 2021<span class="marginnote">Vanderaa, Christophe, and Laurent Gatto. 2021. <span>“Replication of Single-Cell Proteomics Data Reveals Important Computational Challenges.”</span> <em>Expert Rev. Proteomics</em>, October.</span>)</span> as an extension of <code>QFeatures</code>. This second
vignette also provides exercises that give the attendee the
opportunity to apply the learned concepts to reproduce a published
analysis on a subset of a real data set. A <a href="https://github.com/lgatto/2024_scpworkshop_EUBIC">recent
workshop</a>, offered
at the 2024 EuBIC winter school, provides teaching material for the
new <em>scplainer</em> analysis workflow.</p></li>
<li><p>A tutorial presenting <a href="https://jorainer.github.io/MetaboAnnotationTutorials/">Use Cases and Examples for Annotation of
Untargeted Metabolomics
Data</a> using
the <code>MetaboAnnotation</code> and <code>MetaboCoreUtils</code> packages
<span class="citation">(<label for="tufte-mn-17" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-17" class="margin-toggle">Rainer et al. 2022<span class="marginnote">Rainer, Johannes, Andrea Vicini, Liesa Salzer, Jan Stanstrup, Josep M Badia, Steffen Neumann, Michael A Stravs, et al. 2022. <span>“A Modular and Expandable Ecosystem for Metabolomics Data Annotation in <span>R</span>.”</span> <em>Metabolites</em> 12 (2): 173.</span>)</span>.</p></li>
<li><p><a href="https://jorainer.github.io/xcmsTutorials/">Exploring and analyzing LC-MS data with Spectra and
xcms</a> provides an
overview of recent developments in Bioconductor to work with mass
spectrometry
(<a href="https://github.com/RforMassSpectrometry/MsExperiment">MsExperiment</a>,
<a href="https://github.com/RforMassSpectrometry/Spectra">Spectra</a>) and
specifically LC-MS data (<a href="https://github.com/sneumann/xcms">xcms</a>)
and walks through the preprocessing of a small data set emphasizing
on selection of data-dependent settings for the individual
pre-processing steps.</p></li>
</ul>
<p>The <a href="https://jorainer.github.io/SpectraTutorials">SpectraTutorials</a> package provides three different vignettes:</p>
<ul>
<li>
<a href="https://jorainer.github.io/SpectraTutorials/articles/analyzing-MS-data-from-different-sources-with-Spectra.html">Seamless Integration of Mass Spectrometry Data from Different
Sources</a>:
describes import/export of MS data from/to files in different format as well
as processing and handling of MS data with the <em>Spectra</em> package.</li>
<li>
<a href="https://jorainer.github.io/SpectraTutorials/articles/Spectra-backends.html">Spectra: an Expandable Infrastructure to Handle Mass Spectrometry
Data</a>:
explains the concept of backends in <em>Spectra</em>, their properties, use cases
along with performance considerations.</li>
<li>
<a href="https://jorainer.github.io/SpectraTutorials/articles/Spectra-matching-with-MetaboAnnotation.html">MS/MS Spectra Matching with the MetaboAnnotation
Package</a>:
explains how the <em>Spectra</em> package can be used together with the
<em><a href="https://bioconductor.org/packages/3.21/MetaboAnnotation">MetaboAnnotation</a></em> package in LC-MS/MS annotation
workflows for untargeted metabolomics data.</li>
</ul>
</div>
<div id="compiling-the-book-locally" class="section level2" number="7.2">
<h2>
<span class="header-section-number">7.2</span> Compiling the book locally<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('compiling-the-book-locally')" onmouseout="reset_tooltip('compiling-the-book-locally-tooltip')"><span class="tooltiptext" id="compiling-the-book-locally-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>To compile and render the teaching material, you will also need
the <em><a href="https://bioconductor.org/packages/3.21/BiocStyle">BiocStyle</a></em> package and the (slighly
modified) <a href="https://www-huber.embl.de/users/msmith/msmbstyle/">Modern Statistics for Model Biology (msmb) HTML Book
Style</a> by Mike
Smith:</p>
<div class="sourceCode" id="cb458"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb458-1"><a href="sec-si.html#cb458-1" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="fu">c</span>(<span class="st">"bookdown"</span>, <span class="st">"BiocStyle"</span>, <span class="st">"lgatto/msmbstyle"</span>))</span></code></pre></div>
<p>Clone the <a href="https://github.com/Rformassspectrometry/book">book
repository</a> and render
the book with</p>
<div class="sourceCode" id="cb459"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb459-1"><a href="sec-si.html#cb459-1" tabindex="-1"></a>bookdown<span class="sc">::</span><span class="fu">render_book</span>(<span class="st">"."</span>)</span></code></pre></div>
</div>
<div id="session-information" class="section level2" number="7.3">
<h2>
<span class="header-section-number">7.3</span> Session information<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('session-information')" onmouseout="reset_tooltip('session-information-tooltip')"><span class="tooltiptext" id="session-information-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>The following packages have been used to generate this document.</p>
<div class="sourceCode" id="cb460"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb460-1"><a href="sec-si.html#cb460-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.5.0 (2025-04-11)
## Platform: x86_64-pc-linux-gnu
## Running under: Ubuntu 24.04.2 LTS
## 
## Matrix products: default
## BLAS:   /opt/R-4.5/lib/R/lib/libRblas.so 
## LAPACK: /opt/R-4.5/lib/R/lib/libRlapack.so;  LAPACK version 3.12.1
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## time zone: Europe/Brussels
## tzcode source: system (glibc)
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] mzID_1.46.0                 patchwork_1.3.0            
##  [3] factoextra_1.0.7            gplots_3.2.0               
##  [5] limma_3.64.1                lubridate_1.9.4            
##  [7] forcats_1.0.0               stringr_1.5.1              
##  [9] purrr_1.0.4                 readr_2.1.5                
## [11] tidyr_1.3.1                 tibble_3.2.1               
## [13] tidyverse_2.0.0             MSnID_1.42.0               
## [15] cleaver_1.46.0              Biostrings_2.76.0          
## [17] XVector_0.48.0              ggplot2_3.5.2              
## [19] dplyr_1.1.4                 msdata_0.48.0              
## [21] MsDataHub_1.8.0             rpx_2.16.0                 
## [23] MsCoreUtils_1.20.0          QFeatures_1.19.1           
## [25] MultiAssayExperiment_1.34.0 SummarizedExperiment_1.38.1
## [27] Biobase_2.68.0              GenomicRanges_1.60.0       
## [29] GenomeInfoDb_1.44.0         IRanges_2.42.0             
## [31] MatrixGenerics_1.20.0       matrixStats_1.5.0          
## [33] Spectra_1.18.2              BiocParallel_1.42.0        
## [35] S4Vectors_0.46.0            BiocGenerics_0.54.0        
## [37] generics_0.1.4              mzR_2.42.0                 
## [39] Rcpp_1.0.14                 BiocStyle_2.36.0           
## 
## loaded via a namespace (and not attached):
##   [1] RColorBrewer_1.1-3      rstudioapi_0.17.1       jsonlite_2.0.0         
##   [4] magrittr_2.0.3          MALDIquant_1.22.3       farver_2.1.2           
##   [7] rmarkdown_2.29          fs_1.6.6                vctrs_0.6.5            
##  [10] memoise_2.0.1           RCurl_1.98-1.17         rstatix_0.7.2          
##  [13] htmltools_0.5.8.1       S4Arrays_1.8.0          BiocBaseUtils_1.10.0   
##  [16] AnnotationHub_3.16.0    curl_6.2.3              broom_1.0.8            
##  [19] Rhdf5lib_1.30.0         Formula_1.2-5           SparseArray_1.8.0      
##  [22] rhdf5_2.52.0            sass_0.4.10             KernSmooth_2.23-26     
##  [25] bslib_0.9.0             plyr_1.8.9              impute_1.82.0          
##  [28] cachem_1.1.0            igraph_2.1.4            mime_0.13              
##  [31] lifecycle_1.0.4         iterators_1.0.14        pkgconfig_2.0.3        
##  [34] Matrix_1.7-3            R6_2.6.1                fastmap_1.2.0          
##  [37] GenomeInfoDbData_1.2.14 clue_0.3-66             digest_0.6.37          
##  [40] colorspace_2.1-1        pcaMethods_2.0.0        AnnotationDbi_1.70.0   
##  [43] ExperimentHub_2.16.0    RSQLite_2.3.11          ggpubr_0.6.0           
##  [46] filelock_1.0.3          labeling_0.4.3          timechange_0.3.0       
##  [49] httr_1.4.7              abind_1.4-8             compiler_4.5.0         
##  [52] bit64_4.6.0-1           withr_3.0.2             doParallel_1.0.17      
##  [55] backports_1.5.0         carData_3.0-5           DBI_1.2.3              
##  [58] R.utils_2.13.0          ggsignif_0.6.4          MASS_7.3-65            
##  [61] rappdirs_0.3.3          DelayedArray_0.34.1     caTools_1.18.3         
##  [64] gtools_3.9.5            tools_4.5.0             PSMatch_1.13.2         
##  [67] R.oo_1.27.1             glue_1.8.0              R.cache_0.17.0         
##  [70] rhdf5filters_1.20.0     grid_4.5.0              cluster_2.1.8.1        
##  [73] reshape2_1.4.4          gtable_0.3.6            msmbstyle_0.0.22       
##  [76] tzdb_0.5.0              preprocessCore_1.70.0   R.methodsS3_1.8.2      
##  [79] hms_1.1.3               data.table_1.17.4       MetaboCoreUtils_1.16.1 
##  [82] car_3.1-3               xml2_1.3.8              utf8_1.2.5             
##  [85] ggrepel_0.9.6           BiocVersion_3.21.1      foreach_1.5.2          
##  [88] pillar_1.10.2           BiocFileCache_2.16.0    lattice_0.22-6         
##  [91] bit_4.6.0               tidyselect_1.2.1        knitr_1.50             
##  [94] bookdown_0.34.2         ProtGenerics_1.40.0     xfun_0.52              
##  [97] statmod_1.5.0           MSnbase_2.34.1          pheatmap_1.0.12        
## [100] stringi_1.8.7          
##  [ reached 'max' / getOption("max.print") -- omitted 23 entries ]</code></pre>

</div>
</div>
<p style="text-align: center;">
<a href="sec-anx.html"><button class="btn btn-default">Previous</button></a>
</p>
<p class="build-date">Page built: 
2025-06-20
 using 
R version 4.5.0 (2025-04-11)
</p>
</div>
</div>



</body>
</html>
