<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 5 Quantitative data | R for Mass Spectrometry" />
<meta property="og:type" content="book" />





<meta name="author" content="Laurent Gatto, Sebastian Gibb, Johannes Rainer" />

<meta name="date" content="2021-03-23" />


<meta name="description" content="Chapter 5 Quantitative data | R for Mass Spectrometry">

<title>Chapter 5 Quantitative data | R for Mass Spectrometry</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }

code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { background-color: #f8f8f8; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">R for Mass Spectrometry<p><p class="author">Laurent Gatto, Sebastian Gibb, Johannes Rainer</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html"><span class="toc-section-number">1</span> Preamble</a>
<a href="sec-msintro.html"><span class="toc-section-number">2</span> Introduction</a>
<a href="raw-ms-data.html"><span class="toc-section-number">3</span> Raw MS data</a>
<a href="identification-data.html"><span class="toc-section-number">4</span> Identification data</a>
<a id="active-page" href="sec-quant.html"><span class="toc-section-number">5</span> Quantitative data</a><ul class="toc-sections">
<li class="toc"><a href="#quantitation-methodologies"> Quantitation methodologies</a></li>
<li class="toc"><a href="#sec:qf"> QFeatures</a></li>
<li class="toc"><a href="#creating-qfeatures-object"> Creating <code>QFeatures</code> object</a></li>
<li class="toc"><a href="#analysis-pipeline"> Analysis pipeline</a></li>
<li class="toc"><a href="#summary-exercice-1"> Summary exercice</a></li>
</ul>
<a href="sec-si.html"><span class="toc-section-number">6</span> References and session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body>
<div id="sec:quant" class="section level1">
<h1>
<span class="header-section-number">Chapter 5</span> Quantitative data</h1>
<div id="quantitation-methodologies" class="section level2">
<h2>
<span class="header-section-number">5.1</span> Quantitation methodologies</h2>
<p>There are a wide range of proteomics quantitation techniques that can
broadly be classified as labelled vs. label-free, depending whether
the features are labelled prior the MS acquisition and the MS level at
which quantitation is inferred, namely MS1 or MS2.</p>
<table>
<thead><tr class="header">
<th align="left"></th>
<th align="left">Label-free</th>
<th align="left">Labelled</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">MS1</td>
<td align="left">XIC</td>
<td align="left">SILAC, 15N</td>
</tr>
<tr class="even">
<td align="left">MS2</td>
<td align="left">Counting</td>
<td align="left">iTRAQ, TMT</td>
</tr>
</tbody>
</table>
<div id="label-free-ms2-spectral-counting" class="section level3">
<h3>
<span class="header-section-number">5.1.1</span> Label-free MS2: Spectral counting</h3>
<p>In spectral counting, on simply counts the number of quantified
peptides that are assigned to a protein.</p>
<div class="figure">
<span id="fig:sc"></span>
<p class="caption marginnote shownote">
Figure 5.1: Spectral counting. Figure from the <code>Pbase</code> package.
</p>
<img src="img/pbase.png" alt="Spectral counting. Figure from the `Pbase` package." width="75%">
</div>
</div>
<div id="labelled-ms2-isobaric-tagging" class="section level3">
<h3>
<span class="header-section-number">5.1.2</span> Labelled MS2: Isobaric tagging</h3>
<p>Isobaric tagging refers to the labelling using isobaric tags,
i.e. chemical tags that have the same mass and hence can’t be
distinguish by the spectrometer. The peptides of different samples (4,
6, 10, 11 or 16) are labelled with different tags and combined prior
to mass spectrometry acquisition. Given that they are isobaric, all
identical peptides, irrespective of the tag and this the sample of
origin, are co-analysed, up to fragmentation prior to MS2
analysis. During fragmentation, the isobaric tags fall of, fragment
themselves, and result in a set of sample specific peaks. These
specific peaks can be used to infer sample-specific quantitation,
while the rest of the MS2 spectrum is used for identification.</p>
<div class="figure">
<span id="fig:itraq"></span>
<p class="caption marginnote shownote">
Figure 5.2: iTRAQ 4-plex isobaric tagging. Tandem Mass Tags (TMT) offer up to 16 tags.
</p>
<img src="img/itraq.png" alt="iTRAQ 4-plex isobaric tagging. Tandem Mass Tags (TMT) offer up to 16 tags." width="75%">
</div>
</div>
<div id="label-free-ms1-extracted-ion-chromatograms" class="section level3">
<h3>
<span class="header-section-number">5.1.3</span> Label-free MS1: extracted ion chromatograms</h3>
<p>In label-free quantitation, the precursor peaks that match an
identified peptide are integrated of retention time and the area under
that <em>extracted ion chromatogram</em> is used to quantify that peptide in
that sample.</p>
<div class="figure">
<span id="fig:lf"></span>
<p class="caption marginnote shownote">
Figure 5.3: Label-free quantitation. Figure credit <a href="https://github.com/jorainer/">Johannes Rainer</a>.
</p>
<img src="img/chrompeaks.png" alt="Label-free quantitation. Figure credit [Johannes Rainer](https://github.com/jorainer/)." width="75%">
</div>
</div>
<div id="labelled-ms1-silac" class="section level3">
<h3>
<span class="header-section-number">5.1.4</span> Labelled MS1: SILAC</h3>
<p>In SILAc quantitation, sample are grown in a medium that contains
heavy amino acids (typically arginine and lysine). All proteins gown
in this <em>heavy</em> growth medium contain the heavy form of these amino
acids. Two samples, one grown in heavy medium, and one grown in normal
(light) medium are then combined and analysed together. The heavy
peptides precursor peaks are systematically shifted compared to the
light ones, and the ratio between the height of a heavy and light
peaks can be used to calculate peptide and protein fold-changes.</p>
<div class="figure">
<span id="fig:silab"></span>
<p class="caption marginnote shownote">
Figure 5.4: Silac quantitation. Figure credit Wikimedia Commons.
</p>
<img src="img/Silac.png" alt="Silac quantitation. Figure credit Wikimedia Commons." width="75%">
</div>
<p>These different quantitation techniques come with their respective
benefits and distinct challenges, such as large quantities of raw data
processing, data transformation and normalisation, missing values, and
different underlying statistical models for the quantitative data
(count data for spectral counting, continuous data for the others).</p>
<p>In terms of raw data quantitation in R/Bioconductor, most efforts have
been devoted to MS2-level quantitation. Label-free XIC quantitation
has been addressed in the frame of metabolomics data processing by the
<em><a href="https://bioconductor.org/packages/3.12/xcms">xcms</a></em> infrastructure.</p>
<p>Below is a list of suggested packages for some common proteomics
quantitation technologies:</p>
<ul>
<li>Isobaric tagging (iTRAQ and TMT): <em><a href="https://bioconductor.org/packages/3.12/MSnbase">MSnbase</a></em> and <em><a href="https://bioconductor.org/packages/3.12/isobar">isobar</a></em>.</li>
<li>Label-free: <em><a href="https://bioconductor.org/packages/3.12/xcms">xcms</a></em> (metabolomics).</li>
<li>Counting: <em><a href="https://bioconductor.org/packages/3.12/MSnbase">MSnbase</a></em> and <em><a href="https://bioconductor.org/packages/3.12/MSnID">MSnID</a></em> for
peptide-spectrum matching confidence assessment.</li>
<li>
<em><a href="https://github.com/vladpetyuk/N14N15">N14N15</a></em> for heavy Nitrogen-labelled data.</li>
</ul>
</div>
</div>
<div id="sec:qf" class="section level2">
<h2>
<span class="header-section-number">5.2</span> QFeatures</h2>
<p>Mass spectrometry-based quantitative proteomics data can be
representated as a matrix of quantitative values for features (PSMs,
peptides, proteins) arranged along the rows, measured for a set of
samples, arranged along the columns. There is a common representation
for such quantitative data set, namely the <code>SummarizedExperiment</code> (SE)
<span class="citation">(Morgan et al. <a href="#ref-SE">2020</a>)</span> class:</p>
<div class="figure" style="text-align: center">
<span id="fig:sefig"></span>
<p class="caption marginnote shownote">
Figure 5.5: Schematic representation of the anatomy of a <code>SummarizedExperiment</code> object. (Figure taken from the <code>SummarizedExperiment</code> package vignette.)
</p>
<img src="img/SE.png" alt="Schematic representation of the anatomy of a `SummarizedExperiment` object. (Figure taken from the `SummarizedExperiment` package vignette.)" width="100%">
</div>
<ul>
<li>The sample (columns) metadata can be access with the <code>colData()</code>
function.</li>
<li>The features (rows) metadata can be access with the <code>rowData()</code>
column.</li>
<li>If the features represent ranges along genomic coordinates, these
can be accessed with <code>rowRanges()</code>
</li>
<li>Additional metadata describing the overall experiment can be
accessed with <code>metadata()</code>.</li>
<li>The quantiative data can be accessed with <code>assay()</code>.</li>
<li>
<code>assays()</code> returns a list of matrix-like assays.</li>
</ul>
<div id="the-qfeatures-class" class="section level3">
<h3>
<span class="header-section-number">5.2.1</span> The QFeatures class</h3>
<p>While mass spectrometers acquire data for spectra/peptides, the
biological entity of interest are the protein. As part of the data
processing, we are thus required to <strong>aggregate</strong> low-level
quantitative features into higher level data.</p>
<div class="figure">
<span id="fig:featuresplot"></span>
<p class="caption marginnote shownote">
Figure 5.6: Conceptual representation of a <code>QFeatures</code> object and the aggregative relation between different assays.
</p>
<img src="R4MS_files/figure-html/featuresplot-1.png" alt="Conceptual representation of a `QFeatures` object and the aggregative relation between different assays." width="672">
</div>
<p>We are going to start to familiarise ourselves with the <code>QFeatures</code>
class implemented in the
<a href="https://rformassspectrometry.github.io/QFeatures/articles/QFeatures.html"><code>QFeatures</code></a>
package. The class is derived from the Bioconductor
<code>MultiAssayExperiment</code> <span class="citation">(Ramos et al. <a href="#ref-MAE">2017</a>)</span> (MAE) class. Let’s start by loading the
<code>QFeatures</code> package.</p>
<div class="sourceCode" id="cb279"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb279-1" title="1"><span class="kw">library</span>(<span class="st">"QFeatures"</span>)</a></code></pre></div>
<p>Next, we load the <code>feat1</code> test data, which is composed of single
<em>assay</em> of class <code>SummarizedExperiment</code> composed of 10 rows and 2
columns.</p>
<div class="sourceCode" id="cb280"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb280-1" title="1"><span class="kw">data</span>(feat1)</a>
<a class="sourceLine" id="cb280-2" title="2">feat1</a></code></pre></div>
<pre><code>## An instance of class QFeatures containing 1 assays:
##  [1] psms: SummarizedExperiment with 10 rows and 2 columns</code></pre>
<p>Let’s perform some simple operations to familiarise ourselves with the
<code>QFeatures</code> class:</p>
<ul>
<li>Extract the sample metadata using the <code>colData()</code> accessor (like you
have previously done with <code>SummarizedExperiment</code> objects).</li>
</ul>
<div class="sourceCode" id="cb282"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb282-1" title="1"><span class="kw">colData</span>(feat1)</a></code></pre></div>
<pre><code>## DataFrame with 2 rows and 1 column
##        Group
##    &lt;integer&gt;
## S1         1
## S2         2</code></pre>
<p>We can also further annotate the experiment by adding columns to the <code>colData</code> slot:</p>
<div class="sourceCode" id="cb284"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb284-1" title="1"><span class="kw">colData</span>(feat1)<span class="op">$</span>X &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"X1"</span>, <span class="st">"X2"</span>)</a>
<a class="sourceLine" id="cb284-2" title="2">feat1<span class="op">$</span>Y &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"Y1"</span>, <span class="st">"Y2"</span>)</a>
<a class="sourceLine" id="cb284-3" title="3"><span class="kw">colData</span>(feat1)</a></code></pre></div>
<pre><code>## DataFrame with 2 rows and 3 columns
##        Group           X           Y
##    &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## S1         1          X1          Y1
## S2         2          X2          Y2</code></pre>
<ul>
<li>Extract the first (and only) assay composing this <code>QFeaures</code> data
using the <code>[[</code> operator (as you have done to extract elements of a
list) by using the assay’s index or name.</li>
</ul>
<div class="sourceCode" id="cb286"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb286-1" title="1">feat1[[<span class="dv">1</span>]]</a></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 10 2 
## metadata(0):
## assays(1): ''
## rownames(10): PSM1 PSM2 ... PSM9 PSM10
## rowData names(5): Sequence Protein Var location pval
## colnames(2): S1 S2
## colData names(0):</code></pre>
<div class="sourceCode" id="cb288"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb288-1" title="1">feat1[[<span class="st">"psms"</span>]]</a></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 10 2 
## metadata(0):
## assays(1): ''
## rownames(10): PSM1 PSM2 ... PSM9 PSM10
## rowData names(5): Sequence Protein Var location pval
## colnames(2): S1 S2
## colData names(0):</code></pre>
<ul>
<li>Extract the <code>psms</code> assay’s row data and quantitative values.</li>
</ul>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb290-1" title="1"><span class="kw">assay</span>(feat1[[<span class="dv">1</span>]])</a></code></pre></div>
<pre><code>##       S1 S2
## PSM1   1 11
## PSM2   2 12
## PSM3   3 13
## PSM4   4 14
## PSM5   5 15
## PSM6   6 16
## PSM7   7 17
## PSM8   8 18
## PSM9   9 19
## PSM10 10 20</code></pre>
<div class="sourceCode" id="cb292"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb292-1" title="1"><span class="kw">rowData</span>(feat1[[<span class="dv">1</span>]])</a></code></pre></div>
<pre><code>## DataFrame with 10 rows and 5 columns
##            Sequence     Protein       Var      location      pval
##         &lt;character&gt; &lt;character&gt; &lt;integer&gt;   &lt;character&gt; &lt;numeric&gt;
## PSM1       SYGFNAAR       ProtA         1 Mitochondr...     0.084
## PSM2       SYGFNAAR       ProtA         2 Mitochondr...     0.077
## PSM3       SYGFNAAR       ProtA         3 Mitochondr...     0.063
## PSM4       ELGNDAYK       ProtA         4 Mitochondr...     0.073
## PSM5       ELGNDAYK       ProtA         5 Mitochondr...     0.012
## PSM6       ELGNDAYK       ProtA         6 Mitochondr...     0.011
## PSM7  IAEESNFPFI...       ProtB         7       unknown     0.075
## PSM8  IAEESNFPFI...       ProtB         8       unknown     0.038
## PSM9  IAEESNFPFI...       ProtB         9       unknown     0.028
## PSM10 IAEESNFPFI...       ProtB        10       unknown     0.097</code></pre>
</div>
<div id="feature-aggregation" class="section level3">
<h3>
<span class="header-section-number">5.2.2</span> Feature aggregation</h3>
<p>The central functionality of the <code>QFeatures</code> infrastructure is the
aggregation of features into higher-level features while retaining the
link between the different levels. This can be done with the
<a href="https://rformassspectrometry.github.io/QFeatures/reference/QFeatures-aggregate.html"><code>aggregateFeatures()</code> function</a>.</p>
<p>The call below will</p>
<ul>
<li>operate on the <code>psms</code> assay of the <code>feat1</code> objects;</li>
<li>aggregate the rows the assay following the grouping defined in the
<code>peptides</code> row data variables;</li>
<li>perform aggregation using the <code>colMeans()</code> function;</li>
<li>create a new assay named <code>peptides</code> and add it to the <code>feat1</code>
object.</li>
</ul>
<div class="sourceCode" id="cb294"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb294-1" title="1">feat1 &lt;-<span class="st"> </span><span class="kw">aggregateFeatures</span>(feat1, <span class="dt">i =</span> <span class="st">"psms"</span>,</a>
<a class="sourceLine" id="cb294-2" title="2">                           <span class="dt">fcol =</span> <span class="st">"Sequence"</span>,</a>
<a class="sourceLine" id="cb294-3" title="3">                           <span class="dt">name =</span> <span class="st">"peptides"</span>,</a>
<a class="sourceLine" id="cb294-4" title="4">                           <span class="dt">fun =</span> colMeans)</a>
<a class="sourceLine" id="cb294-5" title="5">feat1</a></code></pre></div>
<pre><code>## An instance of class QFeatures containing 2 assays:
##  [1] psms: SummarizedExperiment with 10 rows and 2 columns 
##  [2] peptides: SummarizedExperiment with 3 rows and 2 columns</code></pre>
<ul>
<li>Let’s convince yourself that we understand the effect of feature
aggregation and repeat the calculations manually and check the
content of the new assay’s row data.</li>
</ul>
<div class="sourceCode" id="cb296"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb296-1" title="1"><span class="co">## SYGFNAAR</span></a>
<a class="sourceLine" id="cb296-2" title="2"><span class="kw">colMeans</span>(<span class="kw">assay</span>(feat1[[<span class="dv">1</span>]])[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, ])</a></code></pre></div>
<pre><code>## S1 S2 
##  2 12</code></pre>
<div class="sourceCode" id="cb298"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb298-1" title="1"><span class="kw">assay</span>(feat1[[<span class="dv">2</span>]])[<span class="st">"SYGFNAAR"</span>, ]</a></code></pre></div>
<pre><code>## S1 S2 
##  2 12</code></pre>
<div class="sourceCode" id="cb300"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb300-1" title="1"><span class="co">## ELGNDAYK</span></a>
<a class="sourceLine" id="cb300-2" title="2"><span class="kw">colMeans</span>(<span class="kw">assay</span>(feat1[[<span class="dv">1</span>]])[<span class="dv">4</span><span class="op">:</span><span class="dv">6</span>, ])</a></code></pre></div>
<pre><code>## S1 S2 
##  5 15</code></pre>
<div class="sourceCode" id="cb302"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb302-1" title="1"><span class="kw">assay</span>(feat1[[<span class="dv">2</span>]])[<span class="st">"ELGNDAYK"</span>, ]</a></code></pre></div>
<pre><code>## S1 S2 
##  5 15</code></pre>
<div class="sourceCode" id="cb304"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb304-1" title="1"><span class="co">## IAEESNFPFIK</span></a>
<a class="sourceLine" id="cb304-2" title="2"><span class="kw">colMeans</span>(<span class="kw">assay</span>(feat1[[<span class="dv">1</span>]])[<span class="dv">7</span><span class="op">:</span><span class="dv">10</span>, ])</a></code></pre></div>
<pre><code>##   S1   S2 
##  8.5 18.5</code></pre>
<div class="sourceCode" id="cb306"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb306-1" title="1"><span class="kw">assay</span>(feat1[[<span class="dv">2</span>]])[<span class="st">"IAEESNFPFIK"</span>, ]</a></code></pre></div>
<pre><code>##   S1   S2 
##  8.5 18.5</code></pre>
<div class="sourceCode" id="cb308"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb308-1" title="1"><span class="kw">rowData</span>(feat1[[<span class="dv">2</span>]])</a></code></pre></div>
<pre><code>## DataFrame with 3 rows and 4 columns
##                  Sequence     Protein      location        .n
##               &lt;character&gt; &lt;character&gt;   &lt;character&gt; &lt;integer&gt;
## ELGNDAYK         ELGNDAYK       ProtA Mitochondr...         3
## IAEESNFPFIK IAEESNFPFI...       ProtB       unknown         4
## SYGFNAAR         SYGFNAAR       ProtA Mitochondr...         3</code></pre>
<p>We can now aggregate the peptide-level data into a new protein-level
assay using the <code>colMedians()</code> aggregation function.</p>
<div class="sourceCode" id="cb310"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb310-1" title="1">feat1 &lt;-<span class="st"> </span><span class="kw">aggregateFeatures</span>(feat1, <span class="dt">i =</span> <span class="st">"peptides"</span>,</a>
<a class="sourceLine" id="cb310-2" title="2">                           <span class="dt">fcol =</span> <span class="st">"Protein"</span>,</a>
<a class="sourceLine" id="cb310-3" title="3">                           <span class="dt">name =</span> <span class="st">"proteins"</span>,</a>
<a class="sourceLine" id="cb310-4" title="4">                           <span class="dt">fun =</span> colMedians)</a>
<a class="sourceLine" id="cb310-5" title="5">feat1</a></code></pre></div>
<pre><code>## An instance of class QFeatures containing 3 assays:
##  [1] psms: SummarizedExperiment with 10 rows and 2 columns 
##  [2] peptides: SummarizedExperiment with 3 rows and 2 columns 
##  [3] proteins: SummarizedExperiment with 2 rows and 2 columns</code></pre>
<div class="sourceCode" id="cb312"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb312-1" title="1"><span class="kw">assay</span>(feat1[[<span class="st">"proteins"</span>]])</a></code></pre></div>
<pre><code>##        S1   S2
## ProtA 3.5 13.5
## ProtB 8.5 18.5</code></pre>
</div>
<div id="subsetting-and-filtering" class="section level3">
<h3>
<span class="header-section-number">5.2.3</span> Subsetting and filtering</h3>
<p>The link between the assays becomes apparent when we now subset the
assays for protein A as shown below or using the <code>subsetByFeature()</code>
function. This creates a new instance of class <code>QFeatures</code> containing
assays with the expression data for protein, its peptides and their
PSMs.</p>
<div class="sourceCode" id="cb314"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb314-1" title="1">feat1[<span class="st">"ProtA"</span>, , ]</a></code></pre></div>
<pre><code>## An instance of class QFeatures containing 3 assays:
##  [1] psms: SummarizedExperiment with 6 rows and 2 columns 
##  [2] peptides: SummarizedExperiment with 2 rows and 2 columns 
##  [3] proteins: SummarizedExperiment with 1 rows and 2 columns</code></pre>
<p>The <code>filterFeatures()</code> function can be used to filter rows the assays
composing a <code>QFeatures</code> object using the row data variables. We can
for example retain rows that have a <code>pval</code> &lt; 0.05, which would only
keep rows in the <code>psms</code> assay because the <code>pval</code> is only relevant for
that assay.</p>
<div class="sourceCode" id="cb316"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb316-1" title="1"><span class="kw">filterFeatures</span>(feat1, <span class="op">~</span><span class="st"> </span>pval <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)</a></code></pre></div>
<pre><code>## An instance of class QFeatures containing 3 assays:
##  [1] psms: SummarizedExperiment with 4 rows and 2 columns 
##  [2] peptides: SummarizedExperiment with 0 rows and 2 columns 
##  [3] proteins: SummarizedExperiment with 0 rows and 2 columns</code></pre>
<p>On the other hand, if we filter assay rows for those that localise to
the mitochondrion, we retain the relevant protein, peptides and PSMs.</p>
<div class="sourceCode" id="cb318"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb318-1" title="1"><span class="kw">filterFeatures</span>(feat1, <span class="op">~</span><span class="st"> </span>location <span class="op">==</span><span class="st"> "Mitochondrion"</span>)</a></code></pre></div>
<pre><code>## An instance of class QFeatures containing 3 assays:
##  [1] psms: SummarizedExperiment with 6 rows and 2 columns 
##  [2] peptides: SummarizedExperiment with 2 rows and 2 columns 
##  [3] proteins: SummarizedExperiment with 1 rows and 2 columns</code></pre>
<p>As an exercise, let’s filter rows that do not localise to the
mitochondrion.</p>
<div class="sourceCode" id="cb320"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb320-1" title="1"><span class="kw">filterFeatures</span>(feat1, <span class="op">~</span><span class="st"> </span>location <span class="op">!=</span><span class="st"> "Mitochondrion"</span>)</a></code></pre></div>
<pre><code>## An instance of class QFeatures containing 3 assays:
##  [1] psms: SummarizedExperiment with 4 rows and 2 columns 
##  [2] peptides: SummarizedExperiment with 1 rows and 2 columns 
##  [3] proteins: SummarizedExperiment with 1 rows and 2 columns</code></pre>
<p>You can refer to the <a href="https://rformassspectrometry.github.io/QFeatures/articles/QFeatures.html"><em>Quantitative features for mass spectrometry
data</em></a>
vignette and the <code>QFeature</code> <a href="https://rformassspectrometry.github.io/QFeatures/reference/QFeatures-class.html">manual
page</a>
for more details about the class.</p>
</div>
</div>
<div id="creating-qfeatures-object" class="section level2">
<h2>
<span class="header-section-number">5.3</span> Creating <code>QFeatures</code> object</h2>
<p>While <code>QFeatures</code> objects can be created manually (see <code>?QFeatures</code>
for details), most users will probably possess quantitative data in a
spreadsheet or a dataframe. In such cases, the easiest is to use the
<code>readQFeatures</code> function to extract the quantitative data and metadata
columns. Below, we load the <code>hlpsms</code> dataframe that contains data for
28 PSMs from the TMT-10plex <em>hyper</em>LOPIT spatial
proteomics experiment from <span class="citation">(Christoforou et al. <a href="#ref-Christoforou:2016">2016</a>)</span>. The <code>ecol</code> argument
specifies that columns 1 to 10 contain quantitation data, and that the
assay should be named <code>psms</code> in the returned <code>QFeatures</code> object, to
reflect the nature of the data.</p>
<div class="sourceCode" id="cb322"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb322-1" title="1"><span class="kw">data</span>(hlpsms)</a>
<a class="sourceLine" id="cb322-2" title="2">hl &lt;-<span class="st"> </span><span class="kw">readQFeatures</span>(hlpsms, <span class="dt">ecol =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">name =</span> <span class="st">"psms"</span>)</a>
<a class="sourceLine" id="cb322-3" title="3">hl</a></code></pre></div>
<pre><code>## An instance of class QFeatures containing 1 assays:
##  [1] psms: SummarizedExperiment with 3010 rows and 10 columns</code></pre>
<p>Below, we see that we can extract an assay using its index or its
name. The individual assays are stored as <em>SummerizedExperiment</em>
object and further access its quantitative data and metadata using
the <code>assay</code> and <code>rowData</code> functions</p>
<div class="sourceCode" id="cb324"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb324-1" title="1">hl[[<span class="dv">1</span>]]</a></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 3010 10 
## metadata(0):
## assays(1): ''
## rownames(3010): 1 2 ... 3009 3010
## rowData names(18): Sequence ProteinDescriptions ... RTmin markers
## colnames(10): X126 X127C ... X130N X131
## colData names(0):</code></pre>
<div class="sourceCode" id="cb326"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb326-1" title="1">hl[[<span class="st">"psms"</span>]]</a></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 3010 10 
## metadata(0):
## assays(1): ''
## rownames(3010): 1 2 ... 3009 3010
## rowData names(18): Sequence ProteinDescriptions ... RTmin markers
## colnames(10): X126 X127C ... X130N X131
## colData names(0):</code></pre>
<div class="sourceCode" id="cb328"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb328-1" title="1"><span class="kw">head</span>(<span class="kw">assay</span>(hl[[<span class="st">"psms"</span>]]))</a></code></pre></div>
<pre><code>##         X126      X127C       X127N      X128C       X128N      X129C
## 1 0.12283431 0.08045915 0.070804055 0.09386901 0.051815695 0.13034383
## 2 0.35268185 0.14162381 0.167523880 0.07843497 0.071087436 0.03214548
## 3 0.01546089 0.16142297 0.086938133 0.23120844 0.114664348 0.09610188
## 4 0.04702854 0.09288723 0.102012167 0.11125409 0.067969116 0.14155358
## 5 0.01044693 0.15866147 0.167315736 0.21017494 0.147946673 0.07088253
## 6 0.04955362 0.01215244 0.002477681 0.01297833 0.002988949 0.06253195
##        X129N       X130C      X130N       X131
## 1 0.17540095 0.040068658 0.11478839 0.11961594
## 2 0.06686260 0.031961793 0.02810434 0.02957384
## 3 0.15977819 0.010127118 0.08059400 0.04370403
## 4 0.18015910 0.035329902 0.12166589 0.10014038
## 5 0.17555789 0.007088253 0.02884754 0.02307803
## 6 0.01726511 0.172651119 0.37007905 0.29732174</code></pre>
<div class="sourceCode" id="cb330"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb330-1" title="1"><span class="kw">head</span>(<span class="kw">rowData</span>(hl[[<span class="st">"psms"</span>]]))</a></code></pre></div>
<pre><code>## DataFrame with 6 rows and 18 columns
##      Sequence ProteinDescriptions NbProteins ProteinGroupAccessions
##   &lt;character&gt;         &lt;character&gt;  &lt;integer&gt;            &lt;character&gt;
## 1     SQGEIDk       Tetratrico...          1                 Q8BYY4
## 2     YEAQGDk       Vacuolar p...          1                 P46467
## 3     TTScDTk       C-type man...          1                 Q64449
## 4     aEELESR       Liprin-alp...          1                 P60469
##   Modifications    qValue       PEP  IonScore NbMissedCleavages
##     &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt;         &lt;integer&gt;
## 1 K7(TMT6ple...     0.008   0.11800        27                 0
## 2 K7(TMT6ple...     0.001   0.01070        27                 0
## 3 C4(Carbami...     0.008   0.11800        11                 0
## 4 N-Term(TMT...     0.002   0.04450        24                 0
##   IsolationInterference IonInjectTimems Intensity    Charge      mzDa      MHDa
##               &lt;integer&gt;       &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt;
## 1                     0              70    335000         2   503.274   1005.54
## 2                     0              70    926000         2   520.267   1039.53
## 3                     0              70    159000         2   521.258   1041.51
## 4                     0              70    232000         2   531.785   1062.56
##   DeltaMassPPM     RTmin       markers
##      &lt;numeric&gt; &lt;numeric&gt;   &lt;character&gt;
## 1        -0.38     24.02       unknown
## 2         0.61     18.85       unknown
## 3         1.11     10.17       unknown
## 4         0.35     29.18       unknown
##  [ reached getOption("max.print") -- omitted 2 rows ]</code></pre>
<p>For further details on how to manipulate such objects, refer to the
<em><a href="https://bioconductor.org/packages/3.12/MultiAssayExperiment">MultiAssayExperiment</a></em> <span class="citation">(Ramos et al. <a href="#ref-MAE">2017</a>)</span> and
<em><a href="https://bioconductor.org/packages/3.12/SummerizedExperiment">SummerizedExperiment</a></em> <span class="citation">(Morgan et al. <a href="#ref-SE">2020</a>)</span> packages.</p>
<p>It is also possible to first create a <code>SummerizedExperiment</code>, and then
only include it into a <code>QFeatures</code> object.</p>
<div class="sourceCode" id="cb332"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb332-1" title="1">se &lt;-<span class="st"> </span><span class="kw">readSummarizedExperiment</span>(hlpsms, <span class="dt">ecol =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>)</a>
<a class="sourceLine" id="cb332-2" title="2">se</a></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 3010 10 
## metadata(0):
## assays(1): ''
## rownames(3010): 1 2 ... 3009 3010
## rowData names(18): Sequence ProteinDescriptions ... RTmin markers
## colnames(10): X126 X127C ... X130N X131
## colData names(0):</code></pre>
<div class="sourceCode" id="cb334"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb334-1" title="1"><span class="kw">QFeatures</span>(<span class="kw">list</span>(<span class="dt">psm =</span> se))</a></code></pre></div>
<pre><code>## An instance of class QFeatures containing 1 assays:
##  [1] psm: SummarizedExperiment with 3010 rows and 10 columns</code></pre>
<p>At this stage, i.e. at the beginning of the analysis, and respectively
whether you have a <code>SummerizedExperiment</code> or a <code>QFeatures</code> object, it
is a good time to define the experimental design in the <code>colData</code>
slot.</p>
<div id="exercice-1" class="section level3 unnumbered">
<h3>Exercice</h3>
<p>The CPTAC spike-in study 6 <span class="citation">(Paulovich et al. <a href="#ref-Paulovich:2010">2010</a>)</span> combines the Sigma UPS1
standard containing 48 different human proteins that are spiked in at
5 different concentrations (conditions A to E) into a constant yeast
protein background. The sample were acquired in triplicate on
different instruments in different labs. We are going to start with a
subset of the CPTAC study 6 containing conditions A and B for a single
lab.</p>
<div class="figure" style="text-align: center">
<span id="fig:cptac"></span>
<p class="caption marginnote shownote">
Figure 5.7: The CPTAC spike-in study design (credit Lieven Clement, statOmics, Ghent University).
</p>
<img src="img/cptac.png" alt="The CPTAC spike-in study design (credit Lieven Clement, statOmics, Ghent University)." width="70%">
</div>
<p>The peptide-level data, as processed by MaxQuant <span class="citation">(Cox and Mann <a href="#ref-Cox:2008">2008</a>)</span> is
available in the <code>msdata</code> package:</p>
<div class="sourceCode" id="cb336"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb336-1" title="1"><span class="kw">basename</span>(f &lt;-<span class="st"> </span>msdata<span class="op">::</span><span class="kw">quant</span>(<span class="dt">pattern =</span> <span class="st">"cptac"</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</a></code></pre></div>
<pre><code>## [1] "cptac_a_b_peptides.txt"</code></pre>
<p>Read these data in as either a <code>SummerizedExperiment</code> or a <code>QFeatures</code>
object and annotated the experiment.</p>
<details><p>From the names of the columns, we see that the quantitative columns,
starting with <code>"Intensity."</code> (note the dot!) are at positions 56 to
61.</p>
<div class="sourceCode" id="cb338"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb338-1" title="1"><span class="kw">names</span>(<span class="kw">read.delim</span>(f))</a></code></pre></div>
<pre><code>##  [1] "Sequence"                 "N.term.cleavage.window"  
##  [3] "C.term.cleavage.window"   "Amino.acid.before"       
##  [5] "First.amino.acid"         "Second.amino.acid"       
##  [7] "Second.last.amino.acid"   "Last.amino.acid"         
##  [9] "Amino.acid.after"         "A.Count"                 
## [11] "R.Count"                  "N.Count"                 
## [13] "D.Count"                  "C.Count"                 
## [15] "Q.Count"                  "E.Count"                 
## [17] "G.Count"                  "H.Count"                 
## [19] "I.Count"                  "L.Count"                 
## [21] "K.Count"                  "M.Count"                 
## [23] "F.Count"                  "P.Count"                 
## [25] "S.Count"                  "T.Count"                 
## [27] "W.Count"                  "Y.Count"                 
## [29] "V.Count"                  "U.Count"                 
## [31] "Length"                   "Missed.cleavages"        
## [33] "Mass"                     "Proteins"                
## [35] "Leading.razor.protein"    "Start.position"          
## [37] "End.position"             "Unique..Groups."         
## [39] "Unique..Proteins."        "Charges"                 
## [41] "PEP"                      "Score"                   
## [43] "Identification.type.6A_7" "Identification.type.6A_8"
## [45] "Identification.type.6A_9" "Identification.type.6B_7"
## [47] "Identification.type.6B_8" "Identification.type.6B_9"
## [49] "Experiment.6A_7"          "Experiment.6A_8"         
## [51] "Experiment.6A_9"          "Experiment.6B_7"         
## [53] "Experiment.6B_8"          "Experiment.6B_9"         
## [55] "Intensity"                "Intensity.6A_7"          
## [57] "Intensity.6A_8"           "Intensity.6A_9"          
## [59] "Intensity.6B_7"           "Intensity.6B_8"          
## [61] "Intensity.6B_9"           "Reverse"                 
## [63] "Potential.contaminant"    "id"                      
## [65] "Protein.group.IDs"        "Mod..peptide.IDs"        
## [67] "Evidence.IDs"             "MS.MS.IDs"               
## [69] "Best.MS.MS"               "Oxidation..M..site.IDs"  
## [71] "MS.MS.Count"</code></pre>
<div class="sourceCode" id="cb340"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb340-1" title="1">(i &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">"Intensity</span><span class="ch">\\</span><span class="st">."</span>, <span class="kw">names</span>(<span class="kw">read.delim</span>(f))))</a></code></pre></div>
<pre><code>## [1] 56 57 58 59 60 61</code></pre>
<p>We now read these data using the <code>readSummarizedExperiment</code>
function. This peptide-level expression data will be imported into R
as an instance of class <code>SummarizedExperiment</code>. We also use the
<code>fnames</code> argument to set the row-names of the <code>peptides</code> assay to the
peptide sequences and specify that the file is a tab-separated table.</p>
<div class="sourceCode" id="cb342"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb342-1" title="1">cptac_se &lt;-<span class="st"> </span><span class="kw">readSummarizedExperiment</span>(f, <span class="dt">ecol =</span> i, <span class="dt">fnames =</span> <span class="st">"Sequence"</span>, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</a>
<a class="sourceLine" id="cb342-2" title="2">cptac_se</a></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 11466 6 
## metadata(0):
## assays(1): ''
## rownames(11466): AAAAGAGGAGDSGDAVTK AAAALAGGK ... YYTVFDRDNNR
##   YYTVFDRDNNRVGFAEAAR
## rowData names(65): Sequence N.term.cleavage.window ...
##   Oxidation..M..site.IDs MS.MS.Count
## colnames(6): Intensity.6A_7 Intensity.6A_8 ... Intensity.6B_8
##   Intensity.6B_9
## colData names(0):</code></pre>
<p>Before proceeding, we are going to clean up the sample names and
annotate the experiment:</p>
<div class="sourceCode" id="cb344"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb344-1" title="1"><span class="kw">colnames</span>(cptac_se) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">"I.+</span><span class="ch">\\</span><span class="st">."</span>, <span class="st">""</span>, <span class="kw">colnames</span>(cptac_se))</a>
<a class="sourceLine" id="cb344-2" title="2">cptac_se<span class="op">$</span>condition &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">"_[7-9]"</span>, <span class="st">""</span>, <span class="kw">colnames</span>(cptac_se))</a>
<a class="sourceLine" id="cb344-3" title="3">cptac_se<span class="op">$</span>id &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">"^.+_"</span>, <span class="st">""</span>, <span class="kw">colnames</span>(cptac_se))</a>
<a class="sourceLine" id="cb344-4" title="4"><span class="kw">colData</span>(cptac_se)</a></code></pre></div>
<pre><code>## DataFrame with 6 rows and 2 columns
##        condition          id
##      &lt;character&gt; &lt;character&gt;
## 6A_7          6A           7
## 6A_8          6A           8
## 6A_9          6A           9
## 6B_7          6B           7
## 6B_8          6B           8
## 6B_9          6B           9</code></pre>
</details>
</div>
<div id="exercice-2" class="section level3 unnumbered">
<h3>Exercice</h3>
<p>There are many row variables that aren’t useful here. Get rid or all
of them but <code>Sequence</code>, <code>Proteins</code>, <code>Leading.razor.protein</code>, <code>PEP</code>,
<code>Score</code>, <code>Reverse</code>, and <code>Potential.contaminant</code>.</p>
<details><div class="sourceCode" id="cb346"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb346-1" title="1">keep_var &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"Sequence"</span>, <span class="st">"Proteins"</span>, <span class="st">"Leading.razor.protein"</span>, <span class="st">"PEP"</span>,</a>
<a class="sourceLine" id="cb346-2" title="2">              <span class="st">"Score"</span>, <span class="st">"Reverse"</span>, <span class="st">"Potential.contaminant"</span>)</a>
<a class="sourceLine" id="cb346-3" title="3"></a>
<a class="sourceLine" id="cb346-4" title="4"><span class="kw">rowData</span>(cptac_se) &lt;-<span class="st"> </span><span class="kw">rowData</span>(cptac_se)[, keep_var]</a></code></pre></div>
</details>
</div>
</div>
<div id="analysis-pipeline" class="section level2">
<h2>
<span class="header-section-number">5.4</span> Analysis pipeline</h2>
<p>A typical quantitative proteomics data processing is composed of the
following steps, which we are going to apply to the cptac data created
above.</p>
<ul>
<li>Data import</li>
<li>Exploratory data analysis (PCA)</li>
<li>Missing data management (filtering and/or imputation)</li>
<li>Data cleaning</li>
<li>Transformation and normalisation</li>
<li>Aggregation</li>
<li>Downstream analysis</li>
</ul>
<div class="sourceCode" id="cb347"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb347-1" title="1"><span class="kw">library</span>(<span class="st">"tidyverse"</span>)</a>
<a class="sourceLine" id="cb347-2" title="2"><span class="kw">library</span>(<span class="st">"ggplot2"</span>)</a>
<a class="sourceLine" id="cb347-3" title="3"><span class="kw">library</span>(<span class="st">"QFeatures"</span>)</a>
<a class="sourceLine" id="cb347-4" title="4"><span class="kw">library</span>(<span class="st">"limma"</span>)</a></code></pre></div>
<div id="missing-values" class="section level3">
<h3>
<span class="header-section-number">5.4.1</span> Missing values</h3>
<p>Missing values can be highly frequent in proteomics. These exist two
reasons supporting the existence of missing values, namely biological
or technical.</p>
<ol style="list-style-type: decimal">
<li><p>Values that are missing due to the absence (or extremely low
contentration) of a protein are observed for biological reasons,
and their pattern <strong>aren’t random</strong> (MNAR). A protein missing in
due to the suppression of its expression will not be missing at
random: it will be missing in the condition in which it was
suppressed, and be present in the condition where it is expressed.</p></li>
<li><p>Due to it’s data-dependent acquisition, mass spectrometry isn’t
capable to assaying all peptides in a sample. Peptides that are
less abundant than some of their co-eluting ions, peptides that do
not ionise well or peptides that do not get identified might be
sporadically missing in the final quantitation table, despite their
presence in the biological samples. Their absence patterns are
(completely) <strong>random</strong> (MAR or MCAR) in such cases.</p></li>
</ol>
<p>Often, third party software that produce quantiative data use zeros
instead of properly reporting missing values. We can use the
<code>zeroIsNA()</code> function to replace the <code>0</code> by <code>NA</code> values in our
<code>cptac_se</code> object and then explore the missing data patterns across
columns and rows.</p>
<div class="sourceCode" id="cb348"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb348-1" title="1">cptac_se &lt;-<span class="st"> </span><span class="kw">zeroIsNA</span>(cptac_se)</a>
<a class="sourceLine" id="cb348-2" title="2"><span class="kw">nNA</span>(cptac_se)</a></code></pre></div>
<pre><code>## $nNA
## DataFrame with 1 row and 2 columns
##         nNA       pNA
##   &lt;integer&gt; &lt;numeric&gt;
## 1     31130   45.2497
## 
## $nNArows
## DataFrame with 11466 rows and 3 columns
##                name       nNA       pNA
##         &lt;character&gt; &lt;integer&gt; &lt;numeric&gt;
## 1     AAAAGAGGAG...         4   66.6667
## 2         AAAALAGGK         0    0.0000
## 3        AAAALAGGKK         0    0.0000
## 4     AAADALSDLE...         0    0.0000
## 5     AAADALSDLE...         0    0.0000
## ...             ...       ...       ...
## 11462 YYSIYDLGNN...         6  100.0000
## 11463 YYTFNGPNYN...         3   50.0000
## 11464    YYTITEVATR         4   66.6667
## 11465 YYTVFDRDNN...         6  100.0000
## 11466 YYTVFDRDNN...         6  100.0000
## 
## $nNAcols
## DataFrame with 6 rows and 3 columns
##          name       nNA       pNA
##   &lt;character&gt; &lt;integer&gt; &lt;numeric&gt;
## 1        6A_7      4743   41.3658
## 2        6A_8      5483   47.8196
## 3        6A_9      5320   46.3980
## 4        6B_7      4721   41.1739
## 5        6B_8      5563   48.5174
## 6        6B_9      5300   46.2236</code></pre>
<div class="figure">
<span id="fig:imagena"></span>
<p class="caption marginnote shownote">
Figure 5.8: Distribution of missing value (white). Peptides row with more missing values are moved towards the top of the figure.
</p>
<img src="R4MS_files/figure-html/imagena-1.png" alt="Distribution of missing value (white). Peptides row with more missing values are moved towards the top of the figure." width="672">
</div>
<p>Let’s now explore these missing values:</p>
<ul>
<li>Explore the number or proportion of missing values across peptides
and samples of the <code>cptac_se</code> data.</li>
</ul>
<div class="sourceCode" id="cb350"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb350-1" title="1"><span class="kw">barplot</span>(<span class="kw">nNA</span>(cptac_se)<span class="op">$</span>nNAcols<span class="op">$</span>pNA)</a></code></pre></div>
<p><img src="R4MS_files/figure-html/na2-1.png" width="672"></p>
<div class="sourceCode" id="cb351"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb351-1" title="1"><span class="kw">table</span>(<span class="kw">nNA</span>(cptac_se)<span class="op">$</span>nNArows<span class="op">$</span>nNA)</a></code></pre></div>
<pre><code>## 
##    0    1    2    3    4    5    6 
## 4059  990  884  717  934  807 3075</code></pre>
<ul>
<li>Remove row that have <em>too many</em> missing values. You can do this by
hand or using the <code>filterNA()</code> function.</li>
</ul>
<div class="sourceCode" id="cb353"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb353-1" title="1"><span class="co">## remove rows that have 4 or more NAs out of 6</span></a>
<a class="sourceLine" id="cb353-2" title="2">cptac_se &lt;-<span class="st"> </span><span class="kw">filterNA</span>(cptac_se, <span class="dt">pNA =</span> <span class="dv">4</span><span class="op">/</span><span class="dv">6</span>)</a></code></pre></div>
</div>
<div id="imputation" class="section level3">
<h3>
<span class="header-section-number">5.4.2</span> Imputation</h3>
<p>Imputation is the technique of replacing missing data with probable
values. This can be done with <code>impute()</code> method. As we have discussed
above, there are however two types of missing values in mass
spectrometry-based proteomics, namely data missing at random (MAR),
and data missing not at random (MNAR). These two types of missing
data, those missing at random, and those missing not at random, need
to be imputed with <a href="https://rformassspectrometry.github.io/QFeatures/articles/Processing.html#imputation-1">different types of imputation
methods</a>
<span class="citation">(Lazar et al. <a href="#ref-Lazar:2016">2016</a>)</span>.</p>
<div class="figure">
<span id="fig:miximp"></span>
<p class="caption marginnote shownote">
Figure 5.9: Mixed imputation method. Black cells represent presence of quantitation values and light grey corresponds to missing data. The two groups of interest are depicted in green and blue along the heatmap columns. Two classes of proteins are annotated on the left: yellow are proteins with randomly occurring missing values (if any) while proteins in brown are candidates for non-random missing value imputation.
</p>
<img src="R4MS_files/figure-html/miximp-1.png" alt="Mixed imputation method. Black cells represent presence of quantitation values and light grey corresponds to missing data. The two groups of interest are depicted in green and blue along the heatmap columns. Two classes of proteins are annotated on the left: yellow are proteins with randomly occurring missing values (if any) while proteins in brown are candidates for non-random missing value imputation." width="672">
</div>
<div class="figure">
<span id="fig:lazar"></span>
<p class="caption marginnote shownote">
Figure 5.10: Effect of the nature of missing values on their imputation. Root-mean-square error (RMSE) observations standard deviation ratio (RSR), KNN and MinDet imputation. Lower (blue) is better.
</p>
<img src="img/imp-sim.png" alt="Effect of the nature of missing values on their imputation. Root-mean-square error (RMSE) observations standard deviation ratio (RSR), KNN and MinDet imputation. Lower (blue) is better." width="100%">
</div>
<p>Generally, it is recommended to use <strong>hot deck</strong> methods (nearest
neighbour (<strong>left</strong>), maximum likelihood, …) when data are missing
at random.Conversely, MNAR features should ideally be imputed with a
<strong>left-censor</strong> (minimum value (<strong>right</strong>), but not zero, …) method.</p>
<p>There are various methods to perform data imputation, as described in
<code>?impute</code>. The <em><a href="https://CRAN.R-project.org/package=imp4p">imp4p</a></em> package contains additional
functionality, including some to estimate the randomness of missing
data.</p>
<p>The general syntax for imputation is</p>
<div class="sourceCode" id="cb354"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb354-1" title="1"><span class="kw">data</span>(se_na2)</a>
<a class="sourceLine" id="cb354-2" title="2"><span class="co">## impute missing values using knn imputation</span></a>
<a class="sourceLine" id="cb354-3" title="3"><span class="kw">impute</span>(se_na2, <span class="dt">method =</span> <span class="st">"knn"</span>)</a></code></pre></div>
<pre><code>## Warning in knnimp(x, k, maxmiss = rowmax, maxp = maxp): 12 rows with more than 50 % entries missing;
##  mean imputation used for these rows</code></pre>
<pre><code>## class: SummarizedExperiment 
## dim: 689 16 
## metadata(3): MSnbaseFiles MSnbaseProcessing MSnbaseVersion
## assays(1): ''
## rownames(689): AT1G09210 AT1G21750 ... AT4G11150 AT4G39080
## rowData names(2): nNA randna
## colnames(16): M1F1A M1F4A ... M2F8B M2F11B
## colData names(1): nNA</code></pre>
<div id="exercise-15" class="section level4 unnumbered">
<h4>Exercise</h4>
<p>Following the example above, apply a mixed imputation, using knn for
data missing at random and the deterministic minumum left-cencored
imputation for data missing no at random.</p>
<details><div class="sourceCode" id="cb357"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb357-1" title="1"><span class="kw">impute</span>(se_na2, <span class="st">"mixed"</span>,</a>
<a class="sourceLine" id="cb357-2" title="2">       <span class="dt">randna =</span> <span class="kw">rowData</span>(se_na2)<span class="op">$</span>randna,</a>
<a class="sourceLine" id="cb357-3" title="3">       <span class="dt">mar =</span> <span class="st">"knn"</span>, <span class="dt">mnar =</span> <span class="st">"MinDet"</span>)</a></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 689 16 
## metadata(3): MSnbaseFiles MSnbaseProcessing MSnbaseVersion
## assays(1): ''
## rownames(689): AT1G09210 AT1G21750 ... AT4G11150 AT4G39080
## rowData names(2): nNA randna
## colnames(16): M1F1A M1F4A ... M2F8B M2F11B
## colData names(1): nNA</code></pre>
</details>
</div>
<div id="exercise-16" class="section level4 unnumbered">
<h4>Exercise</h4>
<p>When assessing missing data imputation methods, such as in <a href="https://pubs.acs.org/doi/abs/10.1021/acs.jproteome.5b00981">Lazar et
al. (2016)</a>,
one often replaces values with missing data, imputes these with a
method of choice, then quantifies the difference between original
(expected) and observed (imputed) values. Here, using the <code>se_na2</code>
data, use this strategy to assess the difference between knn and
Bayesian PCA imputation.</p>
<details><div class="sourceCode" id="cb359"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb359-1" title="1">imp1 &lt;-<span class="st"> </span><span class="kw">impute</span>(se_na2, <span class="dt">method =</span> <span class="st">"knn"</span>)</a></code></pre></div>
<pre><code>## Warning in knnimp(x, k, maxmiss = rowmax, maxp = maxp): 12 rows with more than 50 % entries missing;
##  mean imputation used for these rows</code></pre>
<div class="sourceCode" id="cb361"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb361-1" title="1">imp2 &lt;-<span class="st"> </span><span class="kw">impute</span>(se_na2, <span class="dt">method =</span> <span class="st">"bpca"</span>)</a>
<a class="sourceLine" id="cb361-2" title="2"><span class="kw">summary</span>(<span class="kw">abs</span>(<span class="kw">assay</span>(imp1)[<span class="kw">is.na</span>(<span class="kw">assay</span>(se_na2))] <span class="op">-</span><span class="st"> </span><span class="kw">assay</span>(imp2)[<span class="kw">is.na</span>(<span class="kw">assay</span>(se_na2))]))</a></code></pre></div>
<pre><code>##      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
## 5.332e-05 6.594e-03 1.535e-02 2.315e-02 2.855e-02 2.579e-01</code></pre>
<div class="sourceCode" id="cb363"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb363-1" title="1"><span class="kw">summary</span>(<span class="kw">as.numeric</span>(<span class="kw">na.omit</span>(<span class="kw">assay</span>(se_na2))))</a></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##  0.0170  0.1865  0.2440  0.2500  0.3080  0.6587</code></pre>
</details>
</div>
<div id="exercise-17" class="section level4 unnumbered">
<h4>Exercise</h4>
<p>When assessing the impact of missing value imputation on real data,
one can’t use the strategy above. Another useful approach is to assess
the impact of the imputation method on the distribution of the
quantitative data. For instance, here is the intensity distribution of
the <code>se_na2</code> data. Verify the effect of applying <code>knn</code>, <code>zero</code>,
<code>MinDet</code> and <code>bpca</code> on this distribution.</p>
<div class="sourceCode" id="cb365"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb365-1" title="1"><span class="kw">plot</span>(<span class="kw">density</span>(<span class="kw">na.omit</span>(<span class="kw">assay</span>(se_na2))))</a></code></pre></div>
<div class="figure">
<span id="fig:nasetdist"></span>
<p class="caption marginnote shownote">
Figure 5.11: Intensity disctribution of the <code>naset</code> data.
</p>
<img src="R4MS_files/figure-html/nasetdist-1.png" alt="Intensity disctribution of the `naset` data." width="672">
</div>
<details><div class="sourceCode" id="cb366"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb366-1" title="1">cls &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"black"</span>, <span class="st">"red"</span>, <span class="st">"blue"</span>, <span class="st">"steelblue"</span>, <span class="st">"orange"</span>)</a>
<a class="sourceLine" id="cb366-2" title="2"><span class="kw">plot</span>(<span class="kw">density</span>(<span class="kw">na.omit</span>(<span class="kw">assay</span>(se_na2))), <span class="dt">col =</span> cls[<span class="dv">1</span>])</a>
<a class="sourceLine" id="cb366-3" title="3"><span class="kw">lines</span>(<span class="kw">density</span>(<span class="kw">assay</span>(<span class="kw">impute</span>(se_na2, <span class="dt">method =</span> <span class="st">"knn"</span>))), <span class="dt">col =</span> cls[<span class="dv">2</span>])</a></code></pre></div>
<pre><code>## Warning in knnimp(x, k, maxmiss = rowmax, maxp = maxp): 12 rows with more than 50 % entries missing;
##  mean imputation used for these rows</code></pre>
<div class="sourceCode" id="cb368"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb368-1" title="1"><span class="kw">lines</span>(<span class="kw">density</span>(<span class="kw">assay</span>(<span class="kw">impute</span>(se_na2, <span class="dt">method =</span> <span class="st">"zero"</span>))), <span class="dt">col =</span> cls[<span class="dv">3</span>])</a>
<a class="sourceLine" id="cb368-2" title="2"><span class="kw">lines</span>(<span class="kw">density</span>(<span class="kw">assay</span>(<span class="kw">impute</span>(se_na2, <span class="dt">method =</span> <span class="st">"MinDet"</span>))), <span class="dt">col =</span> cls[<span class="dv">4</span>])</a>
<a class="sourceLine" id="cb368-3" title="3"><span class="kw">lines</span>(<span class="kw">density</span>(<span class="kw">assay</span>(<span class="kw">impute</span>(se_na2, <span class="dt">method =</span> <span class="st">"bpca"</span>))), <span class="dt">col =</span> cls[<span class="dv">5</span>])</a>
<a class="sourceLine" id="cb368-4" title="4"><span class="kw">legend</span>(<span class="st">"topright"</span>, <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">"orig"</span>, <span class="st">"knn"</span>, <span class="st">"zero"</span>, <span class="st">"MinDet"</span>, <span class="st">"bpca"</span>),</a>
<a class="sourceLine" id="cb368-5" title="5">       <span class="dt">col =</span> cls, <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">bty =</span> <span class="st">"n"</span>)</a></code></pre></div>
<img src="R4MS_files/figure-html/naex3-1.png" width="672"></details><p><strong>Tip</strong>: When downstream analyses permit, it might be safer not to
impute data and deal explicitly with missing values. Indeed misssing
data impuration is however not a straightforward thing, as is likely
to dramatically fail when a high proportion of data is missing (10s of
%). It is possible to keep NAs when performing hypothesis
tests<label for="tufte-sn-6" class="margin-toggle sidenote-number">6</label><input type="checkbox" id="tufte-sn-6" class="margin-toggle"><span class="sidenote"><span class="sidenote-number">6</span> Still, it is recommended to explore missingness as part of the
exploratory data analysis.</span>, but not to perform a principal component
analysis.</p>
</div>
</div>
<div id="identification-quality-control" class="section level3">
<h3>
<span class="header-section-number">5.4.3</span> Identification quality control</h3>
<p>As discussed in the previous chapter, PSMs are deemed relevant after
comparison against hist from a decoy database. The origin of these
hits is recorded with <code>+</code> in the <code>Reverse</code> variable:</p>
<div class="sourceCode" id="cb369"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb369-1" title="1"><span class="kw">table</span>(<span class="kw">rowData</span>(cptac_se)<span class="op">$</span>Reverse)</a></code></pre></div>
<pre><code>## 
##         + 
## 7572   12</code></pre>
<p>Similarly, a proteomics experiment is also searched against a database
of contaminants:</p>
<div class="sourceCode" id="cb371"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb371-1" title="1"><span class="kw">table</span>(<span class="kw">rowData</span>(cptac_se)<span class="op">$</span>Potential.contaminant)</a></code></pre></div>
<pre><code>## 
##         + 
## 7558   26</code></pre>
<p>Let’s visualise some of the cptac’s metadata using standard <code>ggplot2</code>
code:</p>
<div id="exercise-18" class="section level4 unnumbered">
<h4>Exercise</h4>
<p>Visualise the identification score and the posterior probability
probability (PEP) distributions from forward and reverse hits and
interpret the figure.</p>
<details><div class="sourceCode" id="cb373"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb373-1" title="1"><span class="kw">rowData</span>(cptac_se) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb373-2" title="2"><span class="st">    </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb373-3" title="3"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> Score, <span class="dt">colour =</span> Reverse)) <span class="op">+</span></a>
<a class="sourceLine" id="cb373-4" title="4"><span class="st">    </span><span class="kw">geom_density</span>()</a></code></pre></div>
<p><img src="R4MS_files/figure-html/idqc1-1.png" width="672"></p>
<div class="sourceCode" id="cb374"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb374-1" title="1"><span class="kw">rowData</span>(cptac_se) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb374-2" title="2"><span class="st">    </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb374-3" title="3"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> PEP, <span class="dt">colour =</span> Reverse)) <span class="op">+</span></a>
<a class="sourceLine" id="cb374-4" title="4"><span class="st">    </span><span class="kw">geom_density</span>()</a></code></pre></div>
<p><img src="R4MS_files/figure-html/idqc2-1.png" width="672"></p>
</details>
</div>
</div>
<div id="creating-the-qfeatures-data" class="section level3">
<h3>
<span class="header-section-number">5.4.4</span> Creating the QFeatures data</h3>
<p>We can now create our <code>QFeatures</code> object using the
<code>SummarizedExperiment</code> as show below.</p>
<div class="sourceCode" id="cb375"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb375-1" title="1">cptac &lt;-<span class="st"> </span><span class="kw">QFeatures</span>(<span class="kw">list</span>(<span class="dt">peptides =</span> cptac_se))</a>
<a class="sourceLine" id="cb375-2" title="2">cptac</a></code></pre></div>
<pre><code>## An instance of class QFeatures containing 1 assays:
##  [1] peptides: SummarizedExperiment with 7584 rows and 6 columns</code></pre>
<p>We should also assign the <code>QFeatures</code> column data with the
<code>SummarizedExperiment</code> slot.</p>
<div class="sourceCode" id="cb377"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb377-1" title="1"><span class="kw">colData</span>(cptac) &lt;-<span class="st"> </span><span class="kw">colData</span>(cptac_se)</a></code></pre></div>
<p>Note that it is also possible to directly create a <code>QFeatures</code> object
with the <code>readQFeatures()</code> function and the same arguments as the
<code>readSummarizedExperiment()</code> used above. In addition, most functions
used above and below work on single <code>SummarizedExperiment</code> objects or
assays within a <code>QFeatures</code> object.</p>
</div>
<div id="filtering-out-contaminants-and-reverse-hits" class="section level3">
<h3>
<span class="header-section-number">5.4.5</span> Filtering out contaminants and reverse hits</h3>
<div id="exercise-19" class="section level4 unnumbered">
<h4>Exercise</h4>
<p>Using the <code>filterFeatures()</code> function, filter out the reverse and
contaminant hits, and also retein those that have a posterior error
probability smaller than 0.05.</p>
<details><div class="sourceCode" id="cb378"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb378-1" title="1">cptac &lt;-</a>
<a class="sourceLine" id="cb378-2" title="2"><span class="st">    </span>cptac <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb378-3" title="3"><span class="st">    </span><span class="kw">filterFeatures</span>(<span class="op">~</span><span class="st"> </span>Reverse <span class="op">!=</span><span class="st"> "+"</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb378-4" title="4"><span class="st">    </span><span class="kw">filterFeatures</span>(<span class="op">~</span><span class="st"> </span>Potential.contaminant <span class="op">!=</span><span class="st"> "+"</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb378-5" title="5"><span class="st">    </span><span class="kw">filterFeatures</span>(<span class="op">~</span><span class="st"> </span>PEP <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)</a></code></pre></div>
</details>
</div>
</div>
<div id="log-transformation-and-normaliation" class="section level3">
<h3>
<span class="header-section-number">5.4.6</span> Log-transformation and normaliation</h3>
<p>The two code chunks below log-transform and normalise using the assay
<code>i</code> as input and adding a new one names as defined by <code>name</code>.</p>
<div class="sourceCode" id="cb379"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb379-1" title="1">cptac &lt;-<span class="st"> </span><span class="kw">logTransform</span>(cptac, <span class="dt">i =</span> <span class="st">"peptides"</span>,</a>
<a class="sourceLine" id="cb379-2" title="2">                      <span class="dt">name =</span> <span class="st">"log_peptides"</span>)</a></code></pre></div>
<div id="exercise-20" class="section level4 unnumbered">
<h4>Exercise</h4>
<p>Use the <code>normalize()</code> method to normalise the data. The syntax is the
same as <code>logTransform()</code>. Use the <code>centre.median</code> method.</p>
<details><div class="sourceCode" id="cb380"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb380-1" title="1">cptac &lt;-<span class="st"> </span><span class="kw">normalize</span>(cptac, <span class="dt">i =</span> <span class="st">"log_peptides"</span>,</a>
<a class="sourceLine" id="cb380-2" title="2">                   <span class="dt">name =</span> <span class="st">"lognorm_peptides"</span>,</a>
<a class="sourceLine" id="cb380-3" title="3">                   <span class="dt">method =</span> <span class="st">"center.median"</span>)</a></code></pre></div>
</details>
</div>
<div id="exercise-21" class="section level4 unnumbered">
<h4>Exercise</h4>
<p>Visualise the result of the transformations above. The
<code>plotDensities()</code> function from the <code>limma</code> package is very
convenient, but feel free to use boxplots, violin plots, or any other
visualisation that you deem useful to assess the tranformations.</p>
<details><div class="sourceCode" id="cb381"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb381-1" title="1"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>))</a>
<a class="sourceLine" id="cb381-2" title="2">limma<span class="op">::</span><span class="kw">plotDensities</span>(<span class="kw">assay</span>(cptac[[<span class="st">"peptides"</span>]]))</a>
<a class="sourceLine" id="cb381-3" title="3">limma<span class="op">::</span><span class="kw">plotDensities</span>(<span class="kw">assay</span>(cptac[[<span class="st">"log_peptides"</span>]]))</a>
<a class="sourceLine" id="cb381-4" title="4">limma<span class="op">::</span><span class="kw">plotDensities</span>(<span class="kw">assay</span>(cptac[[<span class="st">"lognorm_peptides"</span>]]))</a></code></pre></div>
<div class="figure">
<span id="fig:plotdens"></span>
<p class="caption marginnote shownote">
Figure 5.12: Three peptide level assays: raw data, log transformed and normalised.
</p>
<img src="R4MS_files/figure-html/plotdens-1.png" alt="Three peptide level assays: raw data, log transformed and normalised." width="1440">
</div>
</details>
</div>
</div>
<div id="aggregation" class="section level3">
<h3>
<span class="header-section-number">5.4.7</span> Aggregation</h3>
<div id="exercise-22" class="section level4 unnumbered">
<h4>Exercise</h4>
<p>Use median aggregation to aggregation peptides into protein
values. This is not necessarily not the best choice, as we will see
later, but a good star.</p>
<details><div class="sourceCode" id="cb382"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb382-1" title="1">cptac &lt;-</a>
<a class="sourceLine" id="cb382-2" title="2"><span class="st">    </span><span class="kw">aggregateFeatures</span>(cptac,</a>
<a class="sourceLine" id="cb382-3" title="3">                      <span class="st">"lognorm_peptides"</span>,</a>
<a class="sourceLine" id="cb382-4" title="4">                      <span class="dt">name =</span> <span class="st">"proteins_med"</span>,</a>
<a class="sourceLine" id="cb382-5" title="5">                      <span class="dt">fcol =</span> <span class="st">"Leading.razor.protein"</span>,</a>
<a class="sourceLine" id="cb382-6" title="6">                      <span class="dt">fun =</span> colMedians,</a>
<a class="sourceLine" id="cb382-7" title="7">                      <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
</details><p>Looking at the <code>.n</code> row variable computed during the aggregation, we
see that most proteins result of the aggregation of 5 peptides or
less, while very few proteins are accounted for by tens of peptides.</p>
<div class="sourceCode" id="cb383"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb383-1" title="1"><span class="kw">table</span>(<span class="kw">rowData</span>(cptac[[<span class="st">"proteins_med"</span>]])<span class="op">$</span>.n)</a></code></pre></div>
<pre><code>## 
##   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20 
## 327 234 167 132  84  73  62  49  49  29  29  24  20  13  15  12   4   6  11   5 
##  21  22  23  24  25  26  28  29  30  31  32  34  37  38  39  42  51  52  62 
##   7   4   7   2   2   3   1   3   1   2   2   1   1   1   1   2   1   1   1</code></pre>
</div>
</div>
<div id="principal-component-analysis" class="section level3">
<h3>
<span class="header-section-number">5.4.8</span> Principal component analysis</h3>
<div class="sourceCode" id="cb385"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb385-1" title="1"><span class="kw">library</span>(<span class="st">"factoextra"</span>)</a>
<a class="sourceLine" id="cb385-2" title="2"></a>
<a class="sourceLine" id="cb385-3" title="3">pca_pep &lt;-</a>
<a class="sourceLine" id="cb385-4" title="4"><span class="st">    </span>cptac[[<span class="st">"lognorm_peptides"</span>]] <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb385-5" title="5"><span class="st">    </span><span class="kw">filterNA</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb385-6" title="6"><span class="st">    </span><span class="kw">assay</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb385-7" title="7"><span class="st">    </span><span class="kw">t</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb385-8" title="8"><span class="st">    </span><span class="kw">prcomp</span>(<span class="dt">scale =</span> <span class="ot">TRUE</span>, <span class="dt">center =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb385-9" title="9"><span class="st">    </span><span class="kw">fviz_pca_ind</span>(<span class="dt">habillage =</span> cptac<span class="op">$</span>condition, <span class="dt">title =</span> <span class="st">"Peptides"</span>)</a>
<a class="sourceLine" id="cb385-10" title="10"></a>
<a class="sourceLine" id="cb385-11" title="11">pca_prot &lt;-</a>
<a class="sourceLine" id="cb385-12" title="12"><span class="st">    </span>cptac[[<span class="st">"proteins_med"</span>]] <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb385-13" title="13"><span class="st">    </span><span class="kw">filterNA</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb385-14" title="14"><span class="st">    </span><span class="kw">assay</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb385-15" title="15"><span class="st">    </span><span class="kw">t</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb385-16" title="16"><span class="st">    </span><span class="kw">prcomp</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb385-17" title="17"><span class="st">    </span><span class="kw">fviz_pca_ind</span>(<span class="dt">habillage =</span> cptac<span class="op">$</span>condition,</a>
<a class="sourceLine" id="cb385-18" title="18">                 <span class="dt">title =</span> <span class="st">"Proteins (median aggregation)"</span>)</a></code></pre></div>
<div class="sourceCode" id="cb386"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb386-1" title="1"><span class="kw">library</span>(<span class="st">"patchwork"</span>)</a>
<a class="sourceLine" id="cb386-2" title="2">pca_pep <span class="op">+</span><span class="st"> </span>pca_prot</a></code></pre></div>
<div class="figure">
<span id="fig:plotpca"></span>
<p class="caption marginnote shownote">
Figure 5.13: Peptide and protein level PCA analyses.
</p>
<img src="R4MS_files/figure-html/plotpca-1.png" alt="Peptide and protein level PCA analyses." width="1152">
</div>
</div>
<div id="visualisation" class="section level3">
<h3>
<span class="header-section-number">5.4.9</span> Visualisation</h3>
<p>Below, we use the <code>longFormat()</code> function to extract the quantitative
and row data in a long format, that can be directly reused by the
tidyverse tools.</p>
<div class="sourceCode" id="cb387"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb387-1" title="1"><span class="kw">longFormat</span>(cptac[<span class="st">"P02787ups|TRFE_HUMAN_UPS"</span>, ,</a>
<a class="sourceLine" id="cb387-2" title="2">                 <span class="kw">c</span>(<span class="st">"lognorm_peptides"</span>, <span class="st">"proteins_med"</span>)]) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb387-3" title="3"><span class="st">    </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb387-4" title="4"><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">condition =</span> <span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">"A"</span>, colname), <span class="st">"A"</span>, <span class="st">"B"</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb387-5" title="5"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> colname, <span class="dt">y =</span> value, <span class="dt">colour =</span> rowname, <span class="dt">shape =</span> condition)) <span class="op">+</span></a>
<a class="sourceLine" id="cb387-6" title="6"><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">3</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb387-7" title="7"><span class="st">    </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="dt">group =</span> rowname)) <span class="op">+</span></a>
<a class="sourceLine" id="cb387-8" title="8"><span class="st">    </span><span class="kw">facet_grid</span>(<span class="op">~</span><span class="st"> </span>assay) <span class="op">+</span></a>
<a class="sourceLine" id="cb387-9" title="9"><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">"P02787ups|TRFE_HUMAN_UPS"</span>)</a></code></pre></div>
<div class="figure">
<span id="fig:vis"></span>
<p class="caption marginnote shownote">
Figure 5.14: Peptide and protein expression profile.
</p>
<img src="R4MS_files/figure-html/vis-1.png" alt="Peptide and protein expression profile." width="1152">
</div>
</div>
<div id="statistical-analysis" class="section level3">
<h3>
<span class="header-section-number">5.4.10</span> Statistical analysis</h3>
<p>R in general and Bioconductor in particular are well suited for the
statistical analysis of data of quantitative proteomics data. Several
packages provide dedicated resources for proteomics data:</p>
<ul>
<li><p><em><a href="https://bioconductor.org/packages/3.12/MSstats">MSstats</a></em> and <em><a href="https://bioconductor.org/packages/3.12/MSstatsTMT">MSstatsTMT</a></em>: A set of tools
for statistical relative protein significanceanalysis in Data dependent
(DDA), SRM, Data independent acquisition (DIA) and TMT experiments.</p></li>
<li><p><em><a href="https://bioconductor.org/packages/3.12/msmsTests">msmsTests</a></em>: Statistical tests for label-free LC-MS/MS
data by spectral counts, to discover differentially expressed
proteins between two biological conditions. Three tests are
available: Poisson GLM regression, quasi-likelihood GLM regression,
and the negative binomial of the <em><a href="https://bioconductor.org/packages/3.12/edgeR">edgeR</a></em>
package. All can be readily applied on <code>MSnSet</code> instances produced,
for example by <code>MSnID</code>.</p></li>
<li><p><em><a href="https://bioconductor.org/packages/3.12/DEP">DEP</a></em> provides an integrated analysis workflow for the
analysis of mass spectrometry proteomics data for differential
protein expression or differential enrichment.</p></li>
<li><p><em><a href="https://github.com/statOmics/MSqRob">MSqRob</a></em>: The <code>MSqRob</code> package
allows a user to do quantitative protein-level statistical inference
on LC-MS proteomics data. More specifically, our package makes use
of peptide-level input data, thus correcting for unbalancedness and
peptide-specific biases. As previously shown (<a href="https://pubs.acs.org/doi/abs/10.1021/pr501223t">Goeminne et
al. (2015)</a>), this
approach is both more sensitive and specific than summarizing
peptide-level input to protein-level values. Model estimates are
stabilized by ridge regression, empirical Bayes variance estimation
and downweighing of outliers. Currently, only label-free proteomics
data types are supported. <code>msqrob2</code> is currently under development
and will be using <code>QFeatures</code> objects.</p></li>
<li><p><em><a href="https://github.com/const-ae/proDA">proDA</a></em> accounts for missing
values in label-free mass spectrometry data without imputation. The
package implements a probabilistic dropout model that ensures that
the information from observed and missing values are properly
combined. It adds empirical Bayesian priors to increase power to
detect differentially abundant proteins.</p></li>
</ul>
<p>Others, while not specfic to proteomics, are also recommended, such as
the <em><a href="https://bioconductor.org/packages/3.12/limma">limma</a></em> package. When analysing spectral counting
data, methods for high throughput sequencing data are
applicable. Below, we illustrate how to apply a typical <code>edgeR</code> test
to count data using the <code>msms.edgeR</code> function from the <code>msmsTests</code>
package.</p>
<p>Below, we are going to perform our statistical analysis on the protein
data using <code>limma</code>.</p>
<div class="sourceCode" id="cb388"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb388-1" title="1">prots &lt;-<span class="st"> </span>cptac[[<span class="st">"proteins_med"</span>]]</a>
<a class="sourceLine" id="cb388-2" title="2"><span class="kw">colData</span>(prots) &lt;-<span class="st"> </span><span class="kw">colData</span>(cptac)</a></code></pre></div>
<p>The <em><a href="https://bioconductor.org/packages/3.12/limma">limma</a></em> package is the precursor package
that enables the consistent application of linear models to normalliy
distributed omics data in general, and microarrays in
particuar.</p>
<p>The <code>limma</code> package implements an empirical Bayes method that provides
borrows information across features to estimate the standard error and
calculate (so called moderate) t statistics. This approach is
demonstrably more powerful that a standard t-tests when the number of
samples is lot.</p>
<p>The code chunk below illstrated how to set up the model, fit it, and
apply the empirical Bayes moderation.</p>
<div class="sourceCode" id="cb389"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb389-1" title="1"><span class="kw">library</span>(<span class="st">"limma"</span>)</a>
<a class="sourceLine" id="cb389-2" title="2">design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span>prots<span class="op">$</span>condition)</a>
<a class="sourceLine" id="cb389-3" title="3">fit &lt;-<span class="st"> </span><span class="kw">lmFit</span>(<span class="kw">assay</span>(prots), design)</a></code></pre></div>
<pre><code>## Warning: Partial NA coefficients for 25 probe(s)</code></pre>
<div class="sourceCode" id="cb391"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb391-1" title="1">fit &lt;-<span class="st"> </span><span class="kw">eBayes</span>(fit)</a></code></pre></div>
<p>Finally, the <code>topTable()</code> function is used the extract the results for
the coefficient of interest.</p>
<div class="sourceCode" id="cb392"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb392-1" title="1">res &lt;-</a>
<a class="sourceLine" id="cb392-2" title="2"><span class="st">    </span><span class="kw">topTable</span>(fit, <span class="dt">coef =</span> <span class="st">"prots$condition6B"</span>, <span class="dt">number =</span> <span class="ot">Inf</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb392-3" title="3"><span class="st">    </span><span class="kw">rownames_to_column</span>(<span class="st">"protein"</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb392-4" title="4"><span class="st">    </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb392-5" title="5"><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">TP =</span> <span class="kw">grepl</span>(<span class="st">"ups"</span>, protein))</a></code></pre></div>
<p>Note the warning about partial <code>NA</code> coefficients for 23 probes:</p>
<div class="sourceCode" id="cb393"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb393-1" title="1">na_coefs &lt;-</a>
<a class="sourceLine" id="cb393-2" title="2"><span class="st">    </span><span class="kw">filter</span>(res, <span class="kw">is.na</span>(t)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb393-3" title="3"><span class="st">    </span><span class="kw">pull</span>(protein)</a>
<a class="sourceLine" id="cb393-4" title="4"><span class="kw">assay</span>(prots[na_coefs, ])</a></code></pre></div>
<pre><code>##                                6A_7      6A_8       6A_9       6B_7       6B_8
## P00167ups|CYB5_HUMAN_UPS        NaN       NaN        NaN -0.7840558 -2.0282987
## P01112ups|RASH_HUMAN_UPS        NaN       NaN        NaN -1.5564896        NaN
## P05413ups|FABPH_HUMAN_UPS       NaN       NaN        NaN -3.3419480        NaN
## P08758ups|ANXA5_HUMAN_UPS       NaN       NaN        NaN -2.7973872 -2.0137585
## sp|P06704|CDC31_YEAST           NaN       NaN        NaN -1.2032046 -2.1252371
## sp|P25574|EMC1_YEAST      -1.506177 -1.983737 -0.7795009        NaN        NaN
## sp|P32608|RTG2_YEAST            NaN       NaN        NaN        NaN -4.4424189
## sp|P32769|HBS1_YEAST            NaN -1.384031 -0.7285780        NaN        NaN
## sp|P34217|PIN4_YEAST            NaN       NaN        NaN -0.8378614 -0.1316397
## sp|P34237|CASP_YEAST            NaN       NaN        NaN -1.5645172 -1.6600291
## sp|P38166|SFT2_YEAST      -1.585685 -1.076707        NaN        NaN        NaN
## sp|P40056|GET2_YEAST            NaN -1.091696 -1.4014211        NaN        NaN
## sp|P40533|TED1_YEAST            NaN       NaN        NaN -2.0491876        NaN
## sp|P43582|WWM1_YEAST            NaN       NaN        NaN -0.5538711 -0.7360990
## sp|P46965|SPC1_YEAST            NaN -3.428771 -3.6321984        NaN        NaN
## sp|P48363|PFD3_YEAST            NaN       NaN        NaN -0.1904905        NaN
##                                 6B_9
## P00167ups|CYB5_HUMAN_UPS  -1.1230809
## P01112ups|RASH_HUMAN_UPS  -1.5618192
## P05413ups|FABPH_HUMAN_UPS -3.8907081
## P08758ups|ANXA5_HUMAN_UPS -2.0894752
## sp|P06704|CDC31_YEAST     -1.5844104
## sp|P25574|EMC1_YEAST             NaN
## sp|P32608|RTG2_YEAST      -2.7873186
## sp|P32769|HBS1_YEAST             NaN
## sp|P34217|PIN4_YEAST      -0.1989392
## sp|P34237|CASP_YEAST      -1.6877463
## sp|P38166|SFT2_YEAST             NaN
## sp|P40056|GET2_YEAST             NaN
## sp|P40533|TED1_YEAST      -1.7474812
## sp|P43582|WWM1_YEAST      -0.7207043
## sp|P46965|SPC1_YEAST             NaN
## sp|P48363|PFD3_YEAST      -0.5087747
##  [ reached getOption("max.print") -- omitted 9 rows ]</code></pre>
<p>We can now visualise the results using a volcano plot:</p>
<div class="sourceCode" id="cb395"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb395-1" title="1">res <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb395-2" title="2"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> logFC, <span class="dt">y =</span> <span class="op">-</span><span class="kw">log10</span>(adj.P.Val))) <span class="op">+</span></a>
<a class="sourceLine" id="cb395-3" title="3"><span class="st">    </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">colour =</span> TP)) <span class="op">+</span></a>
<a class="sourceLine" id="cb395-4" title="4"><span class="st">    </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb395-5" title="5"><span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="op">-</span><span class="kw">log10</span>(<span class="fl">0.05</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb395-6" title="6"><span class="st">    </span><span class="kw">scale_color_manual</span>(<span class="dt">values =</span> <span class="kw">c</span>(<span class="st">"black"</span>,<span class="st">"red"</span>))</a></code></pre></div>
<pre><code>## Warning: Removed 25 rows containing missing values (geom_point).</code></pre>
<div class="figure">
<span id="fig:vp"></span>
<p class="caption marginnote shownote">
Figure 5.15: Volcano plot highlighing spiked-in proteins in red.
</p>
<img src="R4MS_files/figure-html/vp-1.png" alt="Volcano plot highlighing spiked-in proteins in red." width="672">
</div>
<p>Using the pipeline described above, we would would identify a single
differentially expressed protein at an 5 percent FDR but miss out the
other 36 expected spike-in proteins.</p>
<p>We can assess our results in terms of true/false postitves/negatives:</p>
<ul>
<li>True positives: 1</li>
<li>False positives: 0</li>
<li>True negatives: 1330</li>
<li>False negatives: 32</li>
</ul>
</div>
</div>
<div id="summary-exercice-1" class="section level2">
<h2>
<span class="header-section-number">5.5</span> Summary exercice</h2>
<p>As shown below, it is possible to substantially improve these results
by aggregating features using a robust summarisation (available as
<code>MsCoreUtils::robustSummary()</code>), i.e robust regression with
M-estimation using Huber weights, as described in section 2.7 in
<span class="citation">(Sticker et al. <a href="#ref-Sticker:2019">2019</a>)</span>.</p>
<div class="figure">
<span id="fig:unnamed-chunk-67"></span>
<p class="caption marginnote shownote">
Figure 5.16: Aggregation using robust summarisation.
</p>
<img src="img/vp2.png" alt="Aggregation using robust summarisation." width="1048">
</div>
<ul>
<li>True positives: 21</li>
<li>False positives: 2</li>
<li>True negatives: 1340</li>
<li>False negatives: 12</li>
</ul>
<p>Repeat and adapt what we have seen here using, for example, the
<code>robustSummary()</code> function.</p>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-Christoforou:2016">
<p>Christoforou, Andy, Claire M Mulvey, Lisa M Breckels, Aikaterini Geladaki, Tracey Hurrell, Penelope C Hayward, Thomas Naake, et al. 2016. “A Draft Map of the Mouse Pluripotent Stem Cell Spatial Proteome.” <em>Nat Commun</em> 7: 8992. <a href="https://doi.org/10.1038/ncomms9992">https://doi.org/10.1038/ncomms9992</a>.</p>
</div>
<div id="ref-Cox:2008">
<p>Cox, J, and M Mann. 2008. “MaxQuant Enables High Peptide Identification Rates, Individualized P.p.b.-range Mass Accuracies and Proteome-Wide Protein Quantification.” <em>Nat Biotechnol</em> 26 (12): 1367–72. <a href="https://doi.org/10.1038/nbt.1511">https://doi.org/10.1038/nbt.1511</a>.</p>
</div>
<div id="ref-Lazar:2016">
<p>Lazar, C, L Gatto, M Ferro, C Bruley, and T Burger. 2016. “Accounting for the Multiple Natures of Missing Values in Label-Free Quantitative Proteomics Data Sets to Compare Imputation Strategies.” <em>J Proteome Res</em> 15 (4): 1116–25. <a href="https://doi.org/10.1021/acs.jproteome.5b00981">https://doi.org/10.1021/acs.jproteome.5b00981</a>.</p>
</div>
<div id="ref-SE">
<p>Morgan, Martin, Valerie Obenchain, Jim Hester, and Hervé Pagès. 2020. <em>SummarizedExperiment: SummarizedExperiment Container</em>. <a href="https://bioconductor.org/packages/SummarizedExperiment">https://bioconductor.org/packages/SummarizedExperiment</a>.</p>
</div>
<div id="ref-Paulovich:2010">
<p>Paulovich, Amanda G, Dean Billheimer, Amy-Joan L Ham, Lorenzo Vega-Montoto, Paul A Rudnick, David L Tabb, Pei Wang, et al. 2010. “Interlaboratory Study Characterizing a Yeast Performance Standard for Benchmarking LC-MS Platform Performance.” <em>Mol. Cell. Proteomics</em> 9 (2): 242–54.</p>
</div>
<div id="ref-MAE">
<p>Ramos, Marcel, Lucas Schiffer, Angela Re, Rimsha Azhar, Azfar Basunia, Carmen Rodriguez Cabrera, Tiffany Chan, et al. 2017. “Software for the Integration of Multi-Omics Experiments in Bioconductor.” <em>Cancer Research</em> 77(21); e39-42.</p>
</div>
<div id="ref-Sticker:2019">
<p>Sticker, Adriaan, Ludger Goeminne, Lennart Martens, and Lieven Clement. 2019. “Robust Summarization and Inference in Proteome-Wide Label-Free Quantification.” <em>bioRxiv</em>. <a href="https://doi.org/10.1101/668863">https://doi.org/10.1101/668863</a>.</p>
</div>
</div>
</body></html>

<p style="text-align: center;">
<a href="identification-data.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-si.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2021-03-23
</p>
</div>
</div>



</body>
</html>
