<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 4 Identification data | R for Mass Spectrometry" />
<meta property="og:type" content="book" />




<meta name="author" content="Laurent Gatto, Sebastian Gibb, Johannes Rainer" />

<meta name="date" content="2025-06-05" />


<meta name="description" content="Chapter 4 Identification data | R for Mass Spectrometry">

<title>Chapter 4 Identification data | R for Mass Spectrometry</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>
<script>
function copy_link(id) {
  var dummy = document.createElement('input'),
  text = window.location.href.split(/[?#]/)[0] + '#' + id;
  document.body.appendChild(dummy);
  dummy.value = text;
  dummy.select();
  document.execCommand('copy');
  document.body.removeChild(dummy);
  
  var tooltip = document.getElementById(id + '-tooltip');
  tooltip.innerHTML = 'Copied!';
}

function reset_tooltip(id) {
  var tooltip = document.getElementById(id);
  tooltip.innerHTML = 'Copy link';
}
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }

code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #204a87; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #204a87; font-weight: bold; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #ce5c00; font-weight: bold; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">R for Mass Spectrometry<p><p class="author">Laurent Gatto, Sebastian Gibb, Johannes Rainer</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html" id="toc-preamble"><span class="toc-section-number">1</span> Preamble</a>
<a href="sec-msintro.html" id="toc-sec-msintro"><span class="toc-section-number">2</span> Introduction</a>
<a href="sec-raw.html" id="toc-sec-raw"><span class="toc-section-number">3</span> Raw MS data</a>
<a id="active-page" href="sec-id.html" id="toc-sec-id"><span class="toc-section-number">4</span> Identification data</a><ul class="toc-sections">
<li class="toc"><a href="#NA"> Identification data.frame</a></li>
<li class="toc"><a href="#keeping-all-matches"> Keeping all matches</a></li>
<li class="toc"><a href="#filtering-data"> Filtering data</a></li>
<li class="toc"><a href="#adding-identification-data-to-raw-data"> Adding identification data to raw data</a></li>
<li class="toc"><a href="#an-identification-annotated-chromatogram"> An identification-annotated chromatogram</a></li>
<li class="toc"><a href="#visualising-peptide-spectrum-matches"> Visualising peptide-spectrum matches</a></li>
<li class="toc"><a href="#comparing-spectra"> Comparing spectra</a></li>
<li class="toc"><a href="#exploration-and-assessment-of-identifications-using-msnid"> Exploration and Assessment of Identifications using <code>MSnID</code></a></li>
</ul>
<a href="sec-quant.html" id="toc-sec-quant"><span class="toc-section-number">5</span> Quantitative data</a>
<a href="sec-anx.html" id="toc-sec-anx"><span class="toc-section-number">6</span> Annex</a>
<a href="sec-si.html" id="toc-sec-si"><span class="toc-section-number">7</span> Additional materials and session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="sec-id" class="section level1" number="4">
<h1>
<span class="header-section-number">Chapter 4</span> Identification data</h1>
<p>Peptide identification is performed using third-party software - there
is no package to run these searches directly in R. When using command
line search engines it possible to hard-code or automatically generate
the search command lines and run them from R using a <code>system()</code>
call. This allows to generate these reproducibly (especially useful if
many command lines need to be run) and to keep a record in the R
script of the exact command.</p>
<p>The example below illustrates this for 3 mzML files to be searched
using <code>MSGFplus</code>:</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="sec-id.html#cb97-1" aria-hidden="true" tabindex="-1"></a>(mzmls <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"file_"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="st">".mzML"</span>))</span></code></pre></div>
<pre><code>## [1] "file_1.mzML" "file_2.mzML" "file_3.mzML"</code></pre>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="sec-id.html#cb99-1" aria-hidden="true" tabindex="-1"></a>(mzids <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"mzML"</span>, <span class="st">"mzid"</span>, mzmls))</span></code></pre></div>
<pre><code>## [1] "file_1.mzid" "file_2.mzid" "file_3.mzid"</code></pre>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="sec-id.html#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">"java -jar /path/to/MSGFPlus.jar"</span>,</span>
<span id="cb101-2"><a href="sec-id.html#cb101-2" aria-hidden="true" tabindex="-1"></a>       <span class="st">" -s "</span>, mzmls,</span>
<span id="cb101-3"><a href="sec-id.html#cb101-3" aria-hidden="true" tabindex="-1"></a>       <span class="st">" -o "</span>, mzids,</span>
<span id="cb101-4"><a href="sec-id.html#cb101-4" aria-hidden="true" tabindex="-1"></a>       <span class="st">" -d uniprot.fas"</span>,</span>
<span id="cb101-5"><a href="sec-id.html#cb101-5" aria-hidden="true" tabindex="-1"></a>       <span class="st">" -t 20ppm"</span>,</span>
<span id="cb101-6"><a href="sec-id.html#cb101-6" aria-hidden="true" tabindex="-1"></a>       <span class="st">" -m 0"</span>,</span>
<span id="cb101-7"><a href="sec-id.html#cb101-7" aria-hidden="true" tabindex="-1"></a>       <span class="st">" int 1"</span>)</span></code></pre></div>
<pre><code>## [1] "java -jar /path/to/MSGFPlus.jar -s file_1.mzML -o file_1.mzid -d uniprot.fas -t 20ppm -m 0 int 1"
## [2] "java -jar /path/to/MSGFPlus.jar -s file_2.mzML -o file_2.mzid -d uniprot.fas -t 20ppm -m 0 int 1"
## [3] "java -jar /path/to/MSGFPlus.jar -s file_3.mzML -o file_3.mzid -d uniprot.fas -t 20ppm -m 0 int 1"</code></pre>
<div id="identification-data.frame" class="section level2" number="4.1">
<h2>
<span class="header-section-number">4.1</span> Identification data.frame<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('identification-data.frame')" onmouseout="reset_tooltip('identification-data.frame-tooltip')"><span class="tooltiptext" id="identification-data.frame-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Let’s use the identification from <code>msdata</code>:</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="sec-id.html#cb103-1" aria-hidden="true" tabindex="-1"></a>idf <span class="ot">&lt;-</span> MsDataHub<span class="sc">::</span><span class="fu">TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01.20141210.mzid</span>()</span></code></pre></div>
<pre><code>## see ?MsDataHub and browseVignettes('MsDataHub') for documentation</code></pre>
<pre><code>## loading from cache</code></pre>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="sec-id.html#cb106-1" aria-hidden="true" tabindex="-1"></a>idf</span></code></pre></div>
<pre><code>##                                                  EH7807 
## "/home/lgatto/.cache/R/ExperimentHub/27926fb546ae_7857"</code></pre>
<p>Alternatively, the same file is also available from the <code>msdata</code>
package with <code>idf &lt;- msdata::ident(full.names = TRUE)</code>.</p>
<p>The easiest way to read identification data in <code>mzIdentML</code> (often
abbreviated with <code>mzid</code>) into R is to read it with the <code>PSM()</code>
constructor function from the
<a href="https://rformassspectrometry.github.io/PSMatch/"><code>PSMatch</code></a>
package<label for="tufte-sn-5" class="margin-toggle sidenote-number">5</label><input type="checkbox" id="tufte-sn-5" class="margin-toggle"><span class="sidenote"><span class="sidenote-number">5</span> Previously named <code>PSM</code>.</span>. The function will parse the file and return a
<code>DataFrame</code>.</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="sec-id.html#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(PSMatch)</span>
<span id="cb108-2"><a href="sec-id.html#cb108-2" aria-hidden="true" tabindex="-1"></a>id <span class="ot">&lt;-</span> <span class="fu">PSM</span>(idf)</span>
<span id="cb108-3"><a href="sec-id.html#cb108-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(id)</span></code></pre></div>
<pre><code>## [1] 5802   35</code></pre>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="sec-id.html#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(id)</span></code></pre></div>
<pre><code>##  [1] "sequence"                 "spectrumID"              
##  [3] "chargeState"              "rank"                    
##  [5] "passThreshold"            "experimentalMassToCharge"
##  [7] "calculatedMassToCharge"   "peptideRef"              
##  [9] "modNum"                   "isDecoy"                 
## [11] "post"                     "pre"                     
## [13] "start"                    "end"                     
## [15] "DatabaseAccess"           "DBseqLength"             
## [17] "DatabaseSeq"              "DatabaseDescription"     
## [19] "scan.number.s."           "acquisitionNum"          
## [21] "spectrumFile"             "idFile"                  
## [23] "MS.GF.RawScore"           "MS.GF.DeNovoScore"       
## [25] "MS.GF.SpecEValue"         "MS.GF.EValue"            
## [27] "MS.GF.QValue"             "MS.GF.PepQValue"         
## [29] "modPeptideRef"            "modName"                 
## [31] "modMass"                  "modLocation"             
## [33] "subOriginalResidue"       "subReplacementResidue"   
## [35] "subLocation"</code></pre>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Verify that this table contains 5802 matches for 5343
scans and 4938 peptides sequences.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-9" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-9', 'sol-start-9')"></span>
</p>
<div id="sol-body-9" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="sec-id.html#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(id) <span class="do">## number of matches</span></span></code></pre></div>
<pre><code>## [1] 5802</code></pre>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="sec-id.html#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(id<span class="sc">$</span>spectrumID)) <span class="do">## number of scans</span></span></code></pre></div>
<pre><code>## [1] 5343</code></pre>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="sec-id.html#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(id<span class="sc">$</span>sequence))   <span class="do">## number of peptide sequences</span></span></code></pre></div>
<pre><code>## [1] 4938</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>The PSM data are read as is, without any filtering. As we can see
below, we still have all the hits from the forward and reverse (decoy)
databases.</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="sec-id.html#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(id<span class="sc">$</span>isDecoy)</span></code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##  2906  2896</code></pre>
</div>
<div id="keeping-all-matches" class="section level2" number="4.2">
<h2>
<span class="header-section-number">4.2</span> Keeping all matches<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('keeping-all-matches')" onmouseout="reset_tooltip('keeping-all-matches-tooltip')"><span class="tooltiptext" id="keeping-all-matches-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>The data contains also contains multiple matches for several
spectra. The table below shows the number of number of spectra that
have 1, 2, … up to 5 matches.</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="sec-id.html#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">table</span>(id<span class="sc">$</span>spectrumID))</span></code></pre></div>
<pre><code>## 
##    1    2    3    4    5 
## 4936  369   26   10    2</code></pre>
<p>Below, we can see how scan 1774 has 4 matches, all to sequence
<code>RTRYQAEVR</code>, which itself matches to 4 different proteins:</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="sec-id.html#cb122-1" aria-hidden="true" tabindex="-1"></a>i <span class="ot">&lt;-</span> <span class="fu">which</span>(id<span class="sc">$</span>spectrumID <span class="sc">==</span> <span class="st">"controllerType=0 controllerNumber=1 scan=1774"</span>)</span>
<span id="cb122-2"><a href="sec-id.html#cb122-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(id[i, ])[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##              sequence                                    spectrumID chargeState
## EH7807.1381 RTRYQAEVR controllerType=0 controllerNumber=1 scan=1774           2
## EH7807.1382 RTRYQAEVR controllerType=0 controllerNumber=1 scan=1774           2
## EH7807.1383 RTRYQAEVR controllerType=0 controllerNumber=1 scan=1774           2
## EH7807.2477 RTRYQAEVR controllerType=0 controllerNumber=1 scan=1774           2
##             rank passThreshold
## EH7807.1381    1          TRUE
## EH7807.1382    1          TRUE
## EH7807.1383    1          TRUE
## EH7807.2477    1          TRUE</code></pre>
<p>If the goal is to keep all the matches, but arranged by scan/spectrum,
one can <em>reduce</em> the <code>PSM</code> object by the <code>spectrumID</code> variable, so
that each scan correponds to a single row that still stores all
values<label for="tufte-sn-6" class="margin-toggle sidenote-number">6</label><input type="checkbox" id="tufte-sn-6" class="margin-toggle"><span class="sidenote"><span class="sidenote-number">6</span> The rownames aren’t needed here are are removed to reduce
to output in the the next code chunk display parts of <code>id2</code>.</span>:</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="sec-id.html#cb124-1" aria-hidden="true" tabindex="-1"></a>id2 <span class="ot">&lt;-</span> <span class="fu">reducePSMs</span>(id, id<span class="sc">$</span>spectrumID)</span>
<span id="cb124-2"><a href="sec-id.html#cb124-2" aria-hidden="true" tabindex="-1"></a>id2</span></code></pre></div>
<pre><code>## Reduced PSM with 5343 rows and 35 columns.
## names(35): sequence spectrumID ... subReplacementResidue subLocation</code></pre>
<p>The resulting object contains a single entry for scan 1774 with
information for the multiple matches stored as lists within the cells.</p>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb126-1"><a href="sec-id.html#cb126-1" aria-hidden="true" tabindex="-1"></a>j <span class="ot">&lt;-</span> <span class="fu">which</span>(id2<span class="sc">$</span>spectrumID <span class="sc">==</span> <span class="st">"controllerType=0 controllerNumber=1 scan=1774"</span>)</span>
<span id="cb126-2"><a href="sec-id.html#cb126-2" aria-hidden="true" tabindex="-1"></a>id2[j, ]</span></code></pre></div>
<pre><code>## Reduced PSM with 1 rows and 35 columns.
## names(35): sequence spectrumID ... subReplacementResidue subLocation</code></pre>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="sec-id.html#cb128-1" aria-hidden="true" tabindex="-1"></a>id2[j, <span class="st">"DatabaseAccess"</span>]</span></code></pre></div>
<pre><code>## CharacterList of length 1
## [["controllerType=0 controllerNumber=1 scan=1774"]] ECA2104 ECA2867 ECA3427 ECA4142</code></pre>
<p>The is the type of complete identification table that could be used to
annotate an raw mass spectrometry <code>Spectra</code> object, as shown below.</p>
</div>
<div id="filtering-data" class="section level2" number="4.3">
<h2>
<span class="header-section-number">4.3</span> Filtering data<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('filtering-data')" onmouseout="reset_tooltip('filtering-data-tooltip')"><span class="tooltiptext" id="filtering-data-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Often, the PSM data is filtered to only retain reliable matches. The
<code>MSnID</code> package can be used to set thresholds to attain user-defined
PSM, peptide or protein-level FDRs. Here, we will simply filter out
wrong identification manually.</p>
<p>Here, the <code>filter()</code> from the <code>dplyr</code> package comes very handy. We
will thus start by converting the <code>DataFrame</code> to a <code>tibble</code>.</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="sec-id.html#cb130-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"dplyr"</span>)</span>
<span id="cb130-2"><a href="sec-id.html#cb130-2" aria-hidden="true" tabindex="-1"></a>id_tbl <span class="ot">&lt;-</span> tidyr<span class="sc">::</span><span class="fu">as_tibble</span>(id)</span>
<span id="cb130-3"><a href="sec-id.html#cb130-3" aria-hidden="true" tabindex="-1"></a>id_tbl</span></code></pre></div>
<pre><code>## # A tibble: 5,802 × 35
##    sequence    spectrumID chargeState  rank passThreshold experimentalMassToCh…¹
##    &lt;chr&gt;       &lt;chr&gt;            &lt;int&gt; &lt;int&gt; &lt;lgl&gt;                          &lt;dbl&gt;
##  1 RQCRTDFLNY… controlle…           3     1 TRUE                            548.
##  2 ESVALADQVT… controlle…           2     1 TRUE                           1288.
##  3 KELLCLAMQI… controlle…           2     1 TRUE                            744.
##  4 QRMARTSDKQ… controlle…           3     1 TRUE                            913.
##  5 KDEGSTEPLK… controlle…           3     1 TRUE                            927.
##  6 DGGPAIYGHE… controlle…           3     1 TRUE                            969.
##  7 QRMARTSDKQ… controlle…           2     1 TRUE                           1369.
##  8 CIDRARHVEV… controlle…           3     1 TRUE                           1285.
##  9 CIDRARHVEV… controlle…           3     1 TRUE                           1285.
## 10 VGRCRPIINY… controlle…           2     1 TRUE                           1102.
## # ℹ 5,792 more rows
## # ℹ abbreviated name: ¹​experimentalMassToCharge
## # ℹ 29 more variables: calculatedMassToCharge &lt;dbl&gt;, peptideRef &lt;chr&gt;,
## #   modNum &lt;int&gt;, isDecoy &lt;lgl&gt;, post &lt;chr&gt;, pre &lt;chr&gt;, start &lt;int&gt;, end &lt;int&gt;,
## #   DatabaseAccess &lt;chr&gt;, DBseqLength &lt;int&gt;, DatabaseSeq &lt;chr&gt;,
## #   DatabaseDescription &lt;chr&gt;, scan.number.s. &lt;dbl&gt;, acquisitionNum &lt;dbl&gt;,
## #   spectrumFile &lt;chr&gt;, idFile &lt;chr&gt;, MS.GF.RawScore &lt;dbl&gt;, …</code></pre>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li>Remove decoy hits</li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-10" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-10', 'sol-start-10')"></span>
</p>
<div id="sol-body-10" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="sec-id.html#cb132-1" aria-hidden="true" tabindex="-1"></a>id_tbl <span class="ot">&lt;-</span> id_tbl <span class="sc">|&gt;</span></span>
<span id="cb132-2"><a href="sec-id.html#cb132-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span>isDecoy)</span>
<span id="cb132-3"><a href="sec-id.html#cb132-3" aria-hidden="true" tabindex="-1"></a>id_tbl</span></code></pre></div>
<pre><code>## # A tibble: 2,906 × 35
##    sequence    spectrumID chargeState  rank passThreshold experimentalMassToCh…¹
##    &lt;chr&gt;       &lt;chr&gt;            &lt;int&gt; &lt;int&gt; &lt;lgl&gt;                          &lt;dbl&gt;
##  1 RQCRTDFLNY… controlle…           3     1 TRUE                            548.
##  2 ESVALADQVT… controlle…           2     1 TRUE                           1288.
##  3 QRMARTSDKQ… controlle…           3     1 TRUE                            913.
##  4 DGGPAIYGHE… controlle…           3     1 TRUE                            969.
##  5 QRMARTSDKQ… controlle…           2     1 TRUE                           1369.
##  6 CIDRARHVEV… controlle…           3     1 TRUE                           1285.
##  7 CIDRARHVEV… controlle…           3     1 TRUE                           1285.
##  8 VGRCRPIINY… controlle…           2     1 TRUE                           1102.
##  9 QRLDEHCVGV… controlle…           3     1 TRUE                            713.
## 10 VDYQGKKVVI… controlle…           4     1 TRUE                            870.
## # ℹ 2,896 more rows
## # ℹ abbreviated name: ¹​experimentalMassToCharge
## # ℹ 29 more variables: calculatedMassToCharge &lt;dbl&gt;, peptideRef &lt;chr&gt;,
## #   modNum &lt;int&gt;, isDecoy &lt;lgl&gt;, post &lt;chr&gt;, pre &lt;chr&gt;, start &lt;int&gt;, end &lt;int&gt;,
## #   DatabaseAccess &lt;chr&gt;, DBseqLength &lt;int&gt;, DatabaseSeq &lt;chr&gt;,
## #   DatabaseDescription &lt;chr&gt;, scan.number.s. &lt;dbl&gt;, acquisitionNum &lt;dbl&gt;,
## #   spectrumFile &lt;chr&gt;, idFile &lt;chr&gt;, MS.GF.RawScore &lt;dbl&gt;, …</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li>Keep first rank matches</li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-11" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-11', 'sol-start-11')"></span>
</p>
<div id="sol-body-11" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="sec-id.html#cb134-1" aria-hidden="true" tabindex="-1"></a>id_tbl <span class="ot">&lt;-</span> id_tbl <span class="sc">|&gt;</span></span>
<span id="cb134-2"><a href="sec-id.html#cb134-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(rank <span class="sc">==</span> <span class="dv">1</span>)</span>
<span id="cb134-3"><a href="sec-id.html#cb134-3" aria-hidden="true" tabindex="-1"></a>id_tbl</span></code></pre></div>
<pre><code>## # A tibble: 2,751 × 35
##    sequence    spectrumID chargeState  rank passThreshold experimentalMassToCh…¹
##    &lt;chr&gt;       &lt;chr&gt;            &lt;int&gt; &lt;int&gt; &lt;lgl&gt;                          &lt;dbl&gt;
##  1 RQCRTDFLNY… controlle…           3     1 TRUE                            548.
##  2 ESVALADQVT… controlle…           2     1 TRUE                           1288.
##  3 QRMARTSDKQ… controlle…           3     1 TRUE                            913.
##  4 DGGPAIYGHE… controlle…           3     1 TRUE                            969.
##  5 QRMARTSDKQ… controlle…           2     1 TRUE                           1369.
##  6 CIDRARHVEV… controlle…           3     1 TRUE                           1285.
##  7 CIDRARHVEV… controlle…           3     1 TRUE                           1285.
##  8 VGRCRPIINY… controlle…           2     1 TRUE                           1102.
##  9 QRLDEHCVGV… controlle…           3     1 TRUE                            713.
## 10 VDYQGKKVVI… controlle…           4     1 TRUE                            870.
## # ℹ 2,741 more rows
## # ℹ abbreviated name: ¹​experimentalMassToCharge
## # ℹ 29 more variables: calculatedMassToCharge &lt;dbl&gt;, peptideRef &lt;chr&gt;,
## #   modNum &lt;int&gt;, isDecoy &lt;lgl&gt;, post &lt;chr&gt;, pre &lt;chr&gt;, start &lt;int&gt;, end &lt;int&gt;,
## #   DatabaseAccess &lt;chr&gt;, DBseqLength &lt;int&gt;, DatabaseSeq &lt;chr&gt;,
## #   DatabaseDescription &lt;chr&gt;, scan.number.s. &lt;dbl&gt;, acquisitionNum &lt;dbl&gt;,
## #   spectrumFile &lt;chr&gt;, idFile &lt;chr&gt;, MS.GF.RawScore &lt;dbl&gt;, …</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li>Remove shared peptides. Start by identifying scans that match
different proteins. For example scan 4884 matches proteins
<code>XXX_ECA3406</code> and <code>ECA3415</code>. Scan 4099 match <code>XXX_ECA4416_1</code>,
<code>XXX_ECA4416_2</code> and <code>XXX_ECA4416_3</code>. Then remove the scans that
match any of these proteins.</li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-12" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-12', 'sol-start-12')"></span>
</p>
<div id="sol-body-12" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb136-1"><a href="sec-id.html#cb136-1" aria-hidden="true" tabindex="-1"></a>mltm <span class="ot">&lt;-</span></span>
<span id="cb136-2"><a href="sec-id.html#cb136-2" aria-hidden="true" tabindex="-1"></a>    id_tbl <span class="sc">|&gt;</span></span>
<span id="cb136-3"><a href="sec-id.html#cb136-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(spectrumID) <span class="sc">|&gt;</span></span>
<span id="cb136-4"><a href="sec-id.html#cb136-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">nProts =</span> <span class="fu">length</span>(<span class="fu">unique</span>(DatabaseAccess))) <span class="sc">|&gt;</span></span>
<span id="cb136-5"><a href="sec-id.html#cb136-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(nProts <span class="sc">&gt;</span> <span class="dv">1</span>) <span class="sc">|&gt;</span></span>
<span id="cb136-6"><a href="sec-id.html#cb136-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(spectrumID, nProts)</span>
<span id="cb136-7"><a href="sec-id.html#cb136-7" aria-hidden="true" tabindex="-1"></a>mltm</span></code></pre></div>
<pre><code>## # A tibble: 85 × 2
## # Groups:   spectrumID [39]
##    spectrumID                                    nProts
##    &lt;chr&gt;                                          &lt;int&gt;
##  1 controllerType=0 controllerNumber=1 scan=1073      2
##  2 controllerType=0 controllerNumber=1 scan=1073      2
##  3 controllerType=0 controllerNumber=1 scan=6578      2
##  4 controllerType=0 controllerNumber=1 scan=6578      2
##  5 controllerType=0 controllerNumber=1 scan=5617      2
##  6 controllerType=0 controllerNumber=1 scan=5617      2
##  7 controllerType=0 controllerNumber=1 scan=3926      2
##  8 controllerType=0 controllerNumber=1 scan=3926      2
##  9 controllerType=0 controllerNumber=1 scan=4784      2
## 10 controllerType=0 controllerNumber=1 scan=4784      2
## # ℹ 75 more rows</code></pre>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb138-1"><a href="sec-id.html#cb138-1" aria-hidden="true" tabindex="-1"></a>id_tbl <span class="ot">&lt;-</span></span>
<span id="cb138-2"><a href="sec-id.html#cb138-2" aria-hidden="true" tabindex="-1"></a>    id_tbl <span class="sc">|&gt;</span></span>
<span id="cb138-3"><a href="sec-id.html#cb138-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span>spectrumID <span class="sc">%in%</span> mltm<span class="sc">$</span>spectrumID)</span>
<span id="cb138-4"><a href="sec-id.html#cb138-4" aria-hidden="true" tabindex="-1"></a>id_tbl</span></code></pre></div>
<pre><code>## # A tibble: 2,666 × 35
##    sequence    spectrumID chargeState  rank passThreshold experimentalMassToCh…¹
##    &lt;chr&gt;       &lt;chr&gt;            &lt;int&gt; &lt;int&gt; &lt;lgl&gt;                          &lt;dbl&gt;
##  1 RQCRTDFLNY… controlle…           3     1 TRUE                            548.
##  2 ESVALADQVT… controlle…           2     1 TRUE                           1288.
##  3 QRMARTSDKQ… controlle…           3     1 TRUE                            913.
##  4 DGGPAIYGHE… controlle…           3     1 TRUE                            969.
##  5 QRMARTSDKQ… controlle…           2     1 TRUE                           1369.
##  6 CIDRARHVEV… controlle…           3     1 TRUE                           1285.
##  7 CIDRARHVEV… controlle…           3     1 TRUE                           1285.
##  8 VGRCRPIINY… controlle…           2     1 TRUE                           1102.
##  9 QRLDEHCVGV… controlle…           3     1 TRUE                            713.
## 10 VDYQGKKVVI… controlle…           4     1 TRUE                            870.
## # ℹ 2,656 more rows
## # ℹ abbreviated name: ¹​experimentalMassToCharge
## # ℹ 29 more variables: calculatedMassToCharge &lt;dbl&gt;, peptideRef &lt;chr&gt;,
## #   modNum &lt;int&gt;, isDecoy &lt;lgl&gt;, post &lt;chr&gt;, pre &lt;chr&gt;, start &lt;int&gt;, end &lt;int&gt;,
## #   DatabaseAccess &lt;chr&gt;, DBseqLength &lt;int&gt;, DatabaseSeq &lt;chr&gt;,
## #   DatabaseDescription &lt;chr&gt;, scan.number.s. &lt;dbl&gt;, acquisitionNum &lt;dbl&gt;,
## #   spectrumFile &lt;chr&gt;, idFile &lt;chr&gt;, MS.GF.RawScore &lt;dbl&gt;, …</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>Which leaves us with 2666 PSMs.</p>
<p>This can also be achieved with the <code>filterPSMs()</code> function, or the
individual <code>filterPsmRank()</code>, <code>filterPsmDecoy</code> and <code>filterPsmShared()</code>
functions:</p>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb140-1"><a href="sec-id.html#cb140-1" aria-hidden="true" tabindex="-1"></a>id_filtered <span class="ot">&lt;-</span> <span class="fu">filterPSMs</span>(id)</span></code></pre></div>
<pre><code>## Starting with 5802 PSMs:</code></pre>
<pre><code>## Removed 2896 decoy hits.</code></pre>
<pre><code>## Removed 155 PSMs with rank &gt; 1.</code></pre>
<pre><code>## Removed 85 shared peptides.</code></pre>
<pre><code>## 2666 PSMs left.</code></pre>
<p>The <code>describePeptides()</code> and <code>describeProteins()</code> functions from the
<code>PSMatch</code> package provide useful summaries of preptides and proteins
in a PSM search result.</p>
<ul>
<li>
<code>describePeptides()</code> gives the number of unique and shared peptides
and for the latter, the size of their protein groups:</li>
</ul>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="sec-id.html#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="fu">describePeptides</span>(id_filtered)</span></code></pre></div>
<pre><code>## 2324 peptides composed of</code></pre>
<pre><code>##  unique peptides: 2324</code></pre>
<pre><code>##  shared peptides (among protein):</code></pre>
<pre><code>##   ()</code></pre>
<ul>
<li>
<code>describeProteins()</code> gives the number of proteins defined by only
unique, only shared, or a mixture of unique/shared peptides:</li>
</ul>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="sec-id.html#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="fu">describeProteins</span>(id_filtered)</span></code></pre></div>
<pre><code>## 1466 proteins composed of</code></pre>
<pre><code>##  only unique peptides: 1466</code></pre>
<pre><code>##  only shared peptides: 0</code></pre>
<pre><code>##  unique and shared peptides: 0</code></pre>
<p>The <a href="https://rformassspectrometry.github.io/PSMatch/articles/AdjacencyMatrix.html">Understanding protein groups with adjacency
matrices</a>
<code>PSMatch</code> vignette provides additional tools to explore how proteins
were inferred from peptides.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Compare the distribution of raw identification scores of the decoy and
non-decoy hits. Interpret the figure.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-13" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-13', 'sol-start-13')"></span>
</p>
<div id="sol-body-13" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb156"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb156-1"><a href="sec-id.html#cb156-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb156-2"><a href="sec-id.html#cb156-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(id, <span class="fu">aes</span>(<span class="at">x =</span> MS.GF.RawScore,</span>
<span id="cb156-3"><a href="sec-id.html#cb156-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">colour =</span> isDecoy)) <span class="sc">+</span></span>
<span id="cb156-4"><a href="sec-id.html#cb156-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>()</span></code></pre></div>
<p><img src="R4MS_files/figure-html/unnamed-chunk-43-1.png" width="672"></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>The <em><a href="https://CRAN.R-project.org/package=tidyverse">tidyverse</a></em>
tools are fit for data wrangling with identification data. Using the
above identification dataframe, calculate the length of each peptide
(you can use <code>nchar</code> with the peptide sequence <code>sequence</code>) and the
number of peptides for each protein (defined as
<code>DatabaseDescription</code>). Plot the length of the proteins against their
respective number of peptides.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-14" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-14', 'sol-start-14')"></span>
</p>
<div id="sol-body-14" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="sec-id.html#cb157-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(<span class="st">"dplyr"</span>))</span>
<span id="cb157-2"><a href="sec-id.html#cb157-2" aria-hidden="true" tabindex="-1"></a>iddf <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(id_filtered) <span class="sc">|&gt;</span></span>
<span id="cb157-3"><a href="sec-id.html#cb157-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">peplen =</span> <span class="fu">nchar</span>(sequence))</span>
<span id="cb157-4"><a href="sec-id.html#cb157-4" aria-hidden="true" tabindex="-1"></a>npeps <span class="ot">&lt;-</span> iddf <span class="sc">|&gt;</span></span>
<span id="cb157-5"><a href="sec-id.html#cb157-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(DatabaseAccess) <span class="sc">|&gt;</span></span>
<span id="cb157-6"><a href="sec-id.html#cb157-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tally</span>()</span>
<span id="cb157-7"><a href="sec-id.html#cb157-7" aria-hidden="true" tabindex="-1"></a>iddf <span class="ot">&lt;-</span> <span class="fu">full_join</span>(iddf, npeps)</span></code></pre></div>
<pre><code>## Joining with `by = join_by(DatabaseAccess)`</code></pre>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="sec-id.html#cb159-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb159-2"><a href="sec-id.html#cb159-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(iddf, <span class="fu">aes</span>(<span class="at">x =</span> n, <span class="at">y =</span> DBseqLength)) <span class="sc">+</span> <span class="fu">geom_point</span>()</span></code></pre></div>
<div class="figure">
<span style="display:block;" id="fig:answid1"></span>
<p class="caption marginnote shownote">
Figure 4.1: Identifcation data wrangling.
</p>
<img src="R4MS_files/figure-html/answid1-1.png" alt="Identifcation data wrangling." width="672">
</div>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>If you would like to learn more about how the mzid data are handled by
<code>PSMatch</code> via the <em><a href="https://bioconductor.org/packages/3.21/mzR">mzR</a></em> and <em><a href="https://bioconductor.org/packages/3.21/mzID">mzID</a></em>
packages, check out the <a href="sec-anx.html#sec-id2">6.2</a> section in the annex.</p>
</div>
<div id="adding-identification-data-to-raw-data" class="section level2" number="4.4">
<h2>
<span class="header-section-number">4.4</span> Adding identification data to raw data<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('adding-identification-data-to-raw-data')" onmouseout="reset_tooltip('adding-identification-data-to-raw-data-tooltip')"><span class="tooltiptext" id="adding-identification-data-to-raw-data-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>We are goind to use the <code>sp</code> object created in the previous chapter
and the <code>id_filtered</code> variable generated above.</p>
<p>Identification data (as a <code>DataFrame</code>) can be merged into raw data (as
a <code>Spectra</code> object) by adding new spectra variables to the appropriate
MS2 spectra. Scans and peptide-spectrum matches can be matched by
their spectrum identifers.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Identify the spectum identifier columns in the <code>sp</code> the <code>id_filtered</code>
variables.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-15" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-15', 'sol-start-15')"></span>
</p>
<div id="sol-body-15" class="solution-body" style="display: none;">
<p>In the raw data, it is encoded as <code>spectrumId</code>, while in the
identification data, we have <code>spectrumID</code>.</p>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="sec-id.html#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(id_filtered<span class="sc">$</span>spectrumID)</span></code></pre></div>
<pre><code>## [1] "controllerType=0 controllerNumber=1 scan=2949"
## [2] "controllerType=0 controllerNumber=1 scan=6534"
## [3] "controllerType=0 controllerNumber=1 scan=4782"
## [4] "controllerType=0 controllerNumber=1 scan=6118"
## [5] "controllerType=0 controllerNumber=1 scan=4757"
## [6] "controllerType=0 controllerNumber=1 scan=6518"</code></pre>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="sec-id.html#cb162-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sp<span class="sc">$</span>spectrumId)</span></code></pre></div>
<pre><code>## [1] "controllerType=0 controllerNumber=1 scan=1"
## [2] "controllerType=0 controllerNumber=1 scan=2"
## [3] "controllerType=0 controllerNumber=1 scan=3"
## [4] "controllerType=0 controllerNumber=1 scan=4"
## [5] "controllerType=0 controllerNumber=1 scan=5"
## [6] "controllerType=0 controllerNumber=1 scan=6"</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>We still have several PTMs that are matched to a single spectrum
identifier:</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="sec-id.html#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">table</span>(id_filtered<span class="sc">$</span>spectrumID))</span></code></pre></div>
<pre><code>## 
##    1    2    3    4 
## 2630   13    2    1</code></pre>
<p>Let’s look at <code>"controllerType=0 controllerNumber=1 scan=5490"</code>, the
has 4 matching PSMs in detail.</p>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="sec-id.html#cb166-1" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span>(<span class="fu">table</span>(id_filtered<span class="sc">$</span>spectrumID) <span class="sc">==</span> <span class="dv">4</span>)</span></code></pre></div>
<pre><code>## controllerType=0 controllerNumber=1 scan=5490 
##                                          1903</code></pre>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="sec-id.html#cb168-1" aria-hidden="true" tabindex="-1"></a>id_4 <span class="ot">&lt;-</span> id_filtered[id_filtered<span class="sc">$</span>spectrumID <span class="sc">==</span> <span class="st">"controllerType=0 controllerNumber=1 scan=5490"</span>, ] <span class="sc">|&gt;</span></span>
<span id="cb168-2"><a href="sec-id.html#cb168-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>()</span>
<span id="cb168-3"><a href="sec-id.html#cb168-3" aria-hidden="true" tabindex="-1"></a>id_4</span></code></pre></div>
<pre><code>##                   sequence                                    spectrumID
## EH7807.23 KCNQCLKVACTLFYCK controllerType=0 controllerNumber=1 scan=5490
## EH7807.24 KCNQCLKVACTLFYCK controllerType=0 controllerNumber=1 scan=5490
##           chargeState rank passThreshold experimentalMassToCharge
## EH7807.23           3    1          TRUE                 698.6633
## EH7807.24           3    1          TRUE                 698.6633
##           calculatedMassToCharge peptideRef modNum isDecoy post pre start end
## EH7807.23               698.3315     Pep453      4   FALSE    C   K   127 142
## EH7807.24               698.3315     Pep453      4   FALSE    C   K   127 142
##           DatabaseAccess DBseqLength DatabaseSeq          DatabaseDescription
## EH7807.23        ECA0668         302             ECA0668 hypothetical protein
## EH7807.24        ECA0668         302             ECA0668 hypothetical protein
##           scan.number.s. acquisitionNum
## EH7807.23           5490           5490
## EH7807.24           5490           5490
##                                                                  spectrumFile
## EH7807.23 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML
## EH7807.24 TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML
##                      idFile MS.GF.RawScore MS.GF.DeNovoScore MS.GF.SpecEValue
## EH7807.23 27926fb546ae_7857            -22                79     4.555588e-07
## EH7807.24 27926fb546ae_7857            -22                79     4.555588e-07
##           MS.GF.EValue MS.GF.QValue MS.GF.PepQValue modPeptideRef
## EH7807.23     1.307689    0.9006211       0.8901099        Pep453
## EH7807.24     1.307689    0.9006211       0.8901099        Pep453
##                   modName  modMass modLocation subOriginalResidue
## EH7807.23 Carbamidomethyl 57.02146           2               &lt;NA&gt;
## EH7807.24 Carbamidomethyl 57.02146           5               &lt;NA&gt;
##           subReplacementResidue subLocation
## EH7807.23                  &lt;NA&gt;          NA
## EH7807.24                  &lt;NA&gt;          NA
##  [ reached 'max' / getOption("max.print") -- omitted 2 rows ]</code></pre>
<p>We can see that these 4 PSMs differ by the location of the
Carbamidomethyl modification.</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="sec-id.html#cb170-1" aria-hidden="true" tabindex="-1"></a>id_4[, <span class="fu">c</span>(<span class="st">"modName"</span>, <span class="st">"modLocation"</span>)]</span></code></pre></div>
<pre><code>##                   modName modLocation
## EH7807.23 Carbamidomethyl           2
## EH7807.24 Carbamidomethyl           5
## EH7807.25 Carbamidomethyl          10
## EH7807.26 Carbamidomethyl          15</code></pre>
<p>Let’s reduce that PSM table before joining it to the <code>Spectra</code> object,
to make sure we have unique one-to-one matches between the raw spectra
and the PSMs.</p>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="sec-id.html#cb172-1" aria-hidden="true" tabindex="-1"></a>id_filtered <span class="ot">&lt;-</span> <span class="fu">reducePSMs</span>(id_filtered, id_filtered<span class="sc">$</span>spectrumID)</span>
<span id="cb172-2"><a href="sec-id.html#cb172-2" aria-hidden="true" tabindex="-1"></a>id_filtered</span></code></pre></div>
<pre><code>## Reduced PSM with 2646 rows and 35 columns.
## names(35): sequence spectrumID ... subReplacementResidue subLocation</code></pre>
<p>These two data can thus simply be joined using:</p>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="sec-id.html#cb174-1" aria-hidden="true" tabindex="-1"></a>sp <span class="ot">&lt;-</span> <span class="fu">joinSpectraData</span>(sp, id_filtered,</span>
<span id="cb174-2"><a href="sec-id.html#cb174-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">by.x =</span> <span class="st">"spectrumId"</span>,</span>
<span id="cb174-3"><a href="sec-id.html#cb174-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">by.y =</span> <span class="st">"spectrumID"</span>)</span>
<span id="cb174-4"><a href="sec-id.html#cb174-4" aria-hidden="true" tabindex="-1"></a><span class="fu">spectraVariables</span>(sp)</span></code></pre></div>
<pre><code>##  [1] "msLevel"                  "rtime"                   
##  [3] "acquisitionNum"           "scanIndex"               
##  [5] "dataStorage"              "dataOrigin"              
##  [7] "centroided"               "smoothed"                
##  [9] "polarity"                 "precScanNum"             
## [11] "precursorMz"              "precursorIntensity"      
## [13] "precursorCharge"          "collisionEnergy"         
## [15] "isolationWindowLowerMz"   "isolationWindowTargetMz" 
## [17] "isolationWindowUpperMz"   "peaksCount"              
## [19] "totIonCurrent"            "basePeakMZ"              
## [21] "basePeakIntensity"        "ionisationEnergy"        
## [23] "lowMZ"                    "highMZ"                  
## [25] "mergedScan"               "mergedResultScanNum"     
## [27] "mergedResultStartScanNum" "mergedResultEndScanNum"  
## [29] "injectionTime"            "filterString"            
## [31] "spectrumId"               "ionMobilityDriftTime"    
## [33] "scanWindowLowerLimit"     "scanWindowUpperLimit"    
## [35] "electronBeamEnergy"       "rtime_minute"            
## [37] "sequence"                 "chargeState"             
## [39] "rank"                     "passThreshold"           
## [41] "experimentalMassToCharge" "calculatedMassToCharge"  
## [43] "peptideRef"               "modNum"                  
## [45] "isDecoy"                  "post"                    
## [47] "pre"                      "start"                   
## [49] "end"                      "DatabaseAccess"          
## [51] "DBseqLength"              "DatabaseSeq"             
## [53] "DatabaseDescription"      "scan.number.s."          
## [55] "acquisitionNum.y"         "spectrumFile"            
## [57] "idFile"                   "MS.GF.RawScore"          
## [59] "MS.GF.DeNovoScore"        "MS.GF.SpecEValue"        
## [61] "MS.GF.EValue"             "MS.GF.QValue"            
## [63] "MS.GF.PepQValue"          "modPeptideRef"           
## [65] "modName"                  "modMass"                 
## [67] "modLocation"              "subOriginalResidue"      
## [69] "subReplacementResidue"    "subLocation"</code></pre>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Verify that the identification data has been added to the correct
spectra.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-16" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-16', 'sol-start-16')"></span>
</p>
<div id="sol-body-16" class="solution-body" style="display: none;">
<p>Let’s first verify that no identification data has been added to the
MS1 scans.</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="sec-id.html#cb176-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(<span class="fu">is.na</span>(<span class="fu">filterMsLevel</span>(sp, <span class="dv">1</span>)<span class="sc">$</span>sequence))</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>They have indeed been added to 56% of the MS2 spectra.</p>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="sec-id.html#cb178-1" aria-hidden="true" tabindex="-1"></a>sp_2 <span class="ot">&lt;-</span> <span class="fu">filterMsLevel</span>(sp, <span class="dv">2</span>)</span>
<span id="cb178-2"><a href="sec-id.html#cb178-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">is.na</span>(sp_2<span class="sc">$</span>sequence))</span></code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##  2646  3457</code></pre>
<p>Let’s compare the precursor/peptide mass to charges</p>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="sec-id.html#cb180-1" aria-hidden="true" tabindex="-1"></a>sp_2 <span class="ot">&lt;-</span> sp_2[<span class="sc">!</span><span class="fu">is.na</span>(sp_2<span class="sc">$</span>sequence)]</span>
<span id="cb180-2"><a href="sec-id.html#cb180-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(sp_2<span class="sc">$</span>precursorMz <span class="sc">-</span> sp_2<span class="sc">$</span>experimentalMassToCharge)</span></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##  0.0000  0.0000  0.0000  0.0053  0.0000  2.0297</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="an-identification-annotated-chromatogram" class="section level2" number="4.5">
<h2>
<span class="header-section-number">4.5</span> An identification-annotated chromatogram<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('an-identification-annotated-chromatogram')" onmouseout="reset_tooltip('an-identification-annotated-chromatogram-tooltip')"><span class="tooltiptext" id="an-identification-annotated-chromatogram-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Now that we have combined raw data and their associated
peptide-spectrum matches, we can produce an improved total ion
chromatogram, identifying MS1 scans that lead to successful
identifications.</p>
<p>The <code>countIdentifications()</code> function is going to tally the number of
identifications (i.e non-missing characters in the <code>sequence</code> spectra
variable) for each scan. In the case of MS2 scans, these will be
either 1 or 0, depending on the presence of a sequence. For MS1 scans,
the function will count the number of sequences for the descendant MS2
scans, i.e. those produced from precursor ions from each MS1 scan.</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="sec-id.html#cb182-1" aria-hidden="true" tabindex="-1"></a>sp <span class="ot">&lt;-</span> <span class="fu">countIdentifications</span>(sp)</span></code></pre></div>
<p>Below, we see on the second line that 3457 MS2 scans lead to no PSM,
while 2546 lead to an identification. Among all MS1 scans, 833 lead to
no MS2 scans with PSMs. 30 MS1 scans generated one MS2 scan that lead
to a PSM, 45 lead to two PSMs, …</p>
<div class="sourceCode" id="cb183"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb183-1"><a href="sec-id.html#cb183-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">msLevel</span>(sp), sp<span class="sc">$</span>countIdentifications)</span></code></pre></div>
<pre><code>##    
##        0    1    2    3    4    5    6    7    8    9   10
##   1  833   30   45   97  139  132   92   42   17    3    1
##   2 3457 2646    0    0    0    0    0    0    0    0    0</code></pre>
<p>These data can also be visualised on the total ion chromatogram:</p>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb185-1"><a href="sec-id.html#cb185-1" aria-hidden="true" tabindex="-1"></a>sp <span class="sc">|&gt;</span></span>
<span id="cb185-2"><a href="sec-id.html#cb185-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filterMsLevel</span>(<span class="dv">1</span>) <span class="sc">|&gt;</span></span>
<span id="cb185-3"><a href="sec-id.html#cb185-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">spectraData</span>() <span class="sc">|&gt;</span></span>
<span id="cb185-4"><a href="sec-id.html#cb185-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as_tibble</span>() <span class="sc">|&gt;</span></span>
<span id="cb185-5"><a href="sec-id.html#cb185-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> rtime,</span>
<span id="cb185-6"><a href="sec-id.html#cb185-6" aria-hidden="true" tabindex="-1"></a>               <span class="at">y =</span> totIonCurrent)) <span class="sc">+</span></span>
<span id="cb185-7"><a href="sec-id.html#cb185-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>(<span class="at">alpha =</span> <span class="fl">0.25</span>) <span class="sc">+</span></span>
<span id="cb185-8"><a href="sec-id.html#cb185-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">colour =</span> <span class="fu">ifelse</span>(countIdentifications <span class="sc">==</span> <span class="dv">0</span>,</span>
<span id="cb185-9"><a href="sec-id.html#cb185-9" aria-hidden="true" tabindex="-1"></a>                                   <span class="cn">NA</span>, countIdentifications)),</span>
<span id="cb185-10"><a href="sec-id.html#cb185-10" aria-hidden="true" tabindex="-1"></a>               <span class="at">size =</span> <span class="fl">0.75</span>,</span>
<span id="cb185-11"><a href="sec-id.html#cb185-11" aria-hidden="true" tabindex="-1"></a>               <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb185-12"><a href="sec-id.html#cb185-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">colour =</span> <span class="st">"Number of ids"</span>)</span></code></pre></div>
<div class="figure fullwidth">
<img src="R4MS_files/figure-html/nSequencePlot-1.png" alt=" " width="768"><p class="caption marginnote shownote">
</p>
</div>
</div>
<div id="visualising-peptide-spectrum-matches" class="section level2" number="4.6">
<h2>
<span class="header-section-number">4.6</span> Visualising peptide-spectrum matches<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('visualising-peptide-spectrum-matches')" onmouseout="reset_tooltip('visualising-peptide-spectrum-matches-tooltip')"><span class="tooltiptext" id="visualising-peptide-spectrum-matches-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Let’s choose a MS2 spectrum with a high identification score and plot
it.</p>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="sec-id.html#cb186-1" aria-hidden="true" tabindex="-1"></a>i <span class="ot">&lt;-</span> <span class="fu">which</span>(sp<span class="sc">$</span>MS.GF.RawScore <span class="sc">&gt;</span> <span class="dv">100</span>)[<span class="dv">1</span>]</span>
<span id="cb186-2"><a href="sec-id.html#cb186-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSpectra</span>(sp[i])</span></code></pre></div>
<p><img src="R4MS_files/figure-html/unnamed-chunk-53-1.png" width="672"></p>
<p>We have seen above that we can add labels to each peak using the
<code>labels</code> argument in <code>plotSpectra()</code>. The <code>addFragments()</code> function
takes a spectrum as input (that is a <code>Spectra</code> object of length 1) and
annotates its peaks.</p>
<div class="sourceCode" id="cb187"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb187-1"><a href="sec-id.html#cb187-1" aria-hidden="true" tabindex="-1"></a><span class="fu">addFragments</span>(sp[i])</span></code></pre></div>
<pre><code>## Warning in addFragments(sp[i]): 'addFragments' is deprecated.
## Use 'labelFragments' instead.
## See help("Deprecated")</code></pre>
<pre><code>## $THSQEEMQHMQR
##   [1] NA    NA    NA    "b1"  NA    NA    NA    NA    NA    NA    NA    NA   
##  [13] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   
##  [25] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   
##  [37] NA    NA    NA    NA    NA    NA    NA    "y1_" NA    NA    NA    NA   
##  [49] NA    "y1"  NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   
##  [61] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   
##  [73] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   
##  [85] NA    NA    "b2"  NA    NA    NA    NA    NA    NA    NA    NA    NA   
##  [97] NA    NA    NA    NA   
##  [ reached 'max' / getOption("max.print") -- omitted 227 entries ]
## 
## attr(,"group")
## [1] 1</code></pre>
<p>It can be directly used with <code>plotSpectra()</code>:</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="sec-id.html#cb190-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSpectra</span>(sp[i], <span class="at">labels =</span> addFragments,</span>
<span id="cb190-2"><a href="sec-id.html#cb190-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">labelPos =</span> <span class="dv">3</span>, <span class="at">labelCol =</span> <span class="st">"steelblue"</span>)</span></code></pre></div>
<pre><code>## Warning in labels(x): 'labels' is deprecated.
## Use 'labelFragments' instead.
## See help("Deprecated")</code></pre>
<p><img src="R4MS_files/figure-html/unnamed-chunk-55-1.png" width="672"></p>
<p>When a precursor peptide ion is fragmented in a CID cell, it breaks at
specific bonds, producing sets of peaks (<em>a</em>, <em>b</em>, <em>c</em> and <em>x</em>, <em>y</em>,
<em>z</em>) that can be predicted.</p>
<div class="figure">
<p class="caption marginnote shownote">
(#fig:frag_img)Peptide fragmentation.
</p>
<img src="img/frag.png" alt="Peptide fragmentation." width="80%">
</div>
<p>The annotation of spectra is obtained by simulating fragmentation of a
peptide and matching observed peaks to fragments:</p>
<div class="sourceCode" id="cb192"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb192-1"><a href="sec-id.html#cb192-1" aria-hidden="true" tabindex="-1"></a>sp[i]<span class="sc">$</span>sequence</span></code></pre></div>
<pre><code>## [1] "THSQEEMQHMQR"</code></pre>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb194-1"><a href="sec-id.html#cb194-1" aria-hidden="true" tabindex="-1"></a><span class="fu">calculateFragments</span>(sp[i]<span class="sc">$</span>sequence)</span></code></pre></div>
<pre><code>## Fixed modifications used: C=57.02146
## Variable modifications used: None</code></pre>
<pre><code>##           mz ion type pos z         seq      peptide
## 1   102.0550  b1    b   1 1           T THSQEEMQHMQR
## 2   239.1139  b2    b   2 1          TH THSQEEMQHMQR
## 3   326.1459  b3    b   3 1         THS THSQEEMQHMQR
## 4   454.2045  b4    b   4 1        THSQ THSQEEMQHMQR
## 5   583.2471  b5    b   5 1       THSQE THSQEEMQHMQR
## 6   712.2897  b6    b   6 1      THSQEE THSQEEMQHMQR
## 7   843.3301  b7    b   7 1     THSQEEM THSQEEMQHMQR
## 8   971.3887  b8    b   8 1    THSQEEMQ THSQEEMQHMQR
## 9  1108.4476  b9    b   9 1   THSQEEMQH THSQEEMQHMQR
## 10 1239.4881 b10    b  10 1  THSQEEMQHM THSQEEMQHMQR
## 11 1367.5467 b11    b  11 1 THSQEEMQHMQ THSQEEMQHMQR
## 12  175.1190  y1    y   1 1           R THSQEEMQHMQR
## 13  303.1775  y2    y   2 1          QR THSQEEMQHMQR
## 14  434.2180  y3    y   3 1         MQR THSQEEMQHMQR
##  [ reached 'max' / getOption("max.print") -- omitted 44 rows ]</code></pre>
</div>
<div id="comparing-spectra" class="section level2" number="4.7">
<h2>
<span class="header-section-number">4.7</span> Comparing spectra<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('comparing-spectra')" onmouseout="reset_tooltip('comparing-spectra-tooltip')"><span class="tooltiptext" id="comparing-spectra-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>The <code>compareSpectra()</code> function can be used to compare spectra (by default,
computing the normalised dot product).</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ol style="list-style-type: decimal">
<li>Create a new <code>Spectra</code> object containing the MS2 spectra with
sequences <code>"SQILQQAGTSVLSQANQVPQTVLSLLR"</code> and
<code>"TKGLNVMQNLLTAHPDVQAVFAQNDEMALGALR"</code>.</li>
</ol>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-17" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-17', 'sol-start-17')"></span>
</p>
<div id="sol-body-17" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb197-1"><a href="sec-id.html#cb197-1" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="fu">which</span>(sp<span class="sc">$</span>sequence <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"SQILQQAGTSVLSQANQVPQTVLSLLR"</span>, <span class="st">"TKGLNVMQNLLTAHPDVQAVFAQNDEMALGALR"</span>))</span>
<span id="cb197-2"><a href="sec-id.html#cb197-2" aria-hidden="true" tabindex="-1"></a>sp_k <span class="ot">&lt;-</span> sp[k]</span>
<span id="cb197-3"><a href="sec-id.html#cb197-3" aria-hidden="true" tabindex="-1"></a>sp_k</span></code></pre></div>
<pre><code>## MSn data (Spectra) with 5 spectra in a MsBackendMzR backend:
##     msLevel     rtime scanIndex
##   &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;
## 1         2   2687.42      5230
## 2         2   2688.88      5235
## 3         2   2748.75      5397
## 4         2   2765.26      5442
## 5         2   2768.17      5449
##  ... 70 more variables/columns.
## 
## file(s):
## 184155b6efdb_TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ol start="2" style="list-style-type: decimal">
<li>Calculate the 5 by 5 similarity
matrix between all spectra using <code>compareSpectra</code>. See the
<code>?Spectra</code> man page for details. Draw a heatmap of that matrix.</li>
</ol>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-18" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-18', 'sol-start-18')"></span>
</p>
<div id="sol-body-18" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb199-1"><a href="sec-id.html#cb199-1" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> <span class="fu">compareSpectra</span>(sp_k)</span>
<span id="cb199-2"><a href="sec-id.html#cb199-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(mat) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(mat) <span class="ot">&lt;-</span> <span class="fu">strtrim</span>(sp_k<span class="sc">$</span>sequence, <span class="dv">2</span>)</span>
<span id="cb199-3"><a href="sec-id.html#cb199-3" aria-hidden="true" tabindex="-1"></a>mat</span></code></pre></div>
<pre><code>##              TK          TK           SQ          SQ           SQ
## TK 1.0000000000 0.109126094 0.0009373465 0.001261338 0.0008256185
## TK 0.1091260942 1.000000000 0.0025314670 0.001459654 0.0017613212
## SQ 0.0009373465 0.002531467 1.0000000000 0.432133016 0.6879331218
## SQ 0.0012613380 0.001459654 0.4321330158 1.000000000 0.4467153012
## SQ 0.0008256185 0.001761321 0.6879331218 0.446715301 1.0000000000</code></pre>
<div class="sourceCode" id="cb201"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb201-1"><a href="sec-id.html#cb201-1" aria-hidden="true" tabindex="-1"></a>pheatmap<span class="sc">::</span><span class="fu">pheatmap</span>(mat)</span></code></pre></div>
<p><img src="R4MS_files/figure-html/spectra_comp_heatmap-1.png" width="672"></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ol start="3" style="list-style-type: decimal">
<li>Compare the spectra with the plotting function seen previously.</li>
</ol>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-19" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-19', 'sol-start-19')"></span>
</p>
<div id="sol-body-19" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb202-1"><a href="sec-id.html#cb202-1" aria-hidden="true" tabindex="-1"></a><span class="fu">filterIntensity</span>(sp_k, <span class="fl">1e3</span>) <span class="sc">|&gt;</span></span>
<span id="cb202-2"><a href="sec-id.html#cb202-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotSpectra</span>(<span class="at">main =</span> sp_k<span class="sc">$</span>sequence)</span></code></pre></div>
<p><img src="R4MS_files/figure-html/plot_spectra_compared-1.png" width="672"></p>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb203-1"><a href="sec-id.html#cb203-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">1</span>))</span>
<span id="cb203-2"><a href="sec-id.html#cb203-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSpectraMirror</span>(sp_k[<span class="dv">1</span>], sp_k[<span class="dv">2</span>], <span class="at">main =</span> <span class="st">"TK..."</span>)</span>
<span id="cb203-3"><a href="sec-id.html#cb203-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSpectraMirror</span>(sp_k[<span class="dv">3</span>], sp_k[<span class="dv">4</span>], <span class="at">main =</span> <span class="st">"SQ..."</span>)</span>
<span id="cb203-4"><a href="sec-id.html#cb203-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSpectraMirror</span>(sp_k[<span class="dv">3</span>], sp_k[<span class="dv">4</span>], <span class="at">main =</span> <span class="st">"SQ..."</span>)</span></code></pre></div>
<p><img src="R4MS_files/figure-html/unnamed-chunk-56-1.png" width="672"></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<!-- ## Summary exercise -->
<!-- 

<div class = 'question' id=''><p class='question-begin'>&#x25BA; Question</p><div class='question-body'>

 -->
<!-- Download the 3 first mzML and mzID files from the -->
<!-- [PXD022816](https://www.ebi.ac.uk/pride/archive/projects/PXD022816) -->
<!-- project [@Morgenstern:2020]. -->
<!-- 

<p class="question-end"><span class="fa fa-square-o solution-icon"></span></p></div></div>

 -->
<!-- 

<div class = "msmb-solution"><p class="solution-begin">&#x25BA; Solution<span id='sol-start-20' class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-20', 'sol-start-20')"></span></p><div class="solution-body" id = "sol-body-20" style="display: none;">

 -->
<!-- ```{r get_PXD022816} -->
<!-- ## Getting data from PX/PRIDE -->
<!-- library(rpx) -->
<!-- ## https://www.ebi.ac.uk/pride/archive/projects/PXD022816 -->
<!-- ## RawBeans: A Simple, Vendor-Independent, Raw-Data Quality-Control -->
<!-- ## Tool () -->
<!-- PXD022816 <- PXDataset("PXD022816") -->
<!-- PXD022816 -->
<!-- pxfiles(PXD022816) -->
<!-- (mzids <- pxget(PXD022816, grep("mzID", pxfiles(PXD022816))[1:3])) -->
<!-- (mzmls <- pxget(PXD022816, grep("mzML", pxfiles(PXD022816))[1:3])) -->
<!-- ``` -->
<!-- 

<p class="solution-end"><span class="fa fa-square-o solution-icon"></span></p></div></div>

 -->
<!-- 

<div class = 'question' id=''><p class='question-begin'>&#x25BA; Question</p><div class='question-body'>

 -->
<!-- Generate a `Spectra` object and a table of filtered PSMs. Visualise -->
<!-- the total ion chromatograms and check the quality of the -->
<!-- identification data by comparing the density of the decoy and target -->
<!-- PSMs id scores for each file. -->
<!-- 

<p class="question-end"><span class="fa fa-square-o solution-icon"></span></p></div></div>

 -->
<!-- 

<div class = "msmb-solution"><p class="solution-begin">&#x25BA; Solution<span id='sol-start-21' class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-21', 'sol-start-21')"></span></p><div class="solution-body" id = "sol-body-21" style="display: none;">

 -->
<!-- ```{r spectra_PXD022816, message = FALSE} -->
<!-- ## Loading raw data -->
<!-- library("Spectra") -->
<!-- sp <- Spectra(mzmls) -->
<!-- sp -->
<!-- ## number of spectra per file -->
<!-- table(basename(sp$dataOrigin)) -->
<!-- ## all levels are centroided -->
<!-- table(sp$centroided, sp$msLevel) -->
<!-- ``` -->
<!-- ```{r mslevel_PXD022816, message = FALSE, fig.width = 12} -->
<!-- library("ggplot2") -->
<!-- library("tidyr") -->
<!-- library("magrittr") -->
<!-- ## Chromatograms -->
<!-- filterMsLevel(sp, 1) |> -->
<!--     spectraData() |> -->
<!--     as_tibble() |> -->
<!--     ggplot(aes(x = rtime, -->
<!--                y = totIonCurrent, -->
<!--                colour = basename(dataOrigin))) + -->
<!--     geom_line() -->
<!-- ``` -->
<!-- ```{r is_PXD022816, message = FALSE, fig.width = 12} -->
<!-- ## Identification data -->
<!-- library("PSMatch") -->
<!-- id <- PSM(mzids) -->
<!-- ## Number of PSMs per acquisition -->
<!-- table(id$idFile) -->
<!-- tidyr::as_tibble(id) |> -->
<!--     ggplot(aes(x = MetaMorpheus.score, -->
<!--                colour = isDecoy)) + -->
<!--     geom_density() + -->
<!--     facet_wrap(~ spectrumFile) -->
<!-- ``` -->
<!-- ```{r idfilter_PXD022816} -->
<!-- id_filtered <- filterPSMs(id) -->
<!-- max(id_filtered$PSM.level.q.value) -->
<!-- ``` -->
<!-- 

<p class="solution-end"><span class="fa fa-square-o solution-icon"></span></p></div></div>

 -->
<!-- 

<div class = 'question' id=''><p class='question-begin'>&#x25BA; Question</p><div class='question-body'>

 -->
<!-- Join the raw and identification data. Beware though that the joining -->
<!-- must now be performed by spectrum ids and by files. -->
<!-- 

<p class="question-end"><span class="fa fa-square-o solution-icon"></span></p></div></div>

 -->
<!-- 

<div class = "msmb-solution"><p class="solution-begin">&#x25BA; Solution<span id='sol-start-22' class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-22', 'sol-start-22')"></span></p><div class="solution-body" id = "sol-body-22" style="display: none;">

 -->
<!-- ```{r joinsp_PXD022816} -->
<!-- ## primary key for spectra -->
<!-- sp$pkey <- -->
<!--     paste0(sub("^.+_QEP", "QEP", basename(dataOrigin(sp))), -->
<!--            gsub("^.+=", "::", sp$spectrumId)) -->
<!-- head(sp$pkey) -->
<!-- ## primary key for PSMs -->
<!-- id_filtered$pkey <- -->
<!--     paste0(gsub("^.+\\QEP", "QEP", id_filtered$spectrumFile), -->
<!--            sub("^.+=", "::", id_filtered$spectrumID)) -->
<!-- head(id_filtered$pkey) -->
<!-- ## For simplicity, let's keep single hits per spectrum id. -->
<!-- ## Alternatively, explore duplicates and use QFeatures::reduceDataFrame -->
<!-- id_filtered <- id_filtered[!duplicated(id_filtered$pkey), ] -->
<!-- head(id_filtered$pkey) -->
<!-- sp <- joinSpectraData(sp, id_filtered, by.x = "pkey") -->
<!-- ## Number of MS2 scans with a PSM -->
<!-- table(!is.na(filterMsLevel(sp, 2)$sequence)) -->
<!-- ``` -->
<!-- 

<p class="solution-end"><span class="fa fa-square-o solution-icon"></span></p></div></div>

 -->
<!-- 

<div class = 'question' id=''><p class='question-begin'>&#x25BA; Question</p><div class='question-body'>

 -->
<!-- Extract the PSMs that have been matched to peptides from protein -->
<!-- `O43175` and compare and cluster the scans. Hint: once you have -->
<!-- created the smaller `Spectra` object with the scans of interest, -->
<!-- switch to an in-memory backend to seed up the calculations. -->
<!-- 

<p class="question-end"><span class="fa fa-square-o solution-icon"></span></p></div></div>

 -->
<!-- 

<div class = "msmb-solution"><p class="solution-begin">&#x25BA; Solution<span id='sol-start-23' class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-23', 'sol-start-23')"></span></p><div class="solution-body" id = "sol-body-23" style="display: none;">

 -->
<!-- ```{r compare_PXD022816} -->
<!-- sp_O43175 <- sp[which(sp$DatabaseAccess == "O43175")] -->
<!-- sp_O43175 <- setBackend(sp_O43175, MsBackendDataFrame()) -->
<!-- sp_O43175 -->
<!-- cmat <- compareSpectra(sp_O43175) -->
<!-- rownames(cmat) <- -->
<!--     colnames(cmat) <- strtrim(sp_O43175$sequence, 3) -->
<!-- pheatmap::pheatmap(cmat) -->
<!-- (i <- which(rownames(cmat) == "DLP")) -->
<!-- plotSpectra(sp_O43175[i], labels = addFragments, -->
<!--             labelPos = 3, labelCol = "steelblue", -->
<!--             main = sp_O43175$sequence[i]) -->
<!-- spectraData(sp_O43175[i])$precursorCharge -->
<!-- spectraData(sp_O43175[i])$precursorMz -->
<!-- spectraData(sp_O43175[i])$modName -->
<!-- ``` -->
<!-- ```{r plotmirror_PXD022816} -->
<!-- ## Directly compare spectra with/without Carbamyl -->
<!-- plotSpectraMirror(sp_O43175[4], sp_O43175[9]) -->
<!-- plotSpectraMirror(sp_O43175[2], sp_O43175[10]) -->
<!-- ``` -->
<!-- 

<p class="solution-end"><span class="fa fa-square-o solution-icon"></span></p></div></div>

 -->
<!-- 

<div class = 'question' id=''><p class='question-begin'>&#x25BA; Question</p><div class='question-body'>

 -->
<!-- Generate total ion chromatograms for each acquisition and annotate the -->
<!-- MS1 scans with the number of PSMs using the `countIdentifications()` -->
<!-- function, as shown above. The function will automatically perform the -->
<!-- counts in parallel for each acquisition. -->
<!-- 

<p class="question-end"><span class="fa fa-square-o solution-icon"></span></p></div></div>

 -->
<!-- 

<div class = "msmb-solution"><p class="solution-begin">&#x25BA; Solution<span id='sol-start-24' class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-24', 'sol-start-24')"></span></p><div class="solution-body" id = "sol-body-24" style="display: none;">

 -->
<!-- ```{r} -->
<!-- sp <- countIdentifications(sp) -->
<!-- table(msLevel(sp), sp$countIdentifications) -->
<!-- ``` -->
<!-- ```{r tic_PXD022816} -->
<!-- sp |> -->
<!--  filterMsLevel(1) |> -->
<!--  spectraData() |> -->
<!--  as_tibble() |> -->
<!--  ggplot(aes(x = rtime, -->
<!--             y = totIonCurrent)) + -->
<!--      geom_line(alpha = 0.25) + -->
<!--      geom_point(aes(colour = ifelse(countIdentifications == 0, -->
<!--                                     NA, countIdentifications)), -->
<!--                 size = 0.75, -->
<!--                 alpha = 0.5) + -->
<!--      scale_colour_gradient(low = "orange", high = "red") + -->
<!--      facet_grid(sub("^.+_", "", basename(dataOrigin)) ~ .) + -->
<!--      labs(colour = "Number of ids") -->
<!-- ``` -->
<!-- 

<p class="solution-end"><span class="fa fa-square-o solution-icon"></span></p></div></div>

 -->
</div>
<div id="exploration-and-assessment-of-identifications-using-msnid" class="section level2" number="4.8">
<h2>
<span class="header-section-number">4.8</span> Exploration and Assessment of Identifications using <code>MSnID</code>
<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('exploration-and-assessment-of-identifications-using-msnid')" onmouseout="reset_tooltip('exploration-and-assessment-of-identifications-using-msnid-tooltip')"><span class="tooltiptext" id="exploration-and-assessment-of-identifications-using-msnid-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>The <code>MSnID</code> package extracts MS/MS ID data from mzIdentML (leveraging
the <code>mzID</code> package) or text files. After collating the search results
from multiple datasets it assesses their identification quality and
optimises filtering criteria to achieve the maximum number of
identifications while not exceeding a specified false discovery
rate. It also contains a number of utilities to explore the MS/MS
results and assess missed and irregular enzymatic cleavages, mass
measurement accuracy, etc.</p>
<div id="step-by-step-work-flow" class="section level3" number="4.8.1">
<h3>
<span class="header-section-number">4.8.1</span> Step-by-step work-flow<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('step-by-step-work-flow')" onmouseout="reset_tooltip('step-by-step-work-flow-tooltip')"><span class="tooltiptext" id="step-by-step-work-flow-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Let’s reproduce parts of the analysis described the <code>MSnID</code>
vignette. You can explore more with</p>
<div class="sourceCode" id="cb204"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb204-1"><a href="sec-id.html#cb204-1" aria-hidden="true" tabindex="-1"></a><span class="fu">vignette</span>(<span class="st">"msnid_vignette"</span>, <span class="at">package =</span> <span class="st">"MSnID"</span>)</span></code></pre></div>
<p>The <em><a href="https://bioconductor.org/packages/3.21/MSnID">MSnID</a></em> package can be used for post-search filtering
of MS/MS identifications. One starts with the construction of an
<code>MSnID</code> object that is populated with identification results that can
be imported from a <code>data.frame</code> or from <code>mzIdenML</code> files. Here, we
will use the example identification data provided with the package.</p>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="sec-id.html#cb205-1" aria-hidden="true" tabindex="-1"></a>mzids <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"c_elegans.mzid.gz"</span>, <span class="at">package=</span><span class="st">"MSnID"</span>)</span>
<span id="cb205-2"><a href="sec-id.html#cb205-2" aria-hidden="true" tabindex="-1"></a><span class="fu">basename</span>(mzids)</span></code></pre></div>
<pre><code>## [1] "c_elegans.mzid.gz"</code></pre>
<p>We start by loading the package, initialising the <code>MSnID</code> object, and
add the identification result from our <code>mzid</code> file (there could of
course be more than one).</p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb207-1"><a href="sec-id.html#cb207-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"MSnID"</span>)</span>
<span id="cb207-2"><a href="sec-id.html#cb207-2" aria-hidden="true" tabindex="-1"></a>msnid <span class="ot">&lt;-</span> <span class="fu">MSnID</span>(<span class="st">"."</span>)</span></code></pre></div>
<pre><code>## Note, the anticipated/suggested columns in the
## peptide-to-spectrum matching results are:
## -----------------------------------------------
## accession
## calculatedMassToCharge
## chargeState
## experimentalMassToCharge
## isDecoy
## peptide
## spectrumFile
## spectrumID</code></pre>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb209-1"><a href="sec-id.html#cb209-1" aria-hidden="true" tabindex="-1"></a>msnid <span class="ot">&lt;-</span> <span class="fu">read_mzIDs</span>(msnid, mzids)</span></code></pre></div>
<pre><code>## Loaded cached data</code></pre>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="sec-id.html#cb211-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show</span>(msnid)</span></code></pre></div>
<pre><code>## MSnID object
## Working directory: "."
## #Spectrum Files:  1 
## #PSMs: 12263 at 36 % FDR
## #peptides: 9489 at 44 % FDR
## #accessions: 7414 at 76 % FDR</code></pre>
<p>Printing the <code>MSnID</code> object returns some basic information such as</p>
<ul>
<li>Working directory.</li>
<li>Number of spectrum files used to generate data.</li>
<li>Number of peptide-to-spectrum matches and corresponding FDR.</li>
<li>Number of unique peptide sequences and corresponding FDR.</li>
<li>Number of unique proteins or amino acid sequence accessions and corresponding FDR.</li>
</ul>
<p>The package then enables to define, optimise and apply filtering based
for example on missed cleavages, identification scores, precursor mass
errors, etc. and assess PSM, peptide and protein FDR levels. To
properly function, it expects to have access to the following data</p>
<pre><code>## [1] "accession"                "calculatedMassToCharge"  
## [3] "chargeState"              "experimentalMassToCharge"
## [5] "isDecoy"                  "peptide"                 
## [7] "spectrumFile"             "spectrumID"</code></pre>
<p>which are indeed present in our data:</p>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb214-1"><a href="sec-id.html#cb214-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(msnid)</span></code></pre></div>
<pre><code>##  [1] "spectrumID"                "scan number(s)"           
##  [3] "acquisitionNum"            "passThreshold"            
##  [5] "rank"                      "calculatedMassToCharge"   
##  [7] "experimentalMassToCharge"  "chargeState"              
##  [9] "MS-GF:DeNovoScore"         "MS-GF:EValue"             
## [11] "MS-GF:PepQValue"           "MS-GF:QValue"             
## [13] "MS-GF:RawScore"            "MS-GF:SpecEValue"         
## [15] "AssumedDissociationMethod" "IsotopeError"             
## [17] "isDecoy"                   "post"                     
## [19] "pre"                       "end"                      
## [21] "start"                     "accession"                
## [23] "length"                    "description"              
## [25] "pepSeq"                    "modified"                 
## [27] "modification"              "idFile"                   
## [29] "spectrumFile"              "databaseFile"             
## [31] "peptide"</code></pre>
<p>Here, we summarise a few steps and redirect the reader to the
package’s vignette for more details:</p>
</div>
<div id="analysis-of-peptide-sequences" class="section level3" number="4.8.2">
<h3>
<span class="header-section-number">4.8.2</span> Analysis of peptide sequences<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('analysis-of-peptide-sequences')" onmouseout="reset_tooltip('analysis-of-peptide-sequences-tooltip')"><span class="tooltiptext" id="analysis-of-peptide-sequences-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Cleaning irregular cleavages at the termini of the peptides and
missing cleavage site within the peptide sequences. The following two
function calls create the new <code>numMisCleavages</code> and <code>numIrregCleavages</code>
columns in the <code>MSnID</code> object</p>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="sec-id.html#cb216-1" aria-hidden="true" tabindex="-1"></a>msnid <span class="ot">&lt;-</span> <span class="fu">assess_termini</span>(msnid, <span class="at">validCleavagePattern=</span><span class="st">"[KR]</span><span class="sc">\\</span><span class="st">.[^P]"</span>)</span>
<span id="cb216-2"><a href="sec-id.html#cb216-2" aria-hidden="true" tabindex="-1"></a>msnid <span class="ot">&lt;-</span> <span class="fu">assess_missed_cleavages</span>(msnid, <span class="at">missedCleavagePattern=</span><span class="st">"[KR](?=[^P$])"</span>)</span></code></pre></div>
</div>
<div id="trimming-the-data" class="section level3" number="4.8.3">
<h3>
<span class="header-section-number">4.8.3</span> Trimming the data<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('trimming-the-data')" onmouseout="reset_tooltip('trimming-the-data-tooltip')"><span class="tooltiptext" id="trimming-the-data-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Now, we can use the <code>apply_filter</code> function to effectively apply
filters. The strings passed to the function represent expressions that
will be evaluated, thus keeping only PSMs that have 0 irregular
cleavages and 2 or less missed cleavages.</p>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb217-1"><a href="sec-id.html#cb217-1" aria-hidden="true" tabindex="-1"></a>msnid <span class="ot">&lt;-</span> <span class="fu">apply_filter</span>(msnid, <span class="st">"numIrregCleavages == 0"</span>)</span>
<span id="cb217-2"><a href="sec-id.html#cb217-2" aria-hidden="true" tabindex="-1"></a>msnid <span class="ot">&lt;-</span> <span class="fu">apply_filter</span>(msnid, <span class="st">"numMissCleavages &lt;= 2"</span>)</span>
<span id="cb217-3"><a href="sec-id.html#cb217-3" aria-hidden="true" tabindex="-1"></a><span class="fu">show</span>(msnid)</span></code></pre></div>
<pre><code>## MSnID object
## Working directory: "."
## #Spectrum Files:  1 
## #PSMs: 7838 at 17 % FDR
## #peptides: 5598 at 23 % FDR
## #accessions: 3759 at 53 % FDR</code></pre>
</div>
<div id="parent-ion-mass-errors" class="section level3" number="4.8.4">
<h3>
<span class="header-section-number">4.8.4</span> Parent ion mass errors<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('parent-ion-mass-errors')" onmouseout="reset_tooltip('parent-ion-mass-errors-tooltip')"><span class="tooltiptext" id="parent-ion-mass-errors-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Using <code>"calculatedMassToCharge"</code> and <code>"experimentalMassToCharge"</code>, the
<code>mass_measurement_error</code> function calculates the parent ion mass
measurement error in parts per million.</p>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb219-1"><a href="sec-id.html#cb219-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">mass_measurement_error</span>(msnid))</span></code></pre></div>
<pre><code>##       Min.    1st Qu.     Median       Mean    3rd Qu.       Max. 
## -2184.0640    -0.6992     0.0000    17.6146     0.7511  2012.5178</code></pre>
<p>We then filter any matches that do not fit the +/- 20 ppm tolerance</p>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb221-1"><a href="sec-id.html#cb221-1" aria-hidden="true" tabindex="-1"></a>msnid <span class="ot">&lt;-</span> <span class="fu">apply_filter</span>(msnid, <span class="st">"abs(mass_measurement_error(msnid)) &lt; 20"</span>)</span>
<span id="cb221-2"><a href="sec-id.html#cb221-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">mass_measurement_error</span>(msnid))</span></code></pre></div>
<pre><code>##     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
## -19.7797  -0.5866   0.0000  -0.2970   0.5713  19.6758</code></pre>
</div>
<div id="filtering-criteria" class="section level3" number="4.8.5">
<h3>
<span class="header-section-number">4.8.5</span> Filtering criteria<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('filtering-criteria')" onmouseout="reset_tooltip('filtering-criteria-tooltip')"><span class="tooltiptext" id="filtering-criteria-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Filtering of the identification data will rely on</p>
<ul>
<li>-log10 transformed MS-GF+ Spectrum E-value, reflecting the goodness
of match between experimental and theoretical fragmentation patterns</li>
</ul>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb223-1"><a href="sec-id.html#cb223-1" aria-hidden="true" tabindex="-1"></a>msnid<span class="sc">$</span>msmsScore <span class="ot">&lt;-</span> <span class="sc">-</span><span class="fu">log10</span>(msnid<span class="sc">$</span><span class="st">`</span><span class="at">MS-GF:SpecEValue</span><span class="st">`</span>)</span></code></pre></div>
<ul>
<li>the absolute mass measurement error (in ppm units) of the parent ion</li>
</ul>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb224-1"><a href="sec-id.html#cb224-1" aria-hidden="true" tabindex="-1"></a>msnid<span class="sc">$</span>absParentMassErrorPPM <span class="ot">&lt;-</span> <span class="fu">abs</span>(<span class="fu">mass_measurement_error</span>(msnid))</span></code></pre></div>
</div>
<div id="setting-filters" class="section level3" number="4.8.6">
<h3>
<span class="header-section-number">4.8.6</span> Setting filters<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('setting-filters')" onmouseout="reset_tooltip('setting-filters-tooltip')"><span class="tooltiptext" id="setting-filters-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>MS2 filters are handled by a special <code>MSnIDFilter</code> class objects, where
individual filters are set by name (that is present in <code>names(msnid)</code>)
and comparison operator (&gt;, &lt;, = , …) defining if we should retain
hits with higher or lower given the threshold and finally the
threshold value itself.</p>
<div class="sourceCode" id="cb225"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb225-1"><a href="sec-id.html#cb225-1" aria-hidden="true" tabindex="-1"></a>filtObj <span class="ot">&lt;-</span> <span class="fu">MSnIDFilter</span>(msnid)</span>
<span id="cb225-2"><a href="sec-id.html#cb225-2" aria-hidden="true" tabindex="-1"></a>filtObj<span class="sc">$</span>absParentMassErrorPPM <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">comparison=</span><span class="st">"&lt;"</span>, <span class="at">threshold=</span><span class="fl">10.0</span>)</span>
<span id="cb225-3"><a href="sec-id.html#cb225-3" aria-hidden="true" tabindex="-1"></a>filtObj<span class="sc">$</span>msmsScore <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">comparison=</span><span class="st">"&gt;"</span>, <span class="at">threshold=</span><span class="fl">10.0</span>)</span>
<span id="cb225-4"><a href="sec-id.html#cb225-4" aria-hidden="true" tabindex="-1"></a><span class="fu">show</span>(filtObj)</span></code></pre></div>
<pre><code>## MSnIDFilter object
## (absParentMassErrorPPM &lt; 10) &amp; (msmsScore &gt; 10)</code></pre>
<p>We can then evaluate the filter on the identification data object,
which returns the false discovery rate and number of retained
identifications for the filtering criteria at hand.</p>
<div class="sourceCode" id="cb227"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb227-1"><a href="sec-id.html#cb227-1" aria-hidden="true" tabindex="-1"></a><span class="fu">evaluate_filter</span>(msnid, filtObj)</span></code></pre></div>
<pre><code>##           fdr    n
## PSM         0 3807
## peptide     0 2455
## accession   0 1009</code></pre>
</div>
<div id="filter-optimisation" class="section level3" number="4.8.7">
<h3>
<span class="header-section-number">4.8.7</span> Filter optimisation<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('filter-optimisation')" onmouseout="reset_tooltip('filter-optimisation-tooltip')"><span class="tooltiptext" id="filter-optimisation-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Rather than setting filtering values by hand, as shown above, these
can be set automatically to meet a specific false discovery rate.</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb229-1"><a href="sec-id.html#cb229-1" aria-hidden="true" tabindex="-1"></a>filtObj.grid <span class="ot">&lt;-</span> <span class="fu">optimize_filter</span>(filtObj, msnid, <span class="at">fdr.max=</span><span class="fl">0.01</span>,</span>
<span id="cb229-2"><a href="sec-id.html#cb229-2" aria-hidden="true" tabindex="-1"></a>                                <span class="at">method=</span><span class="st">"Grid"</span>, <span class="at">level=</span><span class="st">"peptide"</span>,</span>
<span id="cb229-3"><a href="sec-id.html#cb229-3" aria-hidden="true" tabindex="-1"></a>                                <span class="at">n.iter=</span><span class="dv">500</span>)</span>
<span id="cb229-4"><a href="sec-id.html#cb229-4" aria-hidden="true" tabindex="-1"></a><span class="fu">show</span>(filtObj.grid)</span></code></pre></div>
<pre><code>## MSnIDFilter object
## (absParentMassErrorPPM &lt; 3) &amp; (msmsScore &gt; 7.4)</code></pre>
<div class="sourceCode" id="cb231"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb231-1"><a href="sec-id.html#cb231-1" aria-hidden="true" tabindex="-1"></a><span class="fu">evaluate_filter</span>(msnid, filtObj.grid)</span></code></pre></div>
<pre><code>##                   fdr    n
## PSM       0.004097561 5146
## peptide   0.006447651 3278
## accession 0.021996616 1208</code></pre>
<p>Filters can eventually be applied (rather than just evaluated) using
the <code>apply_filter</code> function.</p>
<div class="sourceCode" id="cb233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb233-1"><a href="sec-id.html#cb233-1" aria-hidden="true" tabindex="-1"></a>msnid <span class="ot">&lt;-</span> <span class="fu">apply_filter</span>(msnid, filtObj.grid)</span>
<span id="cb233-2"><a href="sec-id.html#cb233-2" aria-hidden="true" tabindex="-1"></a><span class="fu">show</span>(msnid)</span></code></pre></div>
<pre><code>## MSnID object
## Working directory: "."
## #Spectrum Files:  1 
## #PSMs: 5146 at 0.41 % FDR
## #peptides: 3278 at 0.64 % FDR
## #accessions: 1208 at 2.2 % FDR</code></pre>
<p>And finally, identifications that matched decoy and contaminant
protein sequences are removed</p>
<div class="sourceCode" id="cb235"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb235-1"><a href="sec-id.html#cb235-1" aria-hidden="true" tabindex="-1"></a>msnid <span class="ot">&lt;-</span> <span class="fu">apply_filter</span>(msnid, <span class="st">"isDecoy == FALSE"</span>)</span>
<span id="cb235-2"><a href="sec-id.html#cb235-2" aria-hidden="true" tabindex="-1"></a>msnid <span class="ot">&lt;-</span> <span class="fu">apply_filter</span>(msnid, <span class="st">"!grepl('Contaminant',accession)"</span>)</span>
<span id="cb235-3"><a href="sec-id.html#cb235-3" aria-hidden="true" tabindex="-1"></a><span class="fu">show</span>(msnid)</span></code></pre></div>
<pre><code>## MSnID object
## Working directory: "."
## #Spectrum Files:  1 
## #PSMs: 5117 at 0 % FDR
## #peptides: 3251 at 0 % FDR
## #accessions: 1179 at 0 % FDR</code></pre>
</div>
<div id="export-msnid-data" class="section level3" number="4.8.8">
<h3>
<span class="header-section-number">4.8.8</span> Export <code>MSnID</code> data<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('export-msnid-data')" onmouseout="reset_tooltip('export-msnid-data-tooltip')"><span class="tooltiptext" id="export-msnid-data-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>The resulting filtered identification data can be exported to a
<code>data.frame</code> (or to a dedicated <code>MSnSet</code> data structure from the
<code>MSnbase</code> package) for quantitative MS data, described below, and
further processed and analysed using appropriate statistical tests.</p>
<div class="sourceCode" id="cb237"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb237-1"><a href="sec-id.html#cb237-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">psms</span>(msnid))</span></code></pre></div>
<pre><code>##   spectrumID scan number(s) acquisitionNum passThreshold rank
## 1 index=7151           8819           7151          TRUE    1
## 2 index=8520          10419           8520          TRUE    1
##   calculatedMassToCharge experimentalMassToCharge chargeState MS-GF:DeNovoScore
## 1               1270.318                 1270.318           3               287
## 2               1426.737                 1426.739           3               270
##   MS-GF:EValue MS-GF:PepQValue MS-GF:QValue MS-GF:RawScore MS-GF:SpecEValue
## 1 1.709082e-24               0            0            239     1.007452e-31
## 2 3.780745e-24               0            0            230     2.217275e-31
##   AssumedDissociationMethod IsotopeError isDecoy post pre end start accession
## 1                       CID            0   FALSE    A   K 283   249   CE02347
## 2                       CID            0   FALSE    A   K 182   142   CE07055
##   length
## 1    393
## 2    206
##                                                                                                                           description
## 1 WBGene00001993; locus:hpd-1; 4-hydroxyphenylpyruvate dioxygenase; status:Confirmed; UniProt:Q22633; protein_id:CAA90315.1; T21C12.2
## 2           WBGene00001755; locus:gst-7; glutathione S-transferase; status:Confirmed; UniProt:P91253; protein_id:AAB37846.1; F11G11.2
##                                      pepSeq modified modification
## 1       AISQIQEYVDYYGGSGVQHIALNTSDIITAIEALR    FALSE         &lt;NA&gt;
## 2 SAGSGYLVGDSLTFVDLLVAQHTADLLAANAALLDEFPQFK    FALSE         &lt;NA&gt;
##              idFile                                   spectrumFile
## 1 c_elegans.mzid.gz c_elegans_A_3_1_21Apr10_Draco_10-03-04_dta.txt
## 2 c_elegans.mzid.gz c_elegans_A_3_1_21Apr10_Draco_10-03-04_dta.txt
##               databaseFile                                       peptide
## 1 ID_004174_E48C5B52.fasta       K.AISQIQEYVDYYGGSGVQHIALNTSDIITAIEALR.A
## 2 ID_004174_E48C5B52.fasta K.SAGSGYLVGDSLTFVDLLVAQHTADLLAANAALLDEFPQFK.A
##   numIrregCleavages numMissCleavages msmsScore absParentMassErrorPPM
## 1                 0                0  30.99678             0.3843772
## 2                 0                0  30.65418             1.3689451
##  [ reached 'max' / getOption("max.print") -- omitted 4 rows ]</code></pre>

</div>
</div>
</div>
<p style="text-align: center;">
<a href="sec-raw.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-quant.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2025-06-05
 using 
R version 4.5.0 (2025-04-11)
</p>
</div>
</div>



</body>
</html>
